{% extends "base.html" %}
{% block title %}Mapping Results{% endblock %}
{% block css %}
    <link rel="stylesheet" type="text/css" href="{{ url_for('css', filename='DataTables/css/jquery.dataTables.css') }}" />
    <link rel="stylesheet" type="text/css" href="/static/packages/DT_bootstrap/DT_bootstrap.css" />
    <link rel="stylesheet" href="https://use.fontawesome.com/releases/v5.3.1/css/all.css" integrity="sha384-mzrmE5qonljUremFsqc01SB46JvROS7bZs3IO2EmfFsd15uHvIt+Y8vEf7N7fWAU" crossorigin="anonymous">
    <link rel="stylesheet" type="text/css" href="/static/packages/purescript_genome_browser/css/purescript-genetics-browser.css" />

    <link rel="stylesheet" type="text/css" href="/static/new/css/marker_regression.css" />

{% endblock %}
{% from "base_macro.html" import header %}
{% block content %}
    <div class="container" style="min-width: 900px;">
        <form method="post" target="_blank" action="/run_mapping" name="marker_regression" id="marker_regression_form">
        <input type="hidden" name="temp_uuid" value="{{ temp_uuid }}">
        {% if temp_trait is defined %}
        <input type="hidden" name="temp_trait" value="{{ temp_trait }}">
        <input type="hidden" name="group" value="{{ group }}">
        {% endif %}
        <input type="hidden" name="trait_id" value="{{ this_trait.name }}">
        <input type="hidden" name="dataset" value="{{ dataset.name }}">
        <input type="hidden" name="genofile" value="{{ genofile_string }}">
        <input type="hidden" name="geno_db_exists" value="{{ geno_db_exists }}">
        <input type="hidden" name="first_run" value="{{ first_run }}">
        {% if output_files is defined %}
        <input type="hidden" name="output_files" value="{{ output_files }}">
        {% endif %}
        {% if reaper_version is defined %}
        <input type="hidden" name="reaper_version" value="{{ reaper_version }}">
        {% endif %}
        <input type="hidden" name="results_path" value="{{ mapping_results_path }}">
        <input type="hidden" name="method" value="{{ mapping_method }}">
        <input type="hidden" name="samples" value="{{ samples|join(",") }}">
        {% for sample in samples %}
        <input type="hidden" name="value:{{ sample }}" value="{{ vals[loop.index - 1] }}">
        {% endfor %}
        <input type="hidden" name="n_samples" value="{{ n_samples }}">
        <input type="hidden" name="maf" value="{{ maf }}">
        <input type="hidden" name="use_loco" value="{{ use_loco }}">
        <input type="hidden" name="selected_chr" value="{{ selectedChr }}">
        <input type="hidden" name="manhattan_plot" value="{{ manhattan_plot }}">
        {% if manhattan_plot == True %}
        <input type="hidden" name="color_scheme" value="alternating">
        {% endif %}
        <input type="hidden" name="num_perm" value="{{ nperm }}">
        <input type="hidden" name="perm_info" value="">
        <input type="hidden" name="perm_strata" value="{{ perm_strata }}">
        <input type="hidden" name="num_bootstrap" value="{{ nboot }}">
        <input type="hidden" name="do_control" value="{{ doControl }}">
        <input type="hidden" name="control_marker" value="{{ controlLocus }}">
        <input type="hidden" name="covariates" value="{{ covariates }}">
        <input type="hidden" name="mapmethod_rqtl_geno" value="{{ mapmethod_rqtl_geno }}">
        <input type="hidden" name="mapmodel_rqtl_geno" value="{{ mapmodel_rqtl_geno }}">
        <input type="hidden" name="pair_scan" value="{{ pair_scan }}">
        <input type="hidden" name="transform" value="{{ transform }}">
        <input type="hidden" name="tool_used" value="Mapping">
        <input type="hidden" name="wanted_inputs" value="">
        <input type="hidden" name="form_url" value="/run_mapping">
        
        <div class="container" style="min-width: 1400px;">
          <div class="col-xs-5">
              <h2>Map Viewer: Whole Genome</h2><br>
              <b>Population:</b> {{ dataset.group.species|capitalize }} {{ dataset.group.name }}<br>
              <b>Database:</b> {{ dataset.fullname }}<br>
              {% if dataset.type == "ProbeSet" %}<b>Trait ID:</b>{% else %}<b>Record ID:</b>{% endif %} <a href="/show_trait?trait_id={{ this_trait.name }}&dataset={{ dataset.name }}">{{ this_trait.name }}</a><br>
              {% if dataset.type == "ProbeSet" %}
              <b>Gene Symbol:</b> <i>{{ this_trait.symbol }}</i><br>
              <b>Location:</b> Chr {{ this_trait.chr }} @ {{ this_trait.mb }} Mb<br>
              {% endif %}
              {% if genofile_string is defined %}
              <b>Genotypes:</b> {{ genofile_string.split(":")[1] }}
              {% endif %}
              <br>
              <br>
              <a id="export_mapping_results" href="#" target="_blank" >Download Full Results</a>
          </div>
          <div id="gn1_map_options" class="col-xs-5" style="outline: 3px double #AAAAAA; padding: 10px; margin: 10px;">
            <div class="col-xs-8" style="padding: 0px;">
              <table>
                <tr>
                  <td><b>Chr:&nbsp;</b></td>
                  <td style="padding: 5px;">
                    <select name="chromosomes" size="1">
                        {% for chr in ChrList %}
                        <option value="{{ chr[1] }}" {% if (chr[1] + 1) == selectedChr %}selected{% endif %}>{{ chr[0] }}</option>
                        {% endfor %}
                    </select>&nbsp;
                    <button type="button" class="btn btn-primary" style="padding-bottom: 2px; padding-top: 2px;" onclick="javascript:remap();">Remap</button>
                  </td>
                </tr>
                <tr>
                  <td ><b>View:&nbsp;</b></td>
                  <td style="padding: 5px;">
                    <input type="text" name="startMb" size="7" value="{% if startMb != -1 %}{{ startMb }}{% endif %}"> to <input type="text" name="endMb" size="7" value="{% if endMb != -1 %}{{ endMb }}{% endif %}">
                  </td>
                </tr>
                <tr>
                  <td><b>Units:&nbsp;</b></td>
                  <td style="padding: 5px;">
                    <label class="radio-inline">
                      <input type="radio" name="LRSCheck" value="LRS" {% if LRS_LOD == "LRS" %}checked{% endif %}>LRS
                    </label>
                    <label class="radio-inline">
                      <input type="radio" name="LRSCheck" value="{% if LRS_LOD == "-logP" %}-logP{% else %}LOD{% endif %}" {% if LRS_LOD == "LOD" or LRS_LOD == "-logP" %}checked{% endif %}>{% if LRS_LOD == "-logP" %}-logP{% else %}LOD{% endif %}
                    </label>
                    <a href="http://genenetwork.org/glossary.html#LOD" target="_blank">
                      <sup style="color:#f00"> ?</sup>
                    </a>
                  </td>
                </tr>
                <tr>
                  <td></td>
                  <td style="padding: 5px;">
                    <input type="text" name="lrsMax" value="{{ '%0.1f' | format(lrsMax|float) }}" size="3"> <span style="font-size: 12px;">units on the y-axis (0 for default)</span>   
                  </td>
                </tr>
                <tr>
                  <td><b>Width:&nbsp;</b></td>
                  <td style="padding: 5px;">
                    <input type="text" name="graphWidth" value="{% if graphWidth is defined %}{{ graphWidth }}{% else %}1600{% endif %}" size="5"><span style="font-size: 12px;"> pixels (minimum=900)</span>
                  </td>
                </tr>
              </table>
              {% if manhattan_plot == True and selectedChr == -1 %}
              <table style="margin-top: 10px;">
                <tr>
                  <td>
                    <b>Manhattan Plot Color Scheme:&nbsp;</b>
                  </td>
                  <td>
                    <select id="color_scheme">
                      <option value="alternating" {% if color_scheme == "alternating" %}selected{% endif %}>Alternating</option>
                      <option value="varied" {% if color_scheme == "varied" %}selected{% endif %}>Varied by Chr</option>
                      <option value="single" {% if color_scheme == "single" %}selected{% endif %}>Single Color</option>
                    </select>
                  </td>
                  <td>
                    <input name="manhattan_single_color" type="hidden" id="point_color" value={% if manhattan_single_color %}{{ manhattan_single_color }}{% else %}"#D9D9D9"{% endif %}>
                    <button style="display: none; margin-left: 5px;" id="point_color_picker" class="jscolor {valueElement: 'point_color'}">Choose Color</button>
                  </td>
                </tr>
              </table>
              {% endif %}
            </div>
            <div class="col-xs-4" style="padding: 0px;">
                  {% if (mapping_method == "reaper" or mapping_method == "rqtl_geno") and nperm > 0 %}
                  <input type="checkbox" name="permCheck" class="checkbox" style="display: inline; margin-top: 0px;" {% if permChecked|upper == "ON" %}value="ON" checked{% endif %}> <span style="font-size: 12px;">Permutation Test 
                    <a href="http://genenetwork.org/glossary.html#Permutation" target="_blank">
                      <sup style="color:#f00"> ?</sup>
                    </a>
                  <br>
                  {% endif %}
                  {% if mapping_method == "reaper" and nboot > 0 %}
                  <input type="checkbox" name="bootCheck" class="checkbox" style="display: inline; margin-top: 0px;" {% if bootChecked|upper == "ON" %}value="ON" checked{% endif %}> <span style="font-size: 12px;">Bootstrap Test 
                    <a href="http://genenetwork.org/glossary.html#bootstrap" target="_blank">
                      <sup style="color:#f00"> ?</sup>
                    </a>
                  <br>
                  {% endif %}
                  {% if mapping_method == "reaper" %}
                  <input type="checkbox" name="additiveCheck" class="checkbox" style="display: inline; margin-top: 0px;" {% if additiveChecked|upper == "ON" %}value="ON" checked{% endif %}> <span style="font-size: 12px;">Allele Effects
                    <a href="http://genenetwork.org/glossary.html#additive" target="_blank">
                      <sup style="color:#f00"> ?</sup>
                    </a>
                  <br>
                  {% endif %}
                  <input type="checkbox" name="showSNP" class="checkbox" style="display: inline; margin-top: 0px;" {% if SNPChecked|upper == "ON" %}value="ON" checked{% endif %}> <span style="font-size: 12px;">SNP Track </span>
                    <a href="http://genenetwork.org/glossary.html#snpSeismograph" target="_blank">
                      <sup style="color:#f00"> ?</sup>
                    </a>
                    <span style="color:red;">*</span>
                  <br>
                  <input type="checkbox" name="showGenes" class="checkbox" style="display: inline; margin-top: 0px;" {% if geneChecked|upper == "ON" %}value="ON" checked{% endif %}> <span style="font-size: 12px;">Gene Track </span> <span style="color:red;">*</span><br>
                  {% if plotScale != "morgan" %}
                  <input type="checkbox" name="haplotypeAnalystCheck" class="checkbox" style="display: inline; margin-top: 0px;" {% if haplotypeAnalystChecked|upper == "ON" %}value="ON" checked{% endif %}> <span style="font-size: 12px;">Haplotype Analyst </span> <span style="color:red;">*</span><br>
                  {% endif %}
                  <input type="checkbox" name="viewLegend" class="checkbox" style="display: inline; margin-top: 0px;" {% if legendChecked|upper == "ON" %}value="ON" checked{% endif %}> <span style="font-size: 12px;">Legend </span>
            </div>
            <div class="col-xs-12" align="center" style="padding: 5px;">
              <span style="color:red;">*</span> <span style="font-size: 12px;">only apply to single chromosome physical mapping</span>
            </div>
          </div>
        </div>
 
        <div class="tabbable" style="margin: 10px;">
          <ul class="nav nav-tabs">
            <li id="gn1_map_tab">
              <a href="#gn1_map" data-toggle="tab" aria-expanded="true">GN1 Map</a>
            </li>
            {% if mapping_method == "gemma" or mapping_method == "reaper" %}
            <li id="browser_tab">
                <a href="#browser_holder" data-toggle="tab" aria-expanded="true">Genome Browser</a>
            </li>
            {% endif %}
          </ul>
          <div class="tab-content" style="overflow-x: auto;">
            <div class="tab-pane active" id="gn1_map">
              <div class="qtlcharts">
                  {{ gifmap|safe }}
                  <img src="/generated/{{ filename }}.png" usemap="#WebQTLImageMap">
                  {% if additiveChecked|upper == "ON" %}
                  <br>
                  <span style="white-space: nowrap;">A positive additive coefficient (green line) indicates that {{ dataset.group.parlist[1] }} alleles increase trait values. In contrast, a negative additive coefficient (orange line) indicates that {{ dataset.group.parlist[0] }} alleles increase trait values.</span>
                  {% endif %}
              </div>
            </div>
            {% if mapping_method == "gemma" or mapping_method == "reaper" %}
            <div class="tab-pane" id="browser_holder" style="height: 600px;">
                <div id="browser" style="margin-right: 20px; width: 90%;">
                    <div id="controls">
                        <button id="scrollLeft"  type="button" >
                          <i class="fas fa-arrow-left"></i>
                        </button>
                        <button id="scrollRight" type="button" >
                          <i class="fas fa-arrow-right"></i>
                        </button>
                        <button id="zoomOut" type="button" >
                          <i class="fas fa-search-minus"></i>
                        </button>
                        <button id="zoomIn"  type="button" >
                          <i class="fas fa-search-plus"></i>
                        </button>
                        <button id="reset"  type="button" >Reset</button>
                    </div>
                    <div id="infoBox"></div>
                </div>
            </div>
            {% endif %}
          </div>
        </div>
        <div>
            {% if nperm > 0 and permChecked == "ON" %}
            <br><br>
            <div id="perm_histogram" class="barchart"></div>
            <!--<img src="/generated/{{ perm_filename }}.gif">-->
            <br><br>
            Total of {{ nperm }} permutations&nbsp;&nbsp;<a href="javascript:;" id="download_perm" target="_blank" >Download Permutation Results</a>
            <br>
            {% endif %}
        </div>

        </form>
        {% if selectedChr == -1 %}
        <div class="container" style="padding-left: 30px; margin-top: 50px; position: relative;">
          <h2>Mapping Statistics</h2>
        <br />
        <button class="btn btn-default" id="select_all"><span class="glyphicon glyphicon-ok"></span> Select All</button>
        <button class="btn btn-default" id="deselect_all"><span class="glyphicon glyphicon-remove"></span> Deselect All</button>
        <button class="btn btn-default" id="invert"><span class="glyphicon glyphicon-resize-vertical"></span> Invert</button>
        <button class="btn btn-success" id="add" disabled><span class="glyphicon glyphicon-plus-sign"></span> Add</button>
        <br />
        <br />
          <div id="table_container" style="width:{% if 'additive' in trimmed_markers[0] %}600{% else %}550{% endif %}px;">
            <table id="trait_table" class="table-hover table-striped cell-border dataTable no-footer">
              <thead>
                <tr>
                  <th></th>
                  <th>Row</th>
                  <th>Marker</th>
                  {% if LRS_LOD == "-logP" %}
                  <th><div style="text-align: right;">–logP</div></th>
                  {% else %}
                  <th><div style="text-align: right;">{{ LRS_LOD }}</div></th>
                  {% endif %}
                  <th><div style="text-align: right;">Position ({% if plotScale == "physic" %}Mb{% else %}cM{% endif %})</div></th>
                  {% if 'additive' in trimmed_markers[0] %}
                  <th><div style="text-align: right;">Add Eff</div></th>
                  {% endif %}
                  {% if 'dominance' in trimmed_markers[0] and dataset.group.genetic_type != "riset" %}
                  <th><div style="text-align: right;">Dom Eff</div></th>
                  {% endif %}
                </tr>
              </thead>
              <tbody>
                {% for marker in trimmed_markers %}
                <tr>
                  <td align="center" style="padding: 1px 0px 1px 0px;">
                    <input type="checkbox" name="selectCheck"
                           class="checkbox trait_checkbox"
                           value="{{ data_hmac('{}:{}Geno'.format(marker.name, dataset.group.name)) }}">
                  </td>
                  <td align="right">{{ loop.index }}</td>
                  <td>{% if geno_db_exists == "True" %}<a href="/show_trait?trait_id={{ marker.name }}&dataset={{ dataset.group.name }}Geno">{{ marker.name }}</a>{% else %}{{ marker.name }}{% endif %}</td>
                  {% if LRS_LOD == "LOD" or LRS_LOD == "-logP" %}
                  {% if 'lod_score' in marker %}
                  <td align="right">{{ '%0.2f' | format(marker.lod_score|float) }}</td>
                  {% else %}
                  <td align="right">{{ '%0.2f' | format(marker.lrs_value|float / 4.61) }}</td>
                  {% endif %}
                  {% else %}
                  {% if 'lod_score' in marker %}
                  <td align="right">{{ '%0.2f' | format(marker.lod_score|float * 4.61) }}</td>
                  {% else %}
                  <td align="right">{{ '%0.2f' | format(marker.lrs_value|float) }}</td>
                  {% endif %}
                  {% endif %}
                  <td align="right">{{ marker.display_pos }}</td>
                  {% if 'additive' in marker %}
                  <td align="right">{{ '%0.3f' | format(marker.additive|float) }}</td>
                  {% endif %}
                  {% if 'dominance' in marker and dataset.group.genetic_type != "riset" %}
                  <td align="right">{{ '%0.2f' | format(marker.dominance|float) }}</td>
                  {% endif %}
                </tr>
                {% endfor %}
              </tbody>
            </table>
          </div>
        </div>
        {% elif selectedChr != -1 and plotScale =="physic" and (dataset.group.species == 'mouse' or dataset.group.species == 'rat') %}
        <div style="width: 100%;">
          <h2>Interval Analyst</h2>
          <div id="table_container">
          <table id="trait_table" class="table-hover table-striped cell-border dataTable" style="float: left; width:100%;">
            <thead>
              <tr>
                {% for header in gene_table_header %}
                <th>{{ header|safe }}</th>
                {% endfor %}
              </tr>
            </thead>
            <tbody>
              {% for row in gene_table_body %}
              <tr>
                {% for n in range(row|length) %}
                {% if n == 0 %}
                <td align="center" style="padding: 1px 0px 1px 0px;">{{ row[n]|safe }}</td>
                {% else %}
                <td>{{ row[n]|safe }}</td>
                {% endif %}
                {% endfor %}
              </tr>
              {% endfor %}
            </tbody>
          </table>
          </div>
        </div>
        {% endif %}
    </div>

    <!-- End of body -->

{% endblock %}

{% block js %}  

    <script type="text/javascript" src="{{ url_for('js', filename='d3js/d3.min.js') }}"></script>
    <script type="text/javascript" src="{{ url_for('js', filename='js_alt/md5.min.js') }}"></script>
    <script type="text/javascript" src="{{ url_for('js', filename='js_alt/underscore.min.js') }}"></script>
    <script type="text/javascript" src="{{ url_for('js', filename='underscore-string/underscore.string.min.js') }}"></script>
    <script type="text/javascript" src="{{ url_for('js', filename='d3-tip/d3-tip.js') }}"></script>
    <script type="text/javascript" src="/static/new/js_external/plotly-latest.min.js"></script>
    {% if manhattan_plot == True and selectedChr == -1 %}
    <script type="text/javascript" src="/static/new/js_external/jscolor.js"></script>
    {% endif %}

    <script language="javascript" type="text/javascript" src="{{ url_for('js', filename='DataTables/js/jquery.dataTables.min.js') }}"></script>
    <script language="javascript" type="text/javascript" src="https://cdn.datatables.net/buttons/1.0.0/js/dataTables.buttons.min.js"></script>
    <script language="javascript" type="text/javascript" src="{{ url_for('js', filename='DataTablesExtensions/plugins/sorting/scientific.js') }}"></script>
    <script language="javascript" type="text/javascript" src="{{ url_for('js', filename='DataTablesExtensions/plugins/sorting/natural.js') }}"></script>
    <script language="javascript" type="text/javascript" src="/static/packages/purescript_genome_browser/js/purescript-genetics-browser.js"></script>

    <script>
        js_data = {{ js_data | safe }}
    </script>

    <script language="javascript" type="text/javascript" src="/static/new/javascript/search_results.js"></script>
    {% if mapping_method == "gemma" or mapping_method == "reaper" %}
    <script language="javascript" type="text/javascript" src="/static/new/javascript/init_genome_browser.js"></script>
    {% endif %}

    <script type="text/javascript" charset="utf-8">
        $(document).ready( function () {
            console.time("Creating table");
            {% if selectedChr == -1 %}
            $('#trait_table').DataTable( {
                "columns": [
                    { "type": "natural", "width": "5%" },
                    { "type": "natural", "width": "8%" },
                    { "type": "natural", "width": "20%" },
                    { "type": "natural" },
                    { "type": "natural" }{% if 'additive' in qtlresults[0] %},
                    { "type": "natural" }{% endif %}{% if 'dominance' in qtlresults[0] and dataset.group.genetic_type != "riset" %},
                    { "type": "natural" }{% endif %}
                ],
                "columnDefs": [ {
                    "targets": 0,
                    "orderable": false
                } ],
                "order": [[1, "asc" ]],
                "sDom": "RZtir",
                "iDisplayLength": -1,
                "autoWidth": false,
                "deferRender": true,
                "bSortClasses": false,
                "scrollCollapse": false,
                "paging": false
            } );
            {% elif selectedChr != -1 and plotScale =="physic" and (dataset.group.species == 'mouse' or dataset.group.species == 'rat') %}
            $('#trait_table').dataTable( {
                "createdRow": function ( row, data, index ) {
                  $('td', row).eq(1).attr("align", "right");
                  $('td', row).eq(3).attr("align", "right");
                  $('td', row).eq(4).attr("align", "right");
                  $('td', row).eq(5).attr("align", "right");
                  $('td', row).eq(6).attr("align", "right");
                  $('td', row).eq(7).attr("align", "right");
                  $('td', row).eq(8).attr("align", "center");
                  $('td', row).eq(9).attr("align", "right");
                },
                "columns": [
                    { "orderDataType": "dom-checkbox" },
                    { "type": "natural"},
                    { "type": "natural" , "orderDataType": "dom-inner-text" },
                    { "type": "natural" , "orderDataType": "dom-inner-text" },
                    { "type": "natural" , "orderDataType": "dom-inner-text" },
                    { "type": "natural" },
                    { "type": "natural" },
                    { "type": "natural-minus-na" },
                    { "type": "natural-minus-na" },
                    { "type": "natural-minus-na" , "orderDataType": "dom-inner-text" },
                    { "type": "natural" }
                ],
                "columnDefs": [ {
                    "targets": 0,
                    "sortable": false
                }],
                "order": [[3, "asc" ]],
                "sDom": "RZtir",
                "iDisplayLength": -1,
                "autoWidth": false,
                "deferRender": true,
                "bSortClasses": false,
                "scrollCollapse": false,
                "paging": false
            } );
            {% endif %}

            $('#vector_map_tab').click(function(){
                $('div#gn1_map_options').hide();
            });

            $('#gn1_map_tab').click(function(){
                $('div#gn1_map_options').show();
            });

        });

        var mapping_input_list = ['temp_uuid', 'trait_id', 'dataset', 'tool_used', 'form_url', 'method', 'transform', 'trimmed_markers', 'selected_chr', 'chromosomes', 'mapping_scale',
                            'score_type', 'suggestive', 'significant', 'num_perm', 'permCheck', 'perm_output', 'perm_strata', 'categorical_vars', 'num_bootstrap', 'bootCheck', 'bootstrap_results',
                            'LRSCheck', 'covariates', 'maf', 'use_loco', 'manhattan_plot', 'color_scheme', 'manhattan_single_color', 'control_marker', 'control_marker_db', 'do_control', 'genofile',
                            'pair_scan', 'startMb', 'endMb', 'graphWidth', 'lrsMax', 'additiveCheck', 'showSNP', 'showGenes', 'viewLegend', 'haplotypeAnalystCheck', 
                            'mapmethod_rqtl_geno', 'mapmodel_rqtl_geno', 'temp_trait', 'group', 'species', 'reaper_version', 'primary_samples', 'n_samples']

        $('input[name=wanted_inputs]').val(mapping_input_list.join(","));

        chrView = function(this_chr, chr_mb_list) {
            $('input[name=selected_chr]').val(this_chr)
            $('input[name=chr_mb_list]').val(chr_mb_list)

            $('#marker_regression_form').attr('action', '/loading');
            $('#marker_regression_form').submit();
        };

        rangeView = function(this_chr, start_mb, end_mb) {
            $('input[name=selected_chr]').val(this_chr)
            $('input[name=startMb]').val(start_mb)
            $('input[name=endMb]').val(end_mb)
            //$('input[name=mb_range]').val(start_mb + "," + end_mb)

            $('#marker_regression_form').attr('action', '/loading');
            $('#marker_regression_form').submit();
        };

        remap = function() { 
            $('input[name=selected_chr]').val($('select[name=chromosomes]').val());
            $('input[name=color_scheme]').val($('select#color_scheme').val());
            $('#marker_regression_form').attr('action', '/loading');
            return $('#marker_regression_form').submit(); 
        };

        {% if manhattan_plot == True and selectedChr == -1 %}
        $('#color_scheme').change(function(){
          if ($(this).val() == "single"){
            $('#point_color_picker').show();
          } else {
            $('#point_color_picker').hide();
          }
        });
        {% endif %}

        {% if mapping_method != "gemma" and mapping_method != "plink" %}
        $('#download_perm').click(function(){
          perm_info_dict = {
            perm_data: js_data.perm_results,
            num_perm: "{{ nperm }}",
            trait_name: "{{ this_trait.display_name }}",
            trait_description: "{{ this_trait.description_display }}",
            cofactors: "{{ covariates }}",
            n_samples: {{ n_samples }},
            n_genotypes: {{ qtl_results|length }},
            {% if genofile_string is defined %}
            genofile: "{{ genofile_string }}",
            {% else %}
            genofile: "",
            {% endif %}
            units_linkage: "{{ LRS_LOD }}",
            strat_cofactors: js_data.categorical_vars
          }
          json_perm_data = JSON.stringify(perm_info_dict);

          $('input[name=perm_info]').val(json_perm_data);
          $('#marker_regression_form').attr('action', '/export_perm_data');
          return $('#marker_regression_form').submit();
        });

        modebar_options = {
          modeBarButtonsToAdd:[{
            name: 'Export as SVG',
            icon: Plotly.Icons.disk,
            click: function(gd) {
              Plotly.downloadImage(gd, {format: 'svg'})
            }
          }],
          modeBarButtonsToRemove:['toImage', 'sendDataToCloud', 'hoverClosest', 'hoverCompare', 'hoverClosestCartesian', 'hoverCompareCartesian', 'lasso2d', 'toggleSpikelines'],
          displaylogo: false
          //modeBarButtons:['toImage2', 'zoom2d', 'pan2d', 'select2d', 'zoomIn2d', 'zoomOut2d', 'autoScale2d', 'resetScale2d'],
        }

        perm_data = js_data.perm_results
        var hist_trace = {
            x: perm_data,
            type: 'histogram'
        };
        histogram_data = [hist_trace];
        histogram_layout = {
          bargap: 0.05,
          title: "<b>Histogram of Permutation Test</b>",
          xaxis: {
                  autorange: true,
                  title: "<b>{{ LRS_LOD }}</b>",
                  titlefont: {
                    family: "arial",
                    size: 20
                  },
                  ticklen: 4,
                  tickfont: {
                    size: 16
                  }
                },
          yaxis: {
                  autorange: true,
                  title: "<b>Count</b>",
                  titlefont: {
                    family: "arial",
                    size: 20
                  },
                  showline: true,
                  ticklen: 4,
                  tickfont: {
                    size: 16
                  },
                  automargin: true
                },
          width: 500,
          height: 300,
          margin: {
              l: 70,
              r: 30,
              t: 100,
              b: 50
          }
        };

        Plotly.newPlot('perm_histogram', histogram_data, histogram_layout, modebar_options);
        {% endif %}

        export_mapping_results = function() {
            $('#marker_regression_form').attr('action', '/export_mapping_results');
            return $('#marker_regression_form').submit();
        }

        $('#export_mapping_results').click(export_mapping_results);

        $('#browser_tab').click(function() {
          $('#gn1_map_options').css("display", "none")
        })
        $('#gn1_map_tab').click(function() {
          $('#gn1_map_options').css("display", "block")
        })
    </script>

{% endblock %}

