{% extends "base.html" %}
{% block title %}Mapping Results{% endblock %}
{% block css %}
    <link rel="stylesheet" type="text/css" href="/static/new/packages/DataTables/css/jquery.dataTables.css" />
    <link rel="stylesheet" type="text/css" href="/static/packages/DT_bootstrap/DT_bootstrap.css" />

    <link rel="stylesheet" type="text/css" href="/static/new/css/marker_regression.css" />
    <link rel="stylesheet" type="text/css" href="/static/new/css/panelutil.css" />
{% endblock %}
{% from "base_macro.html" import header %}
{% block content %}
    <div class="container">
        <form method="post" target="_blank" action="/run_mapping" name="marker_regression" id="marker_regression_form">
        <input type="hidden" name="temp_uuid" value="{{ temp_uuid }}">
        <input type="hidden" name="trait_id" value="{{ this_trait.name }}">
        <input type="hidden" name="dataset" value="{{ dataset.name }}">
        <input type="hidden" name="genofile" value="{{ genofile_string }}">
        <input type="hidden" name="geno_db_exists" value="{{ geno_db_exists }}">
        <input type="hidden" name="results_path" value="{{ mapping_results_path }}">
        <input type="hidden" name="method" value="{{ mapping_method }}">
        {% for sample in samples %}
        <input type="hidden" name="value:{{ sample }}" value="{{ vals[loop.index - 1] }}">
        {% endfor %}
        <input type="hidden" name="maf">
        <input type="hidden" name="use_loco" value="{{ use_loco }}">
        <input type="hidden" name="selected_chr" value="{{ selectedChr }}">
        <input type="hidden" name="manhattan_plot" value="{{ manhattan_plot }}">
        <input type="hidden" name="num_perm" value="{{ nperm }}">
        <input type="hidden" name="perm_results" value="">
        <input type="hidden" name="num_bootstrap" value="{{ nboot }}">
        <input type="hidden" name="do_control" value="{{ doControl }}">
        <input type="hidden" name="control_marker" value="{{ controlLocus }}">
        <input type="hidden" name="covariates" values="{{ covariates }}">
        <input type="hidden" name="mapmethod_rqtl_geno" value="{{ mapmethod_rqtl_geno }}">
        <input type="hidden" name="mapmodel_rqtl_geno" value="{{ mapmodel_rqtl_geno }}">
        <input type="hidden" name="pair_scan" value="{{ pair_scan }}">
        
        <div class="container">
          <div class="col-xs-5">
              <h2>Map Viewer: Whole Genome</h2><br>
              <b>Population:</b> {{ dataset.group.species|capitalize }} {{ dataset.group.name }}<br>
              <b>Database:</b> {{ dataset.fullname }}<br>
              {% if dataset.type == "ProbeSet" %}<b>Trait ID:</b>{% else %}<b>Record ID:</b>{% endif %} <a href="/show_trait?trait_id={{ this_trait.name }}&dataset={{ dataset.name }}">{{ this_trait.name }}</a><br>
              {% if dataset.type == "ProbeSet" %}
              <b>Gene Symbol:</b> <i>{{ this_trait.symbol }}</i><br>
              <b>Location:</b> Chr {{ this_trait.chr }} @ {{ this_trait.mb }} Mb<br>
              {% endif %}
              {% if genofile_string is defined %}
              <b>Genotypes:</b> {{ genofile_string.split(":")[1] }}
              {% endif %}
              <br>
              <br>
              <a href="javascript:export_mapping_results();" target="_blank" >Download Full Results</a>
          </div>
          <div id="gn1_map_options" class="col-xs-5" style="outline: 3px double #AAAAAA; padding: 10px; margin: 10px;">
            <div class="col-xs-8" style="padding: 0px;">
              <table>
                <tr>
                  <td><b>Chr:&nbsp;</b></td>
                  <td>
                    <select name="chromosomes" size="1">
                        {% for chr in ChrList %}
                        <option value="{{ chr[1] }}" {% if (chr[1] + 1) == selectedChr %}selected{% endif %}>{{ chr[0] }}</option>
                        {% endfor %}
                    </select>&nbsp;
                    <button type="button" class="btn btn-primary" style="padding-bottom: 2px; padding-top: 2px;" onclick="javascript:remap();">Remap</button>
                  </td>
                </tr>
                <tr>
                  <td><b>View:&nbsp;</b></td>
                  <td style="padding: 5px;">
                    <input type="text" name="startMb" size="10" value="{% if startMb != -1 %}{{ startMb }}{% endif %}"> to <input type="text" name="endMb" size="10" value="{% if endMb != -1 %}{{ endMb }}{% endif %}">
                  </td>
                </tr>
                <tr>
                  <td><b>Units:&nbsp;</b></td>
                  <td style="padding: 5px;">
                    <label class="radio-inline">
                      <input type="radio" name="LRSCheck" value="LRS" {% if LRS_LOD == "LRS" %}checked{% endif %}>LRS
                    </label>
                    <label class="radio-inline">
                      <input type="radio" name="LRSCheck" value="{% if LRS_LOD == "-log(p)" %}-log(p){% else %}LOD{% endif %}" {% if LRS_LOD == "LOD" or LRS_LOD == "-log(p)" %}checked{% endif %}>{% if LRS_LOD == "-log(p)" %}-log(p){% else %}LOD{% endif %}
                    </label>
                    <a href="http://genenetwork.org/glossary.html#LOD" target="_blank">
                      <sup style="color:#f00"> ?</sup>
                    </a>
                  </td>
                </tr>
                <tr>
                  <td></td>
                  <td style="padding: 5px;">
                    <input type="text" name="lrsMax" value="{{ '%0.1f' | format(lrsMax|float) }}" size="3"> <span style="font-size: 12px;">units on the y-axis (0 for default)</span>   
                  </td>
                </tr>
                <tr>
                  <td><b>Width:&nbsp;</b></td>
                  <td>
                    <input type="text" name="graphWidth" value="{% if graphWidth is defined %}{{ graphWidth }}{% else %}1600{% endif %}" size="5"><span style="font-size: 12px;"> pixels (minimum=900)</span>
                  </td>
                </tr>
              </table>
            </div>
            <div class="col-xs-4" style="padding: 0px;">
                  {% if (mapping_method == "reaper" or mapping_method == "rqtl_geno") and nperm > 0 %}
                  <input type="checkbox" name="permCheck" class="checkbox" style="display: inline; margin-top: 0px;" {% if permChecked|upper == "ON" %}value="ON" checked{% endif %}> <span style="font-size: 12px;">Permutation Test 
                    <a href="http://genenetwork.org/glossary.html#Permutation" target="_blank">
                      <sup style="color:#f00"> ?</sup>
                    </a>
                  <br>
                  {% endif %}
                  {% if mapping_method == "reaper" and nboot > 0 %}
                  <input type="checkbox" name="bootCheck" class="checkbox" style="display: inline; margin-top: 0px;" {% if bootChecked|upper == "ON" %}value="ON" checked{% endif %}> <span style="font-size: 12px;">Bootstrap Test 
                    <a href="http://genenetwork.org/glossary.html#bootstrap" target="_blank">
                      <sup style="color:#f00"> ?</sup>
                    </a>
                  <br>
                  {% endif %}
                  {% if mapping_method == "reaper" %}
                  <input type="checkbox" name="additiveCheck" class="checkbox" style="display: inline; margin-top: 0px;" {% if additiveChecked|upper == "ON" %}value="ON" checked{% endif %}> <span style="font-size: 12px;">Allele Effects
                    <a href="http://genenetwork.org/glossary.html#additive" target="_blank">
                      <sup style="color:#f00"> ?</sup>
                    </a>
                  <br>
                  {% endif %}
                  <input type="checkbox" name="showSNP" class="checkbox" style="display: inline; margin-top: 0px;" {% if SNPChecked|upper == "ON" %}value="ON" checked{% endif %}> <span style="font-size: 12px;">SNP Track </span>
                    <a href="http://genenetwork.org/glossary.html#snpSeismograph" target="_blank">
                      <sup style="color:#f00"> ?</sup>
                    </a>
                    <span style="color:red;">*</span>
                  <br>
                  <input type="checkbox" name="showGenes" class="checkbox" style="display: inline; margin-top: 0px;" {% if geneChecked|upper == "ON" %}value="ON" checked{% endif %}> <span style="font-size: 12px;">Gene Track </span> <span style="color:red;">*</span><br>
                  {% if plotScale != "morgan" and mapping_method != "gemma" and mapping_method != "plink" %}
                  <input type="checkbox" name="haplotypeAnalystCheck" class="checkbox" style="display: inline; margin-top: 0px;" {% if haplotypeAnalystChecked|upper == "ON" %}value="ON" checked{% endif %}> <span style="font-size: 12px;">Haplotype Analyst </span> <span style="color:red;">*</span><br>
                  {% endif %}
                  <input type="checkbox" name="viewLegend" class="checkbox" style="display: inline; margin-top: 0px;" {% if legendChecked|upper == "ON" %}value="ON" checked{% endif %}> <span style="font-size: 12px;">Legend </span>
            </div>
            <div class="col-xs-12" align="center" style="padding: 5px;">
              <span style="color:red;">*</span> <span style="font-size: 12px;">only apply to single chromosome physical mapping</span>
            </div>
          </div>
        </div>
 
        <div class="tabbable" style="margin: 10px;">
          <ul class="nav nav-tabs">
            <li id="gn1_map_tab">
              <a href="#gn1_map" data-toggle="tab" aria-expanded="true">GN1 Map</a>
            </li>
            {% if mapping_method != "gemma" and mapping_method != "plink" %}
            <li id="vector_map_tab">
              <a href="#vector_map" data-toggle="tab" {% if mapping_method != "gemma" and mapping_method != "plink" %}aria-expanded="false"{% else %}aria-expanded="true"{% endif %}>Vector Map</a>
            </li>
            {% endif %}
          </ul>
          <div class="tab-content">
            <div class="tab-pane active" id="gn1_map">
              <div class="qtlcharts">
                  {{ gifmap|safe }}
                  <img src="/generated/{{ filename }}.png" usemap="#WebQTLImageMap">
                  {% if additiveChecked|upper == "ON" %}
                  <br>
                  <span style="white-space: nowrap;">A positive additive coefficient (green line) indicates that {{ dataset.group.parlist[1] }} alleles increase trait values. In contrast, a negative additive coefficient (orange line) indicates that {{ dataset.group.parlist[0] }} alleles increase trait values.</span>
                  {% endif %}
                  {% if nperm > 0 and permChecked == "ON" %}
                  <br><br>
                  <img src="/generated/{{ perm_filename }}.gif">
                  <br><br>
                  Total of {{ nperm }} permutations&nbsp;&nbsp;<a href="javascript:export_perm_data();" target="_blank" >Download Permutation Results</a>
                  <br>
                  {% endif %}
              </div>
            </div>
            {% if mapping_method != "gemma" and mapping_method != "plink" %}
            <div class="tab-pane {% if mapping_method == "gemma" %}active{% endif %}" id="vector_map">
              <div id="chart_container">
                <div class="qtlcharts" id="topchart"></div>
              </div>
            </div>
            {% endif %}
          </div>
        </div>

        </form>
        {% if selectedChr == -1 %}
        <div class="container" style="padding-left: 30px; width:{% if 'additive' in trimmed_markers[0] %}45%{% else %}35%{% endif %};">
          <h2>Mapping Statistics</h2>
        <br />
        <button class="btn btn-default" id="select_all"><span class="glyphicon glyphicon-ok"></span> Select All</button>
        <button class="btn btn-default" id="deselect_all"><span class="glyphicon glyphicon-remove"></span> Deselect All</button>
        <button class="btn btn-default" id="invert"><span class="glyphicon glyphicon-resize-vertical"></span> Invert</button>
        <button class="btn btn-default" id="add" disabled><span class="glyphicon glyphicon-plus-sign"></span> Add</button>
        <br />
        <br />
          <div id="table_container" style="border-style: solid; border-width: 1px; border-color: black;">
          <table id="trait_table" class="table table-hover table-striped nowrap">
            <thead>
              <tr>
                <th></th>
                <th>Row</th>
                <th>Locus</th>
                <th>{{ LRS_LOD }}</th>
                <th>Chr</th>
                {% if plotScale != "physic" %}
                <th>cM</th>
                {% else %}
                <th align="right">Mb</th>
                {% endif %}
                {% if 'additive' in trimmed_markers[0] %}
                <th>Add Eff</th>
                {% endif %}
                {% if 'dominance' in trimmed_markers[0] and dataset.group.genetic_type != "riset" %}
                <th>Dom Eff</th>
                {% endif %}
              </tr>
            </thead>
            <tbody>
              {% for marker in trimmed_markers %}
              <tr>
                <td align="center" style="padding-right: 0px;">
                  <input type="checkbox" name="selectCheck"
                         class="checkbox trait_checkbox"
                         value="{{ data_hmac('{}:{}Geno'.format(marker.name, dataset.group.name)) }}">
                </td>
                <td align="right">{{ loop.index }}</td>
                <td>{% if geno_db_exists == "True" %}<a href="/show_trait?trait_id={{ marker.name }}&dataset={{ dataset.group.name }}Geno">{{ marker.name }}</a>{% else %}{{ marker.name }}{% endif %}</td>
                {% if LRS_LOD == "LOD" or LRS_LOD == "-log(p)" %}
                {% if 'lod_score' in marker %}
                <td align="right">{{ '%0.2f' | format(marker.lod_score|float) }}</td>
                {% else %}
                <td align="right">{{ '%0.2f' | format(marker.lrs_value|float / 4.16) }}</td>
                {% endif %}
                {% else %}
                {% if 'lod_score' in marker %}
                <td align="right">{{ '%0.2f' | format(marker.lod_score|float * 4.16) }}</td>
                {% else %}
                <td align="right">{{ '%0.2f' | format(marker.lrs_value|float) }}</td>
                {% endif %}
                {% endif %}
                <td align="right">{{marker.chr}}</td>
                {% if plotScale != "physic" %}
                <td align="right">{{ '%0.3f' | format(marker.Mb|float) }}</td>
                {% else %}
                <td align="right">{{ '%0.6f' | format(marker.Mb|float) }}</td>
                {% endif %}
                {% if 'additive' in marker %}
                <td align="right">{{ '%0.3f' | format(marker.additive|float) }}</td>
                {% endif %}
                {% if 'dominance' in marker and dataset.group.genetic_type != "riset" %}
                <td align="right">{{ '%0.2f' | format(marker.dominance|float) }}</td>
                {% endif %}
              </tr>
              {% endfor %}
            </tbody>
          </table>
          </div>
        </div>
        {% elif selectedChr != -1 and plotScale =="physic" and (dataset.group.species == 'mouse' or dataset.group.species == 'rat') %}
        <div style="width: 100%;">
          <h2>Interval Analyst</h2>
          <div id="table_container">
          <table id="interval_analyst" class="table-hover table-striped cell-border" style="float: left; width:100%;">
            <thead>
              <tr>
                {% for header in gene_table_header %}
                <th>{{ header|safe }}</th>
                {% endfor %}
              </tr>
            </thead>
            <tbody>
              {% for row in gene_table_body %}
              <tr>
                {% for n in range(row|length) %}
                <td>{{ row[n]|safe }}</td>
                {% endfor %}
              </tr>
              {% endfor %}
            </tbody>
          </table>
          </div>
        </div>
        {% endif %}
    </div>

    <!-- End of body -->

{% endblock %}

{% block js %}  

    <script type="text/javascript" src="http://d3js.org/d3.v3.min.js"></script>
    <script type="text/javascript" src="/static/new/js_external/underscore-min.js"></script>
    <script type="text/javascript" src="/static/new/js_external/underscore.string.min.js"></script>
    <script type="text/javascript" src="/static/new/js_external/d3-tip.min.js"></script>

    <script language="javascript" type="text/javascript" src="https://cdn.datatables.net/1.10.12/js/jquery.dataTables.min.js"></script>
    <script language="javascript" type="text/javascript" src="https://cdn.datatables.net/buttons/1.0.0/js/dataTables.buttons.min.js"></script>
    <script language="javascript" type="text/javascript" src="/static/new/packages/DataTables/js/dataTables.scientific.js"></script>
    <script language="javascript" type="text/javascript" src="/static/new/packages/DataTables/js/dataTables.naturalSort.js"></script>

    {% if mapping_method != "gemma" and mapping_method != "plink" %}
    <script>
        js_data = {{ js_data | safe }}
    </script>
    {% endif %}

    <script language="javascript" type="text/javascript" src="/static/new/javascript/search_results.js"></script>
    {% if mapping_method != "gemma" and mapping_method != "plink" %}
    <script language="javascript" type="text/javascript" src="/static/new/javascript/panelutil.js"></script>
    <script language="javascript" type="text/javascript" src="/static/new/javascript/chr_lod_chart.js"></script> 
    <script language="javascript" type="text/javascript" src="/static/new/javascript/lod_chart.js"></script>
    <script language="javascript" type="text/javascript" src="/static/new/javascript/create_lodchart.js"></script>
    {% endif %}

    <script type="text/javascript" charset="utf-8">
        $(document).ready( function () {
            console.time("Creating table");
            $('#trait_table').DataTable( {
                {% if mapping_method != "reaper" %}
                "columns": [
                    { "type": "natural", "width": "5%" },
                    { "type": "natural", "width": "8%" },
                    { "type": "natural", "width": "20%" },
                    { "type": "natural", "width": "8%" },
                    { "type": "natural", "width": "8%" },
                    { "type": "natural", "width": "15%" }
                ],
                {% elif dataset.group.genetic_type == "riset" %}
                "columns": [
                    { "type": "natural", "width": "5%" },
                    { "type": "natural", "width": "8%" },
                    { "type": "natural", "width": "20%" },
                    { "type": "natural", "width": "8%" },
                    { "type": "natural", "width": "8%" },
                    { "type": "natural", "width": "15%" },
                    { "type": "natural", "width": "8%" }
                ],
                {% else %}
                "columns": [
                    { "type": "natural", "width": "5%" },
                    { "type": "natural", "width": "8%" },
                    { "type": "natural", "width": "20%" },
                    { "type": "natural", "width": "8%" },
                    { "type": "natural", "width": "8%" },
                    { "type": "natural", "width": "15%" },
                    { "type": "natural", "width": "8%" },
                    { "type": "natural", "width": "8%" }
                ],
                {% endif %}
                "columnDefs": [ {
                    "targets": 0,
                    "orderable": false
                } ],
                "order": [[1, "asc" ]],
                "sDom": "RZtir",
                "iDisplayLength": -1,
                "autoWidth": false,
                "deferRender": true,
                "bSortClasses": false,
                "scrollCollapse": false,
                "paging": false
            } );

            $('#interval_analyst').dataTable( {
                "columnDefs": [ {
                    "targets": 0,
                    "sortable": false
                }],
                "order": [[3, "asc" ]],
                "sDom": "RZtir",
                "iDisplayLength": -1,
                "autoWidth": false,
                "deferRender": true,
                "bSortClasses": false,
                "scrollCollapse": false,
                "paging": false
            } );

            $('#vector_map_tab').click(function(){
                $('div#gn1_map_options').hide();
            });

            $('#gn1_map_tab').click(function(){
                $('div#gn1_map_options').show();
            });

        });

        chrView = function(this_chr, chr_mb_list) {
            $('input[name=selected_chr]').val(this_chr)
            $('input[name=chr_mb_list]').val(chr_mb_list)

            return $('#marker_regression_form').submit();
        };

        rangeView = function(this_chr, start_mb, end_mb) {
            $('input[name=selected_chr]').val(this_chr)
            $('input[name=startMb]').val(start_mb)
            $('input[name=endMb]').val(end_mb)
            //$('input[name=mb_range]').val(start_mb + "," + end_mb)

            return $('#marker_regression_form').submit();
        };

        remap = function() { 
            $('input[name=selected_chr]').val($('select[name=chromosomes]').val());
            return $('#marker_regression_form').submit(); 
        };

        {% if mapping_method != "gemma" and mapping_method != "plink" %}
        export_perm_data = function() {
            var num_perm, perm_data;
            num_perm = js_data.num_perm
            perm_data = js_data.perm_results
            json_perm_data = JSON.stringify(perm_data);
            $('input[name=perm_results]').val(json_perm_data);
            $('#marker_regression_form').attr('action', '/export_perm_data');
            return $('#marker_regression_form').submit();
        }
        {% endif %}

        export_mapping_results = function() {
            $('#marker_regression_form').attr('action', '/export_mapping_results');
            return $('#marker_regression_form').submit();
        }

    </script>

{% endblock %}

