<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.0 Transitional//EN">
<HTML><HEAD><TITLE>CXB Genotype / WebQTL</TITLE>
<META http-equiv=Content-Type content="text/html; charset=iso-8859-1">
<LINK REL="stylesheet" TYPE="text/css" HREF='/css/general.css'>
<LINK REL="stylesheet" TYPE="text/css" HREF='/css/menu.css'>

</HEAD>
<BODY  bottommargin="2" leftmargin="2" rightmargin="2" topmargin="2" text=#000000 bgColor=#ffffff>
<TABLE cellSpacing=5 cellPadding=4 width="100%" border=0>
<TBODY>
<TR>
<script language="JavaScript" src="/javascript/header.js"></script>
</TR>
<TR>
<TD bgColor=#eeeeee class="solidBorder">
<Table width= "100%" cellSpacing=0 cellPadding=5><TR>
<!-- Body Start from Here -->
<TD valign="top" height="200" width="100%" bgcolor="#eeeeee">

<P class="title">CXB Genotypes Database (July 2005)

 <A HREF="/webqtl/main.py?FormID=editHtml"><img src="/images/modify.gif" alt="modify this page" border= 0 valign="middle"></A></P>


<P class="subtitle">&nbsp;&nbsp;&nbsp;&nbsp;Summary:</P>

<Blockquote><P>



This CXB genotype data set still consists of 1384 SNP and microsatellite markers with unique strain distribution patterns. This file is used to map all CXB phenotype data sets including approximately 500 phenotypes in the CXB Phenotypes database and 45,000 Hippocampal mRNA expression phenotypes. The present genotype file supercedes an older microsatellite file (405 markers). </Blockquote>



<P class="subtitle">&nbsp;&nbsp;&nbsp;&nbsp;About the genotypes used in these studies:</P>

<Blockquote> WebQTL mapping algorithms rely on genotypes for the CXB strains that include both microsatellite markers (labeled <I>Mit</I> and <I>Msw</I>) and single nucleotide polymorphisms (labeled <I>Gnf</I>). The current set of markers (n = 1384) have been carefully error-checked. Closely linked genetic markers often have the same strain distribution pattern (SDP) across the CXB strains. For computational efficiency, we only use a single marker associated with each SDP. The CXB set is so small that markers on different chromosomes occasionally have almost precisely the same SDP. This produces high non-syntenic association and false linkage between variance in phenotypes and genotypes. Please examine the correlation coeffients of markers close to interest loci with ALL other markers to evaluate the risk of non-syntenic association.
</Blockquote>

<Blockquote>We have genotyped all available CXB strains from The Jackson Laboratory. The entire CXB genotypes data may be <A HREF="http://www.genenetwork.org/genotypes/CXB.geno" class="fs14"> downloaded</A>.
</Blockquote>

<Blockquote>Marker-strain pairs for which we were missing genotypes were often inferred from flanking markers. In marker sets lacking genotypes for a particular strain, a note is included to that effect in the marker set description below.
</Blockquote>


<P class="subtitle">&nbsp;&nbsp;&nbsp;&nbsp;About the marker sets:</P>

<Blockquote> <U><B>Mit</B></U><br>

<I>Mit</I> markers, described by William Dietrich and colleagues (<a href="http://www.broad.mit.edu/cgi-bin/mouse/sts_info?database=mouserelease" target="_blank" class="fs14">1992</a>), are the most widely used of the three marker sets. These markers typically consist of regions of repeated dinucleotides (so-called CA repeat microsatellites) that vary in length among strains. The CA repeat polymorphisms are flanked by unique sequence that can be used to design polymerase chain reaction (PCR) primers that will selectively amplify the intervening variable region. While many of the <i>Mit</i> markers have been typed in the BXD strain set by a number of investigators, the genotypes used here are those reported in the consensus map created by Williams and colleagues (<a href="http://www.genomebiology.com/2001/2/11/research/0046" target="_blank" class="fs14">2001</a>).

<br>
<br><i>Mit</i> marker names: D + (Chr of Marker) + Mit + (Order Found) <UL>
<LI>D indicates that the marker is a DNA segment.
<LI><i>Mit</i>  indicates that the marker was identified at the Massachusetts Institute of Technology.
<LI>Order Found indicates the order in which the markers were identified. </UL>
</Blockquote>

<Blockquote><B><U>Gnf</U></B>
<br><i>Gnf</i> markers are single nucleotide polymorphisms (SNPs) identified between B6 and D2 by genomic sequence sampling. Polymorphisms were typed by Mathew Pletcher and Tim Wiltshire using the Sequenom MassEXTEND system (Wiltshire et al., <A HREF="http://www.ncbi.nlm.nih.gov/entrez/query.fcgi?cmd=Retrieve&db=PubMed&list_uids=12612341&dopt=Abstract" target="_blank" class="fs14">2003</A>). Each of the genotyping reactions was set up in duplicate. Physical positions were determined for each marker and integrated with previous BXD RI mapping data based on a combination of physical and genetic positions. Unsupported double crossovers were verified by manual inspection to ensure accuracy of calls. A full list of SNPs identified in the sequence sampling can be found at <a href="http://www.gnf.org/SNP/" class = "normalsize">http://www.gnf.org/SNP</a>.
<br>
<br><i>Gnf</i> marker names: S + (Chr of Marker) + Gnf + (Mb position) <UL>
<LI>S indicates the marker is a SNP
<LI><i>Gnf</i> indicates that the marker originated at the Genomics Institute of the Novartis Research Foundation.
<LI>Mb position may include decimal values. </UL>
</Blockquote>


<Blockquote><P>

Notes on Nomenclature: The CXB set is the first and oldest group of RI strains of any species. The materal strain is BALB/cBy and the paternal strain is C57BL/6By. Eleven CXB strains were produced at the National Institutes of Health by Donald Bailey (By) starting in 1959, and eight are still extant. After moving to The Jackson Laboratory in 1967, an additional set of five strains were created with the help of Jo Hilgers (Hi). The strains are now labeled numerically. The following are the old strain symbols for CXB1 through CXB7:
<UL>
<LI>CXB1 = CXBD
<LI>CXB2 = CXBE
<LI>CXB3 = CXBG
<LI>CXB4 = CXBH
<LI>CXB5 = CXBI
<LI>CXB6 = CXBJ
<LI>CXB7 = CXBK  (has a 3' UTR polymorphism in mu opioid receptor; PMID: 16708053)
</UL>
</Blockquote>



<P class="subtitle">&nbsp;&nbsp;&nbsp;&nbsp;Acknowledgments:</P>
<Blockquote>
Genotypes for the Mit and Msw marker sets were determined by Jing Gu and

Lu Lu.  <i>Gnf</i> SNP genotypes were generated by Tim

Wiltshire and Mathew Pletcher. The selection of markers to included in the final file was carried out

by Jing Gu.

This text file was originally written by Jeremy Peirce (August 21,

2003). Updated August 22, 2003 by RW/JP/LL. Updated July 31, 2005 by RW.


</Blockquote>

<P class="subtitle">&nbsp;&nbsp;&nbsp;&nbsp;Reference:</P>
<Blockquote><P>Dietrich WF, Katz H, Lincoln SE (<a href="http://www.ncbi.nlm.nih.gov/entrez/query.fcgi?cmd=Retrieve&db=PubMed&list_uids=1353738&dopt=Abstract" class="fs14">1992</a>) A genetic map of the mouse suitable for typing in intraspecific crosses. Genetics 131:423-447.
</P></Blockquote>

<Blockquote><P>
Williams RW, Gu J, Qi S, Lu L (<a href="http://www.genomebiology.com/2001/2/11/research/0046" class=normal>2001</a>) The genetic structure of recombinant inbred mice: High-resolution consensus maps for complex trait analysis. Genome Biology 2:RESEARCH0046
</P></Blockquote>

<Blockquote><P>
Wiltshire T, Pletcher MT, Batalov S, Barnes SW, Tarantino LM, Cooke MP, Wu H, Smylie K, Santrosyan A, Copeland NG, Jenkins NA, Kalush F, Mural RJ, Glynne RJ, Kay SA, Adams MD, Fletcher CF (<A HREF="http://www.ncbi.nlm.nih.gov/entrez/query.fcgi?cmd=Retrieve&db=PubMed&list_uids=12612341&dopt=Abstract" class="fs14">2003</A>) Genome-wide single-nucleotide polymorphism analysis defines haplotype patterns in mouse. Proc Natl Acad Sci USA 100:3380-3385.


</P></Blockquote>
<Blockquote><P>
<P></P>

</TD>
</TR></TABLE>
</TD>
</TR>
<TR>
<TD align=center bgColor=#ddddff class="solidBorder">



<!--Start of footer-->
<TABLE width="90%">
<script language='JavaScript' src='/javascript/footer.js'></script>
</TABLE>
<!--End of footer-->



</TD>
</TR>
</TABLE>
<!-- /Footer -->
<!-- menu script itself. you should not modify this file -->
<script language="JavaScript" src="/javascript/menu_new.js"></script>
<!-- items structure. menu hierarchy and links are stored there -->
<script language="JavaScript" src="/javascript/menu_items.js"></script>
<!-- files with geometry and styles structures -->

<script language="JavaScript" src="/javascript/menu_tpl.js"></script>
<script language="JavaScript">
<!--//
// Note where menu initialization block is located in HTML document.
// Don't try to position menu locating menu initialization block in
// some table cell or other HTML element. Always put it before </body>
// each menu gets two parameters (see demo files)
// 1. items structure
// 2. geometry structure
new menu (MENU_ITEMS, MENU_POS);
// make sure files containing definitions for these variables are linked to the document
// if you got some javascript error like "MENU_POS is not defined", then you've made syntax
// error in menu_tpl.js file or that file isn't linked properly.

// also take a look at stylesheets loaded in header in order to set styles
//-->
</script>
</BODY>
</HTML>
