<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.01//EN" "http://www.w3.org/TR/html4/strict.dtd">
<!--<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.0 Transitional//EN">-->
<HTML><HEAD><TITLE>MDC/CAS/UCL Liver 230v2 (Dec08) RMA **</TITLE>
<META http-equiv=Content-Type content="text/html; charset=iso-8859-1">
<LINK REL="stylesheet" TYPE="text/css" HREF='/css/general.css'>
<LINK REL="stylesheet" TYPE="text/css" HREF='/css/menu.css'>
<LINK RELl="stylesheet" href="ajax/css/ajax-tooltip.css" media="screen" type="text/css">
<script type="text/javascript" src="ajax/js/ajax-dynamic-content.js"></script>
<script type="text/javascript" src="ajax/js/ajax.js"></script>
<script type="text/javascript" src="ajax/js/ajax-tooltip.js"></script>

</HEAD>
<BODY  bottommargin="2" leftmargin="2" rightmargin="2" topmargin="2" text=#000000 bgColor=#ffffff>
<TABLE cellSpacing=5 cellPadding=4 width="100%" border=0>
<TBODY>
<TR>
<script language="JavaScript" src="/javascript/header.js"></script>
</TR>
<TR>
<TD bgColor=#eeeeee class="solidBorder">
<Table width= "100%" cellSpacing=0 cellPadding=5><TR>
<!-- Body Start from Here -->
<TD valign="top" height="200" width="100%" bgcolor="#eeeeee">
<P class="title">MDC/CAS/UCL Liver 230v2 (Dec08) RMA **  <A HREF="/webqtl/main.py?FormID=editHtml"><img src="/images/modify.gif" alt="modify this page" border= 0 valign="middle"></A><BR><BR>Accession number: <A HREF="/webqtl/main.py?FormID=sharinginfo&GN_AccessionId=222">GN222</A></P>
<P><font color="#FF0000">Waiting for the data provider to submit their info file</font></P>

<P class="subtitle">Summary:</P>

<Blockquote>
<P>
<B>SUBTITLE</B>Expression data for normal liver of the HXB/BXH rat RI strains generated using the Affymetrix Rat Genome 230 2.0 array <A HREF="http://www.ncbi.nlm.nih.gov/projects/geo/query/acc.cgi?acc=GPL1355>GEO GPL1355</A></P>


<P>Grieve IC, Dickens NJ, Pravenec M, Kren V, Hubner N, Cook SA, Aitman TJ, Petretto E, Mangion J.
MRC Clinical Sciences Centre, Imperial College, Hammersmith Hospital, London, United Kingdom.

<P>http://www.ncbi.nlm.nih.gov/pubmed/19112506

<P>PLoS One. 2008;3(12):e4033. Epub 2008 Dec 29.

<P>Expression quantitative trait loci (eQTLs) represent genetic control points of gene expression, and can be categorized as cis- and trans-acting, reflecting local and distant regulation of gene expression respectively. Although there is evidence of co-regulation within clusters of trans-eQTLs, the extent of co-expression patterns and their relationship with the genotypes at eQTLs are not fully understood. We have mapped thousands of cis- and trans-eQTLs in four tissues (fat, kidney, adrenal and left ventricle) in a large panel of rat recombinant inbred (RI) strains. Here we investigate the genome-wide correlation structure in expression levels of eQTL transcripts and underlying genotypes to elucidate the nature of co-regulation within cis- and trans-eQTL datasets. Across the four tissues, we consistently found statistically significant correlations of cis-regulated gene expression to be rare (<0.9% of all pairs tested). Most (>80%) of the observed significant correlations of cis-regulated gene expression are explained by correlation of the underlying genotypes. In comparison, co-expression of trans-regulated gene expression is more common, with significant correlation ranging from 2.9%-14.9% of all pairs of trans-eQTL transcripts. We observed a total of 81 trans-eQTL clusters (hot-spots), defined as consisting of > or =10 eQTLs linked to a common region, with very high levels of correlation between trans-regulated transcripts (77.2-90.2%). Moreover, functional analysis of large trans-eQTL clusters (> or =30 eQTLs) revealed significant functional enrichment among genes comprising 80% of the large clusters. The results of this genome-wide co-expression study show the effects of the eQTL genotypes on the observed patterns of correlation, and suggest that functional relatedness between genes underlying trans-eQTLs is reflected in the degree of co-expression observed in trans-eQTL clusters. Our results demonstrate the power of an integrative, systematic approach to the analysis of a large gene expression dataset to uncover underlying structure, and inform future eQTL studies.
<!-- IF NEEDED
<P>Users of these this data set may also find the following complementary resources extremely useful:
<OL>
<LI> <A HREF="http://www.genenetwork.org" target="_blank" class="fs14">Link 1</A>.
<LI> <A HREF="http://www.genenetwork.org" target="_blank" class="fs14">Link 2</A>.
</OL>
-->
</Blockquote>


<P class="subtitle">About the cases used to generate this set of data:</P>

<Blockquote>

<P>Some text here</P>


</Blockquote>


<P class="subtitle">About the tissue used to generate this set of data:</P>

<Blockquote>
<P>Some text here</P>
<P>

<table border="0" cellpadding="0" cellspacing="0" bgcolor="#000000" width="600">
<tr><td>
<table width="600" border="0" cellpadding="5" cellspacing="1" align="left">
<tr bgcolor="royalblue">
<td><font color=#FFFFFF>Index</font></td><td><font color=#FFFFFF>TubeID</font></td><td><font color=#FFFFFF>Group</font></td><td><font color=#FFFFFF>Strain</font></td><td><font color=#FFFFFF>Age</font></td><td><font color=#FFFFFF>Sex</font></td><td><font color=#FFFFFF>Source</font></td></tr>
<tr bgcolor="#eeeeee"><td>1</td><td>R2595E.1</td><td>GDP</td><td>129S1/SvImJ</td><td>59</td><td>F</td><td>UTHSC RW</td></tr>
</table>

</td>
</tr>
</table>
</Blockquote>
</p>

<P class="subtitle">About downloading this data set:</P>
<Blockquote>
<P>Some text here</P>
</Blockquote>


<P class="subtitle">About the array platfrom:</P>
<Blockquote>
<P>Some text here</P>

</Blockquote>


<P class="subtitle">About data values and data processing:</P>

<Blockquote>
<P>Some text here</P>
<P>

<table border="0" cellpadding="0" cellspacing="0" bgcolor="#000000" width="600">
<tr><td>
<table width="600" border="0" cellpadding="5" cellspacing="1" align="left">
<tr bgcolor="royalblue">

<td><font color=#FFFFFF>Index</font></td><td><font color=#FFFFFF>TubeID</font></td><td><font color=#FFFFFF>Strain</font></td><td><font color=#FFFFFF>Original CEL</font></td><td><font color=#FFFFFF>Scale factor</font></td><td><font color=#FFFFFF>Background Average</font></td><td><font color=#FFFFFF>Present</font></td><td><font color=#FFFFFF>Absent</font></td><td><font color=#FFFFFF>Marginal</font></td><td><font color=#FFFFFF>AFFX-b-ActinMur (3'/5')</font></td><td><font color=#FFFFFF>AFFX-GapdhMur (3'/5')</font></td><td><font color=#FFFFFF>Batch Id</font></td><td><font color=#FFFFFF>Used for batch control</font></td></tr>
<tr bgcolor="#eeeeee"><td>1</td><td>R2595E.1</td><td>129S1/SvImJ</td><td>R2595E.1.CEL</td><td>1.79</td><td>115</td><td>61.00%</td><td>37.50%</td><td>1.50%</td><td>1.46</td><td>0.77</td><td>1</td><td>Y</td></tr>
</table>

</td>
</tr>
</table>
</Blockquote>
</p>


<P class="subtitle">Data source acknowledgment:</P>
<Blockquote>

<P>Some text here</P>
</Blockquote>



<P class="subtitle">Information about this text file:</P>
<Blockquote>
<P>Some text here</P>
</Blockquote>

<!-- GEO SERIES DATA GOES HERE -->
<P class="subtitle">GEO Series Data: This section to be used for GEO submission of the whole series of arrays</P>
<Blockquote>
<P><B>GSE Series</B>
<P><B>Status</B>
<P><B>Title</B>
<P><B>Organism(s)</B>
<P><B>Experiment type</B>
<P><B>Summary</B>

<P><B>Overall design</B>
<P><B>Contributor(s)</B>

<P><B>Citation(s)</B>

<P>
<BR><B>Submission date</B>
<BR><B>Contact name</B>
<BR><B>E-mails</B>
<BR><B>Phone</B>
<BR><B>FAX</B>
<BR><B>URL</B>
<BR><B>Organization name</B>
<BR><B>Department(s)</B>
<BR><B>Laboratory(s)</B>
<BR><B>Street address</B>
<BR><B>City</B>
<BR><B>State/province</B>
<BR><B>ZIP/Postal code</B>
<BR><B>Country</B>


<P><B>Platforms</B>
<P><B>Samples</B>
<!-- GEO SERIES DATA ENDS HERE -->

</P></Blockquote>




</TR></TABLE>
</TD>
</TR>
<TR>
<TD align=center bgColor=#ddddff class="solidBorder">
<!--Start of footer-->
<TABLE width="90%">
<script language='JavaScript' src='/javascript/footer.js'></script>
<TR>
<TD colspan=3 class="fs12">
<UL>

</UL>
</TD>
</TR>
</TABLE>
<!--End of footer-->
</TD>
</TR>
</TABLE>
<!-- /Footer -->
<!-- menu script itself. you should not modify this file -->
<script language="JavaScript" src="/javascript/menu_new.js"></script>
<!-- items structure. menu hierarchy and links are stored there -->
<script language="JavaScript" src="/javascript/menu_items.js"></script>
<!-- files with geometry and styles structures -->
<script language="JavaScript" src="/javascript/menu_tpl.js"></script>
<script language="JavaScript">
<!--//
// Note where menu initialization block is located in HTML document.
// Don't try to position menu locating menu initialization block in
// some table cell or other HTML element. Always put it before </body>
// each menu gets two parameters (see demo files)
// 1. items structure
// 2. geometry structure
new menu (MENU_ITEMS, MENU_POS);
// make sure files containing definitions for these variables are linked to the document
// if you got some javascript error like "MENU_POS is not defined", then you've made syntax
// error in menu_tpl.js file or that file isn't linked properly.

// also take a look at stylesheets loaded in header in order to set styles
//-->

</script>
</BODY>
</HTML>
