<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.0 Transitional//EN">
<HTML><HEAD><TITLE>About the HZI Regulatory T Cell mRNA data set of Feb 2011 on GN</TITLE>
<META http-equiv=Content-Type content="text/html; charset=iso-8859-1">
<LINK REL="stylesheet" TYPE="text/css" HREF='/css/general.css'>
<LINK REL="stylesheet" TYPE="text/css" HREF='/css/menu.css'>

</HEAD>
<BODY  bottommargin="2" leftmargin="2" rightmargin="2" topmargin="2" text=#000000 bgColor=#ffffff>
<TABLE cellSpacing=5 cellPadding=4 width="100%" border=0>
<TBODY>
<TR>
<script language="JavaScript" src="/javascript/header.js"></script>
</TR>
<TR>
<TD bgColor=#eeeeee class="solidBorder">
<Table width= "100%" cellSpacing=0 cellPadding=5><TR>
<!-- Body Start from Here -->
<TD valign="top" height="200" width="100%" bgcolor="#eeeeee">
<P class="title">

Helmholtz Zentrum f&uuml;r Infektionsforschung (HZI) T-Regulatory Cell Affymetrix M430v2 February 2011 RMA Data Set <A HREF="/webqtl/main.py?FormID=editHtml"><img src="/images/modify.gif" alt="modify this page" border= 0 valign="middle"></A><BR><BR>Accession number: <A HREF="/webqtl/main.py?FormID=sharinginfo&GN_AccessionId=122">GN122</A></P>



<P class="subtitle">&nbsp;&nbsp;&nbsp;&nbsp;Summary:</P>

<Blockquote>
<Blockquote>
ERROR-CHECKED FIRST PHASE PRIVATE TEST DATA SET.  This data set provides estimates of gene expression in <A HREF="http://en.wikipedia.org/wiki/Regulatory_T_cell" class="fs14" target="_empty">regulatory T cells</A> (CD4+CD25+) of BXD strains. Data were generated by Prof. Dr. Klaus Schughart and colleagues at the <A HREF="http://www.helmholtz-hzi.de/en/" class="fs14" target="_empty">Helmholtz Centre for Infection Research (HZI)</A>. Samples were processed using a total of 35 <span class=SpellE>Affymetrix</span> MOE 430 2.0 short oligomer microarrays, of which 33 passed stringent quality control and error checking.

<p>This is a private test data set. Please contact Dr. Klaus Schughart <klaus.schughart@helmholtz-hzi.de>for early access.


</Blockquote>
</Blockquote>
<P class="subtitle">&nbsp;&nbsp;&nbsp;&nbsp;About the cases used to generate this set of data:</P>
<Blockquote>

<Blockquote>
<P>
Parental and 31 BXD lines were studied. Mice were received from Jackson Laboratory, or from The Oak Ridge National and were bred in the facility of the Neuro-BSIK consortium (VU University Amsterdam). The data set includes expression values for 18 of the BXD strains made by Benjamin Taylor at the Jackson Laboratory in the 1970s and 1990s (BXD1 through BXD40, as well as the two parental strains, C57BL/6J and DBA/2J. All of these strains are fully inbred, many well beyond the 100th filial (F) generation of inbreeding.

<P>BXD spleen sample pools (from 2-3 mice) were obtained from a pathogen-free mice of the Dutch Mouse <span class=SpellE>Phenomics</span> Consortium (MPC) in Amsterdam. Mice were imported into the central animal
facility at the HZI and kept in a pathogen-free vivarium. Mice were euthanized using CO2 and spleenocytes wre prepared. Most mice were between 17 and 22 weeks of age when samples were collected. FACS sorting was used to select the CD4-positive T cells. These cells were further separated into CD4+CD25+ and CD4+CD25- pools.

<P>Error-checking strain identity. A set of more than 20 probe sets with Mendelian segregation patterns in this HZI data set were used to confirm strain identify in early June, 2007. Two errors were detected and rectified. As of June 22, 2007, data are registered correctly. Prior to June 22, 2007, data listed as strains BXD33 and BXD39 were essentially hybrid (mixed) data sets.

<P>On Aug 23, 2007, we loaded the final QTL Reaper data into GeneNetwork for the corrected data set.   The maximum LRS generated by any probe set is 84.6 for 1436240_at (<I>Tra2a</I>). A total of 41 probe sets are associated with QTLs that have LRS values above 46 (LOD > 10).

<P>Sex of samples is listed below in Table 1. In brief, data for BXD14 and 23 are male-only samples, whereas BXD12, 16, 31, 34, 36 and  C57BL/6J are from female-only samples. All other samples (DBA/2J, BXD1, 2, 6, 9, 11, 18, 21 32, 33, 39, 40) consist of one male and one female array. The sex of samples can be independently validated using the <I>Xist</I> probe set (1427262_at).


<P>
<IMG src="http://www.genenetwork.org/images/upload/HZi_sex.gif" valign="top">


<DIR><P><SMALL><B>Figure 1:</B> The expression of <I>Xist</I> can be used as an independent marker for sex. Xist is expressed at very low levels (noise) in male samples (far left) and at high values in females (far right). Sex-balanced samples (middle) have high variance due to the inclusion of one array per sex.  </SMALL></P>
</DIR>



<P class="subtitle">&nbsp;&nbsp;&nbsp;&nbsp;Table 1</P>
<blockquote>
<table border="0" cellpadding="0" cellspacing="0" bgcolor="#000000" width="600">
<tr><td>
<table width="600" border="0" cellpadding="5" cellspacing="1" align="left">
<tr bgcolor="royalblue"><td><font color=#FFFFFF>Index</font></td><td><font color=#FFFFFF>ProbeSet ID</font></td><td><font color=#FFFFFF>Sample Description</font></td><td><font color=#FFFFFF>Sex</font></td><td><font color=#FFFFFF>Strain</font></td><td><font color=#FFFFFF>cd25</font></td><td><font color=#FFFFFF>Microarray</font></td><td><font color=#FFFFFF>Short Description</font></td><td><font color=#FFFFFF>Age</font></td><td><font color=#FFFFFF>Pool No.</font></td><td><font color=#FFFFFF>Pool members (animal number)</font></td><td><font color=#FFFFFF>Date of preparation</font></td></tr>
<tr bgcolor="#eeeeee"><td>1</td><td>HZI1008</td><td>BXD-06f (f1) CD25+</td><td>F</td><td>BXD6</td><td>CD25+</td><td>Yes</td><td>BXD-06f</td><td>17</td><td>f1</td><td>1,3,4</td><td>1-31-2006</td></tr>
<tr bgcolor="#eeeeee"><td>2</td><td>HZI1009</td><td>BXD-06m (m2) CD25+</td><td>M</td><td>BXD6</td><td>CD25+</td><td>Yes</td><td>BXD-06m</td><td>18</td><td>m2</td><td>5,6,7</td><td>1-31-2006</td></tr>
<tr bgcolor="#eeeeee"><td>3</td><td>HZI1010</td><td>BXD-14m (m3) CD25+</td><td>M</td><td>BXD14</td><td>CD25+</td><td>Yes</td><td>BXD-14m</td><td>17</td><td>m3</td><td>1,3,4</td><td>1-31-2006</td></tr>
<tr bgcolor="#eeeeee"><td>4</td><td>HZI1013</td><td>BXD-40f (f6) CD25+</td><td>F</td><td>BXD40</td><td>CD25+</td><td>Yes</td><td>BXD-40f</td><td>17</td><td>f6</td><td>1,2,3</td><td>2-1-2006</td></tr>
<tr bgcolor="#eeeeee"><td>5</td><td>HZI1014</td><td>BXD-40m (m7) CD25+</td><td>M</td><td>BXD40</td><td>CD25+</td><td>Yes</td><td>BXD-40m</td><td>17</td><td>m7</td><td>5,6,7</td><td>2-2-2006</td></tr>
<tr bgcolor="#eeeeee"><td>6</td><td>HZI1015</td><td>BXD-02f (f8) CD25+</td><td>F</td><td>BXD2</td><td>CD25+</td><td>Yes</td><td>BXD-02f</td><td>17</td><td>f8</td><td>1,2,3</td><td>2-14-2006</td></tr>
<tr bgcolor="#eeeeee"><td>7</td><td>HZI1016</td><td>BXD-02m (m20) CD25+</td><td>M</td><td>BXD2</td><td>CD25+</td><td>Yes</td><td>BXD-02m</td><td>21</td><td>m20</td><td>4,5,6</td><td>4-6-2006</td></tr>
<tr bgcolor="#eeeeee"><td>8</td><td>HZI1017</td><td>BXD-11f (f30) CD25+</td><td>F</td><td>BXD11</td><td>CD25+</td><td>Yes</td><td>BXD-11f</td><td>17</td><td>f30</td><td>3,4,5</td><td>5-11-2006</td></tr>
<tr bgcolor="#eeeeee"><td>9</td><td>HZI1018</td><td>BXD-11m (m9) CD25+</td><td>M</td><td>BXD11</td><td>CD25+</td><td>Yes</td><td>BXD-11m</td><td>18</td><td>m9</td><td>1,2</td><td>2-14-2006</td></tr>
<tr bgcolor="#eeeeee"><td>10</td><td>HZI1019</td><td>BXD-12f (f10) CD25+</td><td>F</td><td>BXD12</td><td>CD25+</td><td>Yes</td><td>BXD-12f</td><td>17</td><td>f10</td><td>1,2,3</td><td>2-14-2006</td></tr>
<tr bgcolor="#eeeeee"><td>11</td><td>HZI1020</td><td>BXD-39f (f23) CD25+</td><td>F</td><td>BXD39</td><td>CD25+</td><td>Yes</td><td>BXD-39f</td><td>19</td><td>f23</td><td>4,5,6</td><td>4-11-2006</td></tr>
<tr bgcolor="#eeeeee"><td>12</td><td>HZI1021</td><td>BXD-33m (m11) CD25+</td><td>M</td><td>BXD33</td><td>CD25+</td><td>Yes</td><td>BXD-33m</td><td>17</td><td>m11</td><td>1,2</td><td>2-14-2006</td></tr>
<tr bgcolor="#eeeeee"><td>13</td><td>HZI1022</td><td>BXD-18f (f14) CD25+</td><td>F</td><td>BXD18</td><td>CD25+</td><td>Yes</td><td>BXD-18f</td><td>17</td><td>f14</td><td>3,4,5</td><td>2-15-2006</td></tr>
<tr bgcolor="#eeeeee"><td>14</td><td>HZI1023</td><td>BXD-18m (m13) CD25+</td><td>M</td><td>BXD18</td><td>CD25+</td><td>Yes</td><td>BXD-18m</td><td>18</td><td>m13</td><td>7,8</td><td>2-15-2006</td></tr>
<tr bgcolor="#eeeeee"><td>15</td><td>HZI1024</td><td>BXD-23m (m15) CD25+</td><td>M</td><td>BXD23</td><td>CD25+</td><td>Yes</td><td>BXD-23m</td><td>18</td><td>m15</td><td>1,2,3</td><td>2-15-2006</td></tr>
<tr bgcolor="#eeeeee"><td>16</td><td>HZI1026</td><td>BXD-09f (f17) CD25+</td><td>F</td><td>BXD9</td><td>CD25+</td><td>Yes</td><td>BXD-09f</td><td>21</td><td>f17</td><td>1,2,3</td><td>4-5-2006</td></tr>
<tr bgcolor="#eeeeee"><td>17</td><td>HZI1028</td><td>BXD-09m (m35) CD25+</td><td>M</td><td>BXD9</td><td>CD25+</td><td>Yes</td><td>BXD-09m</td><td>15</td><td>m35</td><td>7,8,9</td><td>7-7-2006</td></tr>
<tr bgcolor="#eeeeee"><td>18</td><td>HZI1029</td><td>BXD-32f (f18) CD25+</td><td>F</td><td>BXD32</td><td>CD25+</td><td>Yes</td><td>BXD-32f</td><td>21</td><td>f18</td><td>1,2,3</td><td>4-6-2006</td></tr>
<tr bgcolor="#eeeeee"><td>19</td><td>HZI1030</td><td>BXD-32m (m19) CD25+</td><td>M</td><td>BXD32</td><td>CD25+</td><td>Yes</td><td>BXD-32m</td><td>22</td><td>m19</td><td>1,2,3</td><td>4-6-2006</td></tr>
<tr bgcolor="#eeeeee"><td>20</td><td>HZI1031</td><td>BXD-33f (f22) CD25+</td><td>F</td><td>BXD33</td><td>CD25+</td><td>Yes</td><td>BXD-33f</td><td>18</td><td>f22</td><td>2,3,4</td><td>4-11-2006</td></tr>
<tr bgcolor="#eeeeee"><td>21</td><td>HZI1032</td><td>BXD-39m (m29) CD25+</td><td>M</td><td>BXD39</td><td>CD25+</td><td>Yes</td><td>BXD-39m</td><td>17</td><td>m29</td><td>5,6,7</td><td>5-10-2006</td></tr>
<tr bgcolor="#eeeeee"><td>22</td><td>HZI1033</td><td>BXD-01f (f32) CD25+</td><td>F</td><td>BXD1</td><td>CD25+</td><td>Yes</td><td>BXD-01f</td><td>18</td><td>f32</td><td>3,4</td><td>7-6-2006</td></tr>
<tr bgcolor="#eeeeee"><td>23</td><td>HZI1034</td><td>BXD-01m (m31) CD25+</td><td>M</td><td>BXD1</td><td>CD25+</td><td>Yes</td><td>BXD-01m</td><td>18</td><td>m31</td><td>1,2</td><td>7-6-2006</td></tr>
<tr bgcolor="#eeeeee"><td>24</td><td>HZI1035</td><td>BXD-16f (f26) CD25+</td><td>F</td><td>BXD16</td><td>CD25+</td><td>Yes</td><td>BXD-16f</td><td>18</td><td>f26</td><td>1,2,3</td><td>4-12-2006</td></tr>
<tr bgcolor="#eeeeee"><td>25</td><td>HZI1036</td><td>BXD-21f (f25) CD25+</td><td>F</td><td>BXD21</td><td>CD25+</td><td>Yes</td><td>BXD-21f</td><td>19</td><td>f25</td><td>5,6,7</td><td>4-12-2006</td></tr>
<tr bgcolor="#eeeeee"><td>26</td><td>HZI1037</td><td>BXD-21m (m24) CD25+</td><td>M</td><td>BXD21</td><td>CD25+</td><td>Yes</td><td>BXD-21m</td><td>18</td><td>m24</td><td>1,2,3</td><td>4-12-2006</td></tr>
<tr bgcolor="#eeeeee"><td>27</td><td>HZI1039</td><td>BXD-31f (f34) CD25+</td><td>F</td><td>BXD31</td><td>CD25+</td><td>Yes</td><td>BXD-31f</td><td>16</td><td>f34</td><td>1,2,3</td><td>7-7-2006</td></tr>
<tr bgcolor="#eeeeee"><td>28</td><td>HZI1040</td><td>C57BL/6Jf (f28) CD25+</td><td>F</td><td>C57BL/6J</td><td>CD25+</td><td>Yes</td><td>C57BL/6Jf</td><td>16</td><td>f28</td><td>1,2,3</td><td>5-10-2006</td></tr>
<tr bgcolor="#eeeeee"><td>29</td><td>HZI1041</td><td>DBA/2Jf (f27) CD25+</td><td>F</td><td>DBA/2J</td><td>CD25+</td><td>Yes</td><td>DBA/2Jf</td><td>16</td><td>f27</td><td>5,6,7</td><td>5-10-2006</td></tr>
<tr bgcolor="#eeeeee"><td>30</td><td>HZI1042</td><td>DBA/2Jm (m21) CD25+</td><td>M</td><td>DBA/2J</td><td>CD25+</td><td>Yes</td><td>DBA/2Jm</td><td>21</td><td>m21</td><td>1,2,3</td><td>4-11-2006</td></tr>
<tr bgcolor="#eeeeee"><td>31</td><td>HZI1487</td><td>BXD-08f (f67) CD25+</td><td>F</td><td>BXD8</td><td>CD25+</td><td>Yes</td><td>BXD-08f</td><td>11</td><td>f67</td><td>4,5,6</td><td>6-25-2007</td></tr>
<tr bgcolor="#eeeeee"><td>32</td><td>HZI1488</td><td>BXD-08m (m66) CD25+</td><td>M</td><td>BXD8</td><td>CD25+</td><td>Yes</td><td>BXD-08m</td><td>17</td><td>m66</td><td>1,2,3</td><td>6-25-2007</td></tr>
<tr bgcolor="#eeeeee"><td>33</td><td>HZI1489</td><td>BXD-16m (m36) CD25+</td><td>M</td><td>BXD16</td><td>CD25+</td><td>Yes</td><td>BXD-16m</td><td>20, 16</td><td>m36</td><td>5,6,7</td><td>8-28-2006</td></tr>
<tr bgcolor="#eeeeee"><td>34</td><td>HZI1490</td><td>BXD-12m (m42) CD25+</td><td>M</td><td>BXD12</td><td>CD25+</td><td>Yes</td><td>BXD-12m</td><td>20</td><td>m42</td><td>5,6,7</td><td>10-23-2006</td></tr>
<tr bgcolor="#eeeeee"><td>35</td><td>HZI1491</td><td>BXD-13f (f44) CD25+</td><td>F</td><td>BXD13</td><td>CD25+</td><td>Yes</td><td>BXD-13f</td><td>15</td><td>f44</td><td>1,2,3</td><td>12-13-2006</td></tr>
<tr bgcolor="#eeeeee"><td>36</td><td>HZI1492</td><td>BXD-13m (m45) CD25+</td><td>M</td><td>BXD13</td><td>CD25+</td><td>Yes</td><td>BXD-13m</td><td>15</td><td>m45</td><td>4,5,6,7</td><td>12-13-2006</td></tr>
<tr bgcolor="#eeeeee"><td>37</td><td>HZI1493</td><td>BXD-14f (f48) CD25+</td><td>F</td><td>BXD14</td><td>CD25+</td><td>Yes</td><td>BXD-14f</td><td>16</td><td>f48</td><td>5,6,7</td><td>2-15-2007</td></tr>
<tr bgcolor="#eeeeee"><td>38</td><td>HZI1494</td><td>BXD-19f (f64) CD25+</td><td>F</td><td>BXD19</td><td>CD25+</td><td>Yes</td><td>BXD-19f</td><td>19</td><td>f64</td><td>7,8,9</td><td>6-20-2007</td></tr>
<tr bgcolor="#eeeeee"><td>39</td><td>HZI1495</td><td>BXD-19m (m46) CD25+</td><td>M</td><td>BXD19</td><td>CD25+</td><td>Yes</td><td>BXD-19m</td><td>16</td><td>m46</td><td>4,5,6</td><td>12-15-2006</td></tr>
<tr bgcolor="#eeeeee"><td>40</td><td>HZI1499</td><td>BXD-28m (m43) CD25+</td><td>M</td><td>BXD28</td><td>CD25+</td><td>Yes</td><td>BXD-28m</td><td>17,2</td><td>m43</td><td>1,2,3</td><td>10-23-2006</td></tr>
<tr bgcolor="#eeeeee"><td>41</td><td>HZI1500</td><td>BXD-42f (f49) CD25+</td><td>F</td><td>BXD42</td><td>CD25+</td><td>Yes</td><td>BXD-42f</td><td>17</td><td>f49</td><td>??</td><td>3-8-2007</td></tr>
<tr bgcolor="#eeeeee"><td>42</td><td>HZI1502</td><td>F1 (BXD)m (f50) CD25+</td><td>M</td><td>B6D2F1</td><td>CD25+</td><td>Yes</td><td>F1 (BXD)m</td><td>15</td><td>m51</td><td>1,2,3,</td><td>4-18-2007</td></tr>
<tr bgcolor="#eeeeee"><td>43</td><td>HZI1503</td><td>F1 (BXD)m (m51) CD25+</td><td>F</td><td>B6D2F1</td><td>CD25+</td><td>Yes</td><td>F1 (BXD)f</td><td>15</td><td>f50</td><td>1,2,3</td><td>4-18-2007</td></tr>
<tr bgcolor="#eeeeee"><td>44</td><td>HZI1504</td><td>BXD-86f (f52) CD25+</td><td>F</td><td>BXD86</td><td>CD25+</td><td>Yes</td><td>BXD-86f</td><td>16</td><td>f52</td><td>1,2,3</td><td>4-18-2007</td></tr>
<tr bgcolor="#eeeeee"><td>45</td><td>HZI1505</td><td>BXD-43f (f53) CD25+</td><td>F</td><td>BXD43</td><td>CD25+</td><td>Yes</td><td>BXD-43f</td><td>16</td><td>f53</td><td>1,2,3</td><td>4-23-2007</td></tr>
<tr bgcolor="#eeeeee"><td>46</td><td>HZI1506</td><td>BXD-44f (f54) CD25+</td><td>F</td><td>BXD44</td><td>CD25+</td><td>Yes</td><td>BXD-44f</td><td>18</td><td>f54</td><td>1,2,3</td><td>4-23-2007</td></tr>
<tr bgcolor="#eeeeee"><td>47</td><td>HZI1507</td><td>BXD-45f (f55) CD25+</td><td>F</td><td>BXD45</td><td>CD25+</td><td>Yes</td><td>BXD-45f</td><td>19</td><td>f55</td><td>1,2,3</td><td>4-23-2007</td></tr>
<tr bgcolor="#eeeeee"><td>48</td><td>HZI1508</td><td>BXD-62f (f56) CD25+</td><td>F</td><td>BXD62</td><td>CD25+</td><td>Yes</td><td>BXD-62f</td><td>17</td><td>f56</td><td>1,2,3</td><td>4-26-2007</td></tr>
<tr bgcolor="#eeeeee"><td>49</td><td>HZI1509</td><td>BXD-73f (f57) CD25+</td><td>F</td><td>BXD73</td><td>CD25+</td><td>Yes</td><td>BXD-73f</td><td>18</td><td>f57</td><td>1,2,3</td><td>4-26-2007</td></tr>
<tr bgcolor="#eeeeee"><td>50</td><td>HZI1510</td><td>BXD-51f (f59) CD25+</td><td>F</td><td>BXD51</td><td>CD25+</td><td>Yes</td><td>BXD-51f</td><td>22</td><td>f59</td><td>1,2,3</td><td>6-18-2007</td></tr>
<tr bgcolor="#eeeeee"><td>51</td><td>HZI1523</td><td>BXD-75f (f58) CD25+</td><td>F</td><td>BXD75</td><td>CD25+</td><td>Yes</td><td>BXD-75f</td><td>15,17</td><td>f58</td><td>1,2,3</td><td>4-26-2007</td></tr>
<tr bgcolor="#eeeeee"><td>52</td><td>HZI1525</td><td>BXD-29m (m37) CD25+</td><td>M</td><td>BXD29</td><td>CD25+</td><td>Yes</td><td>BXD-29m</td><td>20, 16</td><td>m37</td><td>1,2,3</td><td>8-29-2006</td></tr>
<tr bgcolor="#eeeeee"><td>53</td><td>HZI1526</td><td>BXD-34f (f4) CD25+</td><td>F</td><td>BXD34</td><td>CD25+</td><td>Yes</td><td>BXD-34f</td><td>17</td><td>f4</td><td>1,2,3</td><td>2-1-2006</td></tr>
<tr bgcolor="#eeeeee"><td>54</td><td>HZI1940</td><td>BXD-27m (m39) CD25+</td><td>M</td><td>BXD27</td><td>CD25+</td><td>Yes</td><td>BXD-27m</td><td>18 - 20</td><td>m39</td><td>1,3,4</td><td>9-1-2006</td></tr>
<tr bgcolor="#eeeeee"><td>55</td><td>HZI1941</td><td>BXD-42m (m47) CD25+</td><td>M</td><td>BXD42</td><td>CD25+</td><td>Yes</td><td>BXD-42m</td><td>15,16</td><td>m47</td><td>1,2,3</td><td>12-15-2006</td></tr>
<tr bgcolor="#eeeeee"><td>56</td><td>HZI1942</td><td>BXD-34m (m5) CD25+</td><td>M</td><td>BXD34</td><td>CD25+</td><td>Yes</td><td>BXD-34m</td><td>17</td><td>m5</td><td>5,7,8</td><td>2-1-2006</td></tr>
<tr bgcolor="#eeeeee"><td>57</td><td>HZI1943</td><td>BXD-38f (f70) CD25+</td><td>F</td><td>BXD38</td><td>CD25+</td><td>Yes</td><td>BXD-38f</td><td>13</td><td>f70</td><td>4,5,6,7</td><td>2-1-2008</td></tr>
<tr bgcolor="#eeeeee"><td>58</td><td>HZI1944</td><td>BXD-31m (m69) CD25+</td><td>M</td><td>BXD31</td><td>CD25+</td><td>Yes</td><td>BXD-31m</td><td>14</td><td>m69</td><td>4,5,6</td><td>2-1-2008</td></tr>
<tr bgcolor="#eeeeee"><td>59</td><td>HZI1945</td><td>BXD-27f (f12) CD25+</td><td>F</td><td>BXD27</td><td>CD25+</td><td>Yes</td><td>BXD-27f</td><td>18</td><td>f12</td><td>1,2</td><td>2-15-2006</td></tr>
<tr bgcolor="#eeeeee"><td>60</td><td>HZI1946</td><td>BXD-38m (m63) CD25+</td><td>M</td><td>BXD38</td><td>CD25+</td><td>Yes</td><td>BXD-38m</td><td>18</td><td>m63</td><td>1,2,3</td><td>6-20-2007</td></tr>
<tr bgcolor="#eeeeee"><td>61</td><td>HZI1947</td><td>BXD-23f (f62) CD25+</td><td>F</td><td>BXD23</td><td>CD25+</td><td>Yes</td><td>BXD-23f</td><td>21</td><td>f62</td><td>1,2,3</td><td>6-20-2007</td></tr>
<tr bgcolor="#eeeeee"><td>62</td><td>HZI1948</td><td>BXD-28f (f61) CD25+</td><td>F</td><td>BXD28</td><td>CD25+</td><td>Yes</td><td>BXD-28f</td><td>22</td><td>f61</td><td>1,2,3</td><td>6-18-2007</td></tr>
</table>
</blockquote>

		</TR></TABLE>
	

<P class="subtitle">
About the Array Platform:
<Blockquote>


<p>The Affymetrix M430 2.0 array consists of approximately 992,936 useful 25-nucleotide probes that estimate the expression of approximately 39,000 transcripts, including a majority of known genes and expressed sequence tags. The array sequences were selected late in 2002 using NCBI Build 107 by Affymetrix. The UTHSC GN group continuously reannotated probe sets on this array, producing more accurate data on probe and probe set targets. All probes have also be aligned to the most recent assembly of the Mouse Genome using Jim Kent's BLAT program.
</Blockquote>



<P class="subtitle">Methods:</P>

<Blockquote>
 <p>Parental and BXD lines were received from Jackson Laboratory, or from Oak Ridge Laboratory (BXD43, BXD51, BXD61, BXD62, BXD65, BXD68, BXD69, BXD73, BXD75, BXD87, BXD90), and were bred in the facility of the Neuro-BSIK consortium (VU University Amsterdam). Female mice 3 per strain were housed on sawdust in standard Makrolon type II cages with food (Harlan Teklad 2018) and water ad libitum under specific pathogen free conditions. For the analysis, mice were transferred to the animal facility in Braunschweig and adapted for at least two weeks to the new environment before preparing the spleen cells. All protocols involving mice were approved by national animal welfare committees.<p>
  For sorting of Tregs and Th cells, splenocytes from 31 BXD recombinant inbred strains as well as from the parental mouse lines DBA/2J and C57BL/6J were isolated by flushing the spleens with erythrocyte-lysis-buffer. Cells were collected by centrifugation, re-suspended in cold FACS-buffer (PBS / 2% FCS / 0,5 mM EDTA). After passing the cells through a 100 µm cell strainer and an additional washing step with FACS-buffer, splenocytes were stained with anti-CD4-APC and anti-CD25-PE for 10 minutes at 4°C, washed and re-suspended in FACS-buffer. CD4+ T cells were separated into CD4+CD25+ Tregs and CD4+CD25- Th cells using a MoFlo cell sorter (Cytomation) and purity of the sorted T cell subsets reached 95-97%.<p>
  Quality and integrity of the total RNA isolated from 1x10<sup>5</sup> cells was controlled by running all samples on an Agilent Technologies 2100 Bioanalyzer (Agilent Technologies; Waldbronn, Germany). RNA amplification and labeling was done according to manufactures protocol (Small Sample Target Labeling Assay Version II, Affymetrix; Santa Clara, CA).  The concentration of biotin-labeled cRNA was determined by UV absorbance. In all cases, 10 µg of each biotinylated cRNA preparation were fragmented and placed in a hybridization cocktail containing four biotinylated hybridization controls (BioB, BioC, BioD, and Cre) as recommended by the manufacturer. Samples were hybridized to an identical lot of Affymetrix MOE430 2.0 for 16 hours at 46°C. After hybridisation the GeneChips were washed and stained using the Affymetrix´s recommended EukGE-WS2v5 protocol for GeneChip®  Fluidics FS400 station.  Images were scanned using GeneChip® Scanner 3000 under the control of GCOS 1.3 software package (Affymetrix; Santa Clara, CA).</p>
</Blockquote>
<p class=MsoNormal style='mso-margin-top-alt:auto;mso-margin-bottom-alt:auto'><b><span
lang=EN-GB style='font-size:11.0pt;font-family:Verdana;color:#000082;
mso-ansi-language:EN-GB'>About the data processing:
</b>
<Blockquote>

<p>Microarray data then was preprocessed using the RMA method [bolstad] and subsequently batch corrected [Alberts et al]. In this study, RNA was extracted at three different points in time for the Treg samples and also microarray processing was performed at three different points in time. Similarly, the Th samples were processed in two batches. Therefore, we performed a batch correction for both cell types using the following ANOVA model before further analysis of the data.<br />
<em>y<sub>i</sub> = </em><em>μ</em><em> + B<sub>i</sub> + e<sub>i</sub></em><br />
Where <em>y<sub>i</sub></em> is the expression level of the <em>i</em><sup>th</sup> microarray, <em>μ</em> is the overall mean, B<em><sub>i</sub> </em>is the batch to which the <em>i</em>th individual belongs and <em>e<sub>i</sub> </em>is the residual error.<br />
Batch corrected data sets were then preprocessed before transferring them to the GeneNetwork (GN) database: Adding an offset of 1 unit to each signal intensity value to ensure that the logarithm of all values were positive, computing the log2 value, performing a quantile normalization of the log2 values for the total set of arrays using the same initial steps used by the RMA transform, computing the Z scores for each cell value, multiplying all Z scores by 2 and adding 8 to the value of all Z scores. The advantage of this variant of a Z transformation is that all values are positive and that 1 unit represents approximately a 2-fold difference in expression as determined using the spike-in control probe sets. The mean values were subsequently calculated if multiple samples from one BXD line were recorded (male and females or replicates). </p>

</Blockquote>
<P class="subtitle">Acknowledgment:</P>
<Blockquote>

<p>These data were generated by Prof. Dr. Klaus Schughart (Department of Experimental Mouse Genetics) and Dr. Dunja Bruder  (Research Group Immune Regulation) at the Helmholtz Center for Infection Research with the help of Dr. Lothar Gr&ouml;be (FACS sorting, Research Group Mucosal Immunity).

<p>Funding was provided by the Helmholtz Association and publicly funded research projects awarded to Drs. Klaus
Schughart and Dunja Bruder.</p>

</Blockquote>

<P class="subtitle">About this text file:</p>
<Blockquote>

<P>This text file was generated by KS on July, 18 2011.
</P></Blockquote>

<P>



<b style='mso-bidi-font-weight:normal'><span lang=EN-US style='font-size:14.0pt;
font-family:"Courier New";mso-fareast-font-family:"Times New Roman";mso-ansi-language:
EN-US;mso-fareast-language:DE;mso-bidi-language:AR-SA'><br clear=all
style='page-break-before:always'>
</span></b>
</div>
</TR></TABLE>
</TD>
</TR>
<TR>
<TD align=center bgColor=#ddddff class="solidBorder">




<!--Start of footer-->
<TABLE width="90%">
<script language='JavaScript' src='/javascript/footer.js'></script>
</TABLE>
<!--End of footer-->




</TD>
</TR>
</TABLE>
<!-- /Footer -->
<!-- menu script itself. you should not modify this file -->
<script language="JavaScript" src="/javascript/menu_new.js"></script>
<!-- items structure. menu hierarchy and links are stored there -->
<script language="JavaScript" src="/javascript/menu_items.js"></script>
<!-- files with geometry and styles structures -->
<script language="JavaScript" src="/javascript/menu_tpl.js"></script>
<script language="JavaScript">
<!--//
// Note where menu initialization block is located in HTML document.
// Don't try to position menu locating menu initialization block in
// some table cell or other HTML element. Always put it before </body>
// each menu gets two parameters (see demo files)
// 1. items structure
// 2. geometry structure
new menu (MENU_ITEMS, MENU_POS);
// make sure files containing definitions for these variables are linked to the document
// if you got some javascript error like "MENU_POS is not defined", then you've made syntax
// error in menu_tpl.js file or that file isn't linked properly.

// also take a look at stylesheets loaded in header in order to set styles
//-->
</script>
</BODY>
</HTML>
