<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.0 Transitional//EN">
<HTML><HEAD><TITLE>Arabidopsis Cross Information</TITLE>
<META http-equiv=Content-Type content="text/html; charset=iso-8859-1">
<LINK REL="stylesheet" TYPE="text/css" HREF='css/general.css'>
<LINK REL="stylesheet" TYPE="text/css" HREF='css/menu.css'>
<SCRIPT SRC="javascript/webqtl.js"></SCRIPT>


</HEAD>
<BODY  bottommargin="2" leftmargin="2" rightmargin="2" topmargin="2" text=#000000 bgColor=#ffffff>
<TABLE cellSpacing=5 cellPadding=4 width="100%" border=0>
	<TBODY>
	<TR>
		<script language="JavaScript" src="/javascript/header.js"></script>
	</TR>
	<TR>
		<TD bgColor=#eeeeee class="solidBorder">
		<Table width= "100%" cellSpacing=0 cellPadding=5><TR>
		<!-- Body Start from Here -->
		<TD valign="top" height="200" width="100%" bgcolor="#eeeeee">
		<P class="title"><I>Arabidopsis</I> Cross Information <A HREF="/webqtl/main.py?FormID=editHtml"><img src="images/modify.gif" alt="modify this page" border= 0 valign="middle"></A></P>


&nbsp;&nbsp;&nbsp;&nbsp;<class="subtitle">
Bay-0 x Shahdara (BXS):</A> 
		
<Blockquote>
The Bay-0 x Shahdara genetic reference population of 420 recombinant inbred lines (RIL) was created by Olivier Loudet and Sylvain Chaillou between 1997 and 2000 at the <A HREF=" http://w3.inra.fr/ " target="_blank" class="fs14">INRA</A> in Versailles, France. It is one of the largest publically available <I>Arabidopsis</I> RIL sets and more than 10 times larger than most rodent RILs. <I>Arabidopsis</I> is commonly known as mustard weed, and is one of the favored model organisms used by geneticists and developmental biologists. This plant has a genome consisting of <A HREF="http://www.inra.fr/qtlat/BayxSha/BayxShaMarker.htm" target="_blank" class="fs14">five</A> chromsomes and a total of 125 megabases of DNA, equivalent in length to a single human chromosome. However, the genome is nonetheless rich and contains approximately <A HREF="http://www.tigr.org/tdb/e2k1/ath1/intro.shtml" target="_blank" class="fs14">26,000</A> genes. Full sequence data are available for this species. (The Col-0 accession was sequenced.)

<P>Please site the following publication when using the Bay-0 x Shahdara (BXS) data sets:

<Blockquote>
Loudet O, Chaillou S, Camilleri C, Bouchez D, Daniel-Vedele F (<A HREF="http://www.inra.fr/qtlat/BayxSha/Loudet2002.pdf" target="_blank" class="fs14">2002</A>) Bay-0 x Shahdara recombinant inbred line population: a powerful tool for the genetic dissection of complex traits in Arabidopsis. <U>Theoretical and Applied Genetics</U> 104:1173-1184  (<A HREF="http://www.inra.fr/qtlat/BayxSha/Loudet2002.pdf" target="_blank" class="fs14">pdf</A>)
</Blockquote>

<P>The original set of 420 RILs were derived from a cross between Bay-0 (accession N954) and Shahdara (accession N929); two accessions obtained from the <A HREF="http://arabidopsis.info/" class="fs14" target="_blank">NASC</A> European Arabidopsis Stock Centre. Bay-0 and Shahdara were chosen because of their well characterized genetic, geographical, and ecological differences. Lines were propagated by single seed descent through the sixth generation (F6) without selection. One plant per line was then used for genotyping (420 RILs x 38 markers) and selfed to obtain F7 seeds. F8 seed stock generated by bulk multiplication of F7 plants are available for analysis for 411 of these RILs. Data sets in WebQTL include up to 415 BXS accessions and the two parental stock.</P>

<P><B>How to obtain these lines</B>: The entire extant set of 411 lines was donated to the <A HREF="http://seeds.nottingham.ac.uk/Nasc/action.lasso?-search" class="fs14" target="_blank">NASC</A> and <A HREF="http://www.arabidopsis.org/abrc/" class="fs14" target="_blank">ABRC</A> in 2002 and is available (Stock number <A HREF="http://seeds.nottingham.ac.uk/Nasc/action.lasso?-response=detail/detail_general.lasso&-token.maxrecords=50&-token.return=cart/carthitlist.lasso&-token.code=57920" class="fs14" target="_blank">N57920</A> and CS57920) for &#163;450 / US $720 (academic fee). 

<P>A core set of 18 lines are also available from the <A HREF="http://seeds.nottingham.ac.uk/Nasc/action.lasso?-search" class="fs14" target="_blank">NASC</A> and and ABRC (stock numbers <A HREF="http://seeds.nottingham.ac.uk/Nasc/action.lasso?-response=detail/detail_general.lasso&-token.maxrecords=50&-token.return=cart/carthitlist.lasso&-token.code=57922" class="fs14" target="_blank">N57922</A> and CS57922) for &#163;25 / US $40 (academic fee). This "Core-Pop18" set is not intended for systems genetics, but rather is used to evaluate trait variability and transgression. This set includes the parental accessions. 

<P>Single lines may also be purchased directly from NASC for &#163;2.25 each plus &#163;8.50 per order and from ABRC for US $4 per order (prices current as of April 2005). 

<P>Please contact <A HREF="mailto:loudet@versailles.inra.fr" class="fs14">Olivier Loudet</A> at loudet@versailles.inra.fr</a> if you have further questions on the availability of seed stock and comments or questions on the genotype or phenotype data sets deposited in the GeneNetwork.
</P>

<P><B>Note on nomenclature</B>: The widely accepted names of these accessions have the form: Bay-0 x Shahdara - 1, Bay-0 x Shahdara - 2, Bay-0 x Shahdara - 432, etc.  In the GeneNetwork we have shortened this long form to BXS001, BXS002, BXS432. 

<P>For more information on the Bay-0 x Shadara <I>Arabidopsis thaliana</I> cross, the genotypic data, the genetic maps, and the trait data please visit the INRA <A HREF="http://www.inra.fr/qtlat/BayxSha/index.htm" target="_blank" class="fs14">Versailles QTL Arabidopsis</A> site.

<P>For more background on the history, generation, and use of RILs as genetic reference populations for systems genetics please see Silver <A HREF="http://www.complextrait.org/archive/2001/HTML/silverbook/frame1.1.shtml" target="_blank" class="fs14">(1995)</A>. Additional useful literature links are provided in the <B>References</B> link at the top center of this page.
</P>

</Blockquote>


<P class="subtitle">&nbsp;&nbsp;&nbsp;&nbsp;About this file:</P> 
<Blockquote><P> The file started, April 12, 2005 by Olivier Loudet and RWW. Last update RWW, April 15, 2005.</P>
<TABLE width="100%"><TR><TD align="left"> <A HREF="http://dbsgap.versailles.inra.fr/vnat/"><IMG SRC="images/VNATCrop.gif"></A></TD><TD align="right">
<A HREF="http://www.dpw.wau.nl/natural/"><IMG SRC="images/naturalbig.jpg"></A></TD></TABLE></Blockquote>
		</TD>
		</TR></TABLE>
		</TD>
	</TR>
	<TR>
		<TD align=center bgColor=#ddddff class="solidBorder">
		<!--Start of footer-->
		<TABLE width="90%">
		<script language='JavaScript' src='/javascript/footer.js'></script>
		</TABLE>
		<!--End of footer-->
		</TD>
	</TR>
</TABLE>
<!-- /Footer -->
<script language="JavaScript" src="javascript/menu_new.js"></script>
<script language="JavaScript" src="javascript/menu_items.js"></script>
<script language="JavaScript" src="javascript/menu_tpl.js"></script>
<script language="JavaScript">
	<!--//
	new menu (MENU_ITEMS, MENU_POS);
	//-->
</script>
<script src="http://www.google-analytics.com/urchin.js" type="text/javascript">
</script>
<script type="text/javascript">
_uacct = "UA-3782271-1";
urchinTracker();
</script>
</BODY>
</HTML>
