<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.0 Transitional//EN">
<HTML><HEAD><TITLE>Scriptable Interface</TITLE>
<META http-equiv=Content-Type content="text/html; charset=iso-8859-1">
<LINK REL="stylesheet" TYPE="text/css" HREF='css/general.css'>
<LINK REL="stylesheet" TYPE="text/css" HREF='css/menu.css'>
<SCRIPT SRC="javascript/webqtl.js"></SCRIPT>


</HEAD>
<BODY  bottommargin="2" leftmargin="2" rightmargin="2" topmargin="2" text=#000000 bgColor=#ffffff>
<TABLE cellSpacing=5 cellPadding=4 width="100%" border=0>
	<TBODY>
	<TR>
	<script language="JavaScript" src="/javascript/header.js"></script>
	</TR>
<TR>
		<TD bgColor=#eeeeee class="solidBorder">
		<Table width= "100%" cellSpacing=0 cellPadding=5><TR>
<!-- Body Start from Here -->
		
<td width="62%" bgColor=#eeeeee valign=top>
		 

<P class="title">Overview  <A HREF="/webqtl/main.py?FormID=editHtml">
			<img src="images/modify.gif" alt="modify this page" border= 0 valign="middle"></A></P>
		 
<Blockquote>GeneNetwork (GN) provides a scriptable interface that allows other services and web users to retrieve data, map QTLs, generate lists of covariates, or to link to the most appropriate page of data for a particular tissue or transcript. This page documents the GN scriptable interface. GN functions are invoked using commands, which usually takes the form of a URL beginning

<br>
<span class="keywords">http://robot.genenetwork.org/webqtl/main.py?</span>...</Blockquote>
		 
<Blockquote>At the end of this type of command, after the question mark, you need to add a set of keywords and values separated by the <span class="keywords">=</span> character, for example, <span class="keywords">cmd=get</span> or <span class="keywords">probeset=100011_at</span>. Keywords and values that GN will recognize are listed and explained in the table below.
</Blockquote>
		 

<Blockquote>The keyword-value pairs should be separated by the <span class="keywords">&</span> character. A complete query might look like
<br>
<span class="keywords">cmd=get&probeset=100011_at&probe=all&format=column</span>

<br>
The order of the keyword-value pairs in the list is unimportant. However, the <span class="keywords">cmd</span> keyword is required and its value determines what other keywords are required or allowed. In examples, we always use the <span class="keywords">cmd</span> keyword first.</Blockquote>
		 
<Blockquote>
Putting the whole query together results in a string that looks like <span class="keywords">http://robot.genenetwork.org/webqtl/main.py?cmd=get&probeset=100011_at&probe=all&format=column</span>

<br>
Although this command is shown on two lines of text, it is really just one line, and there should be no carriage return or line break.

</Blockquote></td>
<td width="38%" bgColor=#eeeeee valign=top> 
<P class="title">Specific commands</P>

<Blockquote><span class="keywords"><a href="CGIDoc.html#genotype">cmd=genotype</a></span> retrieves genotype data for any of the supported sets of recombinant inbred lines.</Blockquote>

<Blockquote><span class="keywords"><a href="CGIDoc.html#trait">cmd=get</a></span> or <span class="keywords"><a href="CGIDoc.html#trait">cmd=trait</a></span> retrieves trait data from databases of published phenotypes or gene expression data. It can also retrieve marker genotypes as numerical codes for correlation searches. 
</Blockquote>

<Blockquote><span class="keywords"><a href="CGIDoc.html#map">cmd=map</a></span> retrieves trait data and searches for associations among BXD marker genotypes, returning marker names, regression coefficients and likelihood ratio statistics.</Blockquote>
		 
<Blockquote><span class="keywords"><a href="CGIDoc.html#interval">cmd=interval</a></span> retrieves trait data and performs interval mappng, returning map positions of peak likelihood ratio statistic.</Blockquote>
		 
<Blockquote><span class="keywords"><a href="CGIDoc.html#correlation">cmd=correlation</a></span> retrieves trait data and searches for correlations with traits in another 
(or the same) trait database. <span class="keywords">cmd=correlation</span> and <span class="keywords">cmd=pearson</span> use Pearson product-moment correlation. <span class="keywords">cmd=spearman</span> uses Spearman rank-sum correlation.</Blockquote></td>
</tr>
	




<TR>
		<TD colspan=2 bgColor=#eeeeee>
		<P class="title">Commands and Keywords</P>


<TABLE border="0" cellpadding="0" cellspacing="0" width="95%" bgcolor="#999999" align=center>
		<TR>
			<TD>
			<TABLE border="0" cellpadding="5" cellspacing="1" width="100%">
		 <tr bgcolor="royalblue"> 
			<td align="center">
				<FONT color="#FFFFFF"><STRONG>Function</strong></Font>
			</td>
			<td align="center">
				<FONT color="#FFFFFF"><STRONG>Command<br>
				keyword-value</strong></Font>
			</td>
			<td align="center"> 
				<FONT color="#FFFFFF"><STRONG>Other keywords</strong>
				<br>(<strong>bold</strong> keywords are required)</Font>
			</td>
			<td align="center">
				<FONT color="#FFFFFF"><STRONG>Allowed values</strong>
				<br>[Comments or descriptions in square brackets]</Font>
			</td>
		 </tr>
		 <tr bgcolor="#eeeeee"> 
	
<!--RETRIEVE GENOTYPE DATA-->		

<td><a name="genotype">Retrieve genotype data</a></td>
			<td class="keywords">cmd=genotype<br>
			 cmd=gen</td>
			<td class="keywords"><strong>riset</strong></td>
			<td><span class="keywords">axb</span>, <span class="keywords">bxa</span>, 
			 <span class="keywords">bxd</span>, <span class="keywords">bxh</span>, <span class="keywords">cxb</span></td>
		 </tr>
		 <tr bgcolor="#eeeeee"> 
			<td>&nbsp;</td>
			<td>&nbsp;</td>
			<td class="keywords">chr <a href="CGIDoc.html#Note3" class="keywords"><sup>(3)</sup></a></td>
			<td><span class="keywords">1</span>, <span class="keywords">2</span>, 
			 <span class="keywords">3</span>, ..., <span class="keywords">19</span>, <span class="keywords">X</span></td>
		 </tr>
		 <tr bgcolor="#eeeeee"> 
			<td>&nbsp;</td>
			<td>&nbsp;</td>
			<td class="keywords">format <a href="CGIDoc.html#Note4" class="keywords"><sup>(4)</sup></a></td>
			<td><span class="keywords">row</span>, <span class="keywords">column</span></td>
		 </tr>
		 <tr bgcolor="#eeeeee"> 

<td colspan="4" >Example: <a target="_blank" href="http://robot.genenetwork.org/webqtl/main.py?cmd=genotype&riset=BXD&chr=1&format=column" class="keywords">http://robot.genenetwork.org/webqtl/main.py?cmd=genotype&riset=BXD&chr=1&format=column</a></td>
		 </tr>
<!--RETRIEVE TRAIT DATA-->
<tr bgcolor="#eeeeee"> 
			<td colspan="4">&nbsp;</td>
		 </tr>
		 <tr bgcolor="#eeeeee"> 
			<td><a name="trait">Retrieve trait data</a></td>
			<td class="keywords">cmd=get<br>
			 cmd=trait<br>
			 cmd=tra</td>
			<td class="keywords"><strong>probeset</strong> <a href="CGIDoc.html#Note1" class="keywords"><sup>(1)</sup></a></td>
			<td>[Probe or probeset identifier, trait record number 
			 (for db=3), or marker name (for db=4)]</td>
		 </tr>
		 <tr bgcolor="#eeeeee"> 
			<td>&nbsp;</td>
			<td>&nbsp;</td>
			<td class="keywords"><strong>db</strong></td>
			
<td>
Each database stored in GN is assigned a short <A HREF="http://robot.genenetwork.org/webqtl/main.py?cmd=help" class="keywords" target="_blank">code</A> required to identify the particular database from which you would like to retrieve data.
</td>
		
		 </tr>
		 <tr bgcolor="#eeeeee"> 
			<td>&nbsp;</td>
			<td>&nbsp;</td>
			<td class="keywords"><strong>probe</strong> <a href="CGIDoc.html#Note1" class="keywords"><sup>(1)</sup></a></td>
			<td><span class="keywords">all pm mm</span> <br>OR <br>
			 [Affymetrix cell location code] <a href="CGIDoc.html#Note2" class="keywords"><sup>(3)</sup></a></td>
		 </tr>
		 <tr bgcolor="#eeeeee"> 
			<td>&nbsp;</td>
			<td>&nbsp;</td>
			<td class="keywords">format <a href="CGIDoc.html#Note4" class="keywords"><sup>(4)</sup></a></td>
			<td><span class="keywords">row</span>, <span class="keywords">column</span></td>
		 </tr>
		 <tr bgcolor="#eeeeee"> 
			<td colspan="4" >Example: <A target="_blank" Href="http://robot.genenetwork.org/webqtl/main.py?cmd=get&probeset=98332_at&db=bra08-03MAS5&probe=119637&format=col" class="keywords">http://robot.genenetwork.org/webqtl/main.py?cmd=get&probeset=98332_at&db=bra08-03MAS5&probe=119637&format=col</a></td>
		 </tr>

<!--SINGLE-MARKER REGRESSION-->

<tr bgcolor="#eeeeee"> 
			<td colspan="4">&nbsp;</td>
		 </tr>
		 <tr bgcolor="#eeeeee"> 
			<td><a name="map">Single-marker regression</a></td>
			<td class="keywords">cmd=map</td>
			<td class="keywords"><strong>probeset</strong> <a href="CGIDoc.html#Note1" class="keywords"><sup>(1)</sup></a></td>
			<td>[Probe or probeset identifier, trait record number 
			 (for db=3) or marker name (for db=4)]</td>
		 </tr>
		 <tr bgcolor="#eeeeee"> 
			<td>&nbsp;</td>
			<td>&nbsp;</td>
			<td class="keywords"><strong>db</strong></td>
			
<td>
Each database stored in GN is assigned a short <A HREF="http://robot.genenetwork.org/webqtl/main.py?cmd=help" class="keywords" target="_blank">code</A> required to identify the particular database from which you would like to retrieve data.
</td>

		</tr>
		 <tr bgcolor="#eeeeee"> 
			<td>&nbsp;</td>
			<td>&nbsp;</td>
			<td class="keywords"><strong>probe</strong> <a href="CGIDoc.html#Note1" class="keywords"><sup>(1)</sup></a></td>
			<td>all<br>
			 [Affymetrix cell location code] <a href="CGIDoc.html#Note2" class="keywords"><sup>(2)</sup></a></td>
		 </tr>
		 <tr bgcolor="#eeeeee"> 
			<td>&nbsp;</td>
			<td>&nbsp;</td>
			<td class="keywords">return</td>
			<td>number of LRS values to be reported; all LRS are reported if omitted</td>
		 </tr>
		 <tr bgcolor="#eeeeee"> 
			<td>&nbsp;</td>
			<td>&nbsp;</td>
			<td class="keywords">sort</td>
			<td><span class="keywords">lrs</span> [sorts by descending pvalue]<br>
				[sorts by ascending position by default]</td>
		 </tr>
		 <tr bgcolor="#eeeeee"> 

<td colspan="4" >Example: <a target="_blank" href="http://robot.genenetwork.org/webqtl/main.py?cmd=map&probeset=98332_at&db=bra08-03MAS5&sort=lrs&return=20" class="keywords">http://robot.genenetwork.org/webqtl/main.py?cmd=map&probeset=98332_at&db=bra08-03MAS5&sort=lrs&return=20</a></td>
		 </tr>

<!--QTL INTERVAL MAPPING-->

 <tr bgcolor="#eeeeee"> 
			<td colspan="4">&nbsp;</td>
		 </tr>
		 <tr bgcolor="#eeeeee"> 
			<td><a name="interval">QTL interval mapping</a></td>
			<td class="keywords">cmd=interval<br>
			 cmd=int</td>
			<td class="keywords"><strong>probeset</strong> <a href="CGIDoc.html#Note1" class="keywords"><sup>(1)</sup></a></td>
			<td>[Probe or probeset identifier, trait record number 
			 (for db=3), or marker name (for db=4)]</td>
		 </tr>
		 <tr bgcolor="#eeeeee"> 
			<td>&nbsp;</td>
			<td>&nbsp;</td>
			<td class="keywords"><strong>db</strong></td>
			
<td>
Each database stored in GN is assigned a short <A HREF="http://robot.genenetwork.org/webqtl/main.py?cmd=help" class="keywords" target="_blank">code</A> required to identify the particular database from which you would like to retrieve data.
</td>
		
		</tr>
		 <tr bgcolor="#eeeeee"> 
			<td>&nbsp;</td>
			<td>&nbsp;</td>
			<td class="keywords"><strong>probe</strong> <a href="CGIDoc.html#Note1" class="keywords"><sup>(1)</sup></a></td>
			<td>[Affymetrix cell location code] <a href="CGIDoc.html#Note2" class="keywords"><sup>(2)</sup></a></td>
		 </tr>
		 <tr bgcolor="#eeeeee"> 
			<td>&nbsp;</td>
			<td>&nbsp;</td>
			<td class="keywords">chr</td>
			<td><span class="keywords">1</span>, <span class="keywords">2</span>, 
			 <span class="keywords">3</span>, ..., <span class="keywords">19</span>, <span class="keywords">X</span></td>
		 </tr>
		 <tr bgcolor="#eeeeee"> 
			<td>&nbsp;</td>
			<td>&nbsp;</td>
			<td class="keywords">sort</td>
			<td><span class="keywords">lrs</span> [sorts by descending pvalue]<br>
				[sorts by ascending position by default]</td>
		 </tr>
		 <tr bgcolor="#eeeeee"> 
			<td colspan="4" >Example: <a target="_blank" href="http://robot.genenetwork.org/webqtl/main.py?cmd=interval&probeset=98332_at&db=bra08-03MAS5&chr=1" class="keywords">http://robot.genenetwork.org/webqtl/main.py?cmd=interval&probeset=98332_at&db=bra08-03MAS5&chr=1</a></td>
		 </tr>
		 <tr bgcolor="#eeeeee"> 
			<td colspan="4">&nbsp;</td>
		 </tr>

<!--TRAIT CORRELATION SEARCH-->

 <tr bgcolor="#eeeeee"> 
			<td><a name="correlation">Trait correlation search</a></td>
			<td><span class="keywords">cmd=correlation<br>cmd=cor<br>cmd=pearson<br>cmd=pea<br>cmd=spearman<br>cmd=spe</span></td>
			<td class="keywords"><strong>probeset</strong> <a href="CGIDoc.html#Note1" class="keywords"><sup>(1)</sup></a></td>
			<td>[Affymetrix probeset identifier, trait record number 
			 (for db=3) or marker name (for db=4)]</td>
		 </tr>
		 <tr bgcolor="#eeeeee"> 
			<td>&nbsp;</td>
			<td>&nbsp;</td>
			<td class="keywords"><strong>db</strong></td>
			
<td>
Each database stored in GN is assigned a short <A HREF="http://robot.genenetwork.org/webqtl/main.py?cmd=help" class="keywords" target="_blank">code</A> required to identify the particular database from which you would like to retrieve data.
</td>
		
		</tr>
		 <tr bgcolor="#eeeeee"> 
			<td>&nbsp;</td>
			<td>&nbsp;</td>
			<td class="keywords"><strong>searchdb</strong></td>

<td>
Each database stored in GN is assigned a short <A HREF="http://robot.genenetwork.org/webqtl/main.py?cmd=help" class="keywords" target="_blank">code</A> required to identify the particular database from which you would like to retrieve data.
</td>
		

 </tr>
		 <tr bgcolor="#eeeeee"> 
			<td>&nbsp;</td>
			<td>&nbsp;</td>
			<td class="keywords"><strong>probe</strong> <a href="CGIDoc.html#Note1" class="keywords"><sup>(1)</sup></a></td>
			<td>[Affymetrix cell location code] <a href="CGIDoc.html#Note2" class="keywords"><sup>(2)</sup></a></td>
		 </tr>
		 <tr bgcolor="#eeeeee"> 
			<td>&nbsp;</td>
			<td>&nbsp;</td>
			<td class="keywords">return</td>
			<td>[Maximum number of correlations to be reported]</td>
		 </tr>
		 <tr bgcolor="#eeeeee"> 
			<td>&nbsp;</td>
			<td>&nbsp;</td>
			<td class="keywords">sort</td>
			<td> <p><span class="keywords">pvalue</span> [sorts by ascending 
				pvalue]<br>
				[sorts by descending correlation by default]</p></td>
		 </tr>
		 <tr bgcolor="#eeeeee"> 
			<td>&nbsp;</td>
			<td>&nbsp;</td>
			<td class="keywords">id</td>
			<td><span class="keywords">yes</span> [Includes NCBI Entrez Gene ID for each transcript in the output (for expression data only).]<br><span class="keywords">only</span> [Returns only NCBI Entrez Gene ID for each transcript (for expression data only).]</td>
		 </tr>
		 <tr bgcolor="#eeeeee"> 
			<td colspan="4" >Example: <a target="_blank" href="http://robot.genenetwork.org/webqtl/main.py?cmd=cor&probeset=100001_at&probe=136415&db=bra08-03MAS5&searchdb=BXDPublish&return=500&sort=pvalue" class="keywords">http://robot.genenetwork.org/webqtl/main.py?cmd=cor&probeset=100001_at&probe=136415&db=bra08-03MAS5&searchdb=BXDPublish&return=500&sort=pvalue</a></td>
		 </tr>

<!--OPEN TRAIT DATA AND ANALYSIS PAGE-->
	 
		 <tr bgcolor="#eeeeee"> 
			<td colspan="4">&nbsp;</td>
		 </tr>
		 <tr bgcolor="#eeeeee"> 
			<td><a name="interval">Open Trait Data and Analysis Form</a></td>
			<td class="keywords">cmd=show<br>
			 cmd=shw</td>
			<td class="keywords"><strong>probeset</strong> <a href="CGIDoc.html#Note1" class="keywords"><sup>(1)</sup></a></td>
			<td>[Affymetrix probeset identifier, trait record number 
			 (for db=3) or marker name (for db=4)]</td>
		 </tr>
		 <tr bgcolor="#eeeeee"> 
			<td>&nbsp;</td>
			<td>&nbsp;</td>
			<td class="keywords"><strong>db</strong></td>
<td>
Each database stored in GN is assigned a short <A HREF="http://www.genenetwork.org/webqtl/main.py?cmd=help" class="keywords" target="_blank">code</A> required to identify the particular database from which you would like to retrieve data.
</td>
		
</tr>
		 
<tr bgcolor="#eeeeee"> 
			<td>&nbsp;</td>
			<td>&nbsp;</td>
			<td class="keywords"><strong>probe</strong> <a href="CGIDoc.html#Note1" class="keywords"><sup>(1)</sup></a></td>
			<td>[Affymetrix cell location code] <a href="CGIDoc.html#Note2" class="keywords"><sup>(2)</sup></a></td>
		 </tr>
		 <tr bgcolor="#eeeeee"> 
			<td colspan="4" >Example: <a target="_blank" href="http://www.genenetwork.org/webqtl/main.py?cmd=show&db=HC_M2_0606_M&probeset=1415670_at&probe=269753" class="keywords">http://www.genenetwork.org/webqtl/main.py?cmd=show&db=HC_M2_0606_M&probeset=1415670_at&probe=269753</a></td>
		 </tr>
		 <tr bgcolor="#eeeeee"> 
			<td colspan="8">&nbsp;</td>
		 </tr>

<!--OPEN BEST TRAIT DATA STATISTICS PAGE-->

<tr bgcolor="#eeeeee"> <td colspan="4">The script below is used to filter a list of probes or probe sets that target the same gene or transcript to find the single probe or probe set with highest expression and to then display the Statistics Page for that particular data. This script does not use the cgi-bin format.</td></tr>

<tr bgcolor="#eeeeee"> 
			<td><a name="showBest">Show Best Trait Data Statistics Page</a></td>
			
<td class="keywords">parameters<br>
			 &nbsp;</td>

<td class="keywords">examples<br>
			 &nbsp;</td>

<td>Notes</td>
		
</tr>

<tr bgcolor="#eeeeee"> 
			<td>&nbsp;</td>

<td class="keywords"><strong>gene= <BR>refseq= <BR>geneid= <BR>database= <BR>searchAlias=</strong></td>
			
<td class="keywords">gene=Rho <BR>refseq=NM_145383 <BR>geneid=212541 <BR>database=Eye_M2_0906_R <BR>searchAlias=1</td>
			
<td>There are three supported search term types: 1. Gene symbol, 2. RefSeq identifier, 3. Entrez gene identifier. Although not recommended, the string "&searchAlias=1" can be added at the end of the command to retrieve data using the alias of a proper gene symbol when the original does not work. Thus "gene=RP4" will resolve to "gene=Rho" only if you add "&searchAlias=1" at the end of the command. 
<BR>Each database stored in GN is assigned a short <A HREF="http://www.genenetwork.org/webqtl/main.py?cmd=help" class="keywords" target="_blank">code</A> required to identify the particular database from which you would like to retrieve data.
</td>
		</tr>
<tr bgcolor="#eeeeee"> 
<td colspan="4" >Example 1: <a target="_blank" href="http://www.genenetwork.org/webqtl/main.py?FormID=showBest&gene=Rho&database=HC_M2_0606_P" class="keywords">http://www.genenetwork.org/webqtl/main.py?FormID=showBest&gene=Rho&database=HC_M2_0606_P</a></td>
		 </tr>

<tr bgcolor="#eeeeee"> 
			<td colspan="4" >Example 2: <a target="_blank" href="http://www.genenetwork.org/webqtl/main.py?FormID=showBest&refseq=NM_145383&database=HC_M2_0606_P" class="keywords">http://www.genenetwork.org/webqtl/main.py?FormID=showBest&refseq=NM_145383&database=HC_M2_0606_P</a></td>
		 </tr>

<tr bgcolor="#eeeeee"> 
			<td colspan="4" >Example 3: <a target="_blank" href="http://www.genenetwork.org/webqtl/main.py?FormID=showBest&geneid=212541&database=HC_M2_0606_P" class="keywords">http://www.genenetwork.org/webqtl/main.py?FormID=showBest&geneid=212541&database=HC_M2_0606_P</a></td>
		 </tr>

<tr bgcolor="#eeeeee"> 
			<td colspan="4" >Example 4: <a target="_blank" href="http://www.genenetwork.org/webqtl/main.py?FormID=showBest&gene=RP4&database=HC_M2_0606_P&searchAlias=1" class="keywords">http://www.genenetwork.org/webqtl/main.py?FormID=showBest&gene=RP4&database=HC_M2_0606_P&searchAlias=1</a></td>
		 </tr>

<tr bgcolor="#eeeeee"> <td colspan="4">This script was originally written to allow connections from the <A HREF="http://neibank.nei.nih.gov/" "target="_empty" class="fs14" >NEIBank</A> to GeneNetwork by H. Li and R. Williams (see <A HREF="http://www.genenetwork.org/whats_new.html" "target="_empty" class="fs14" >News of April 25, 2007</A>). </td></tr>

<!--SEARCH FOR GENE DATA-->
 
		 <tr bgcolor="#eeeeee"> 
			<td colspan="4">&nbsp;</td>
		 </tr>
		 <tr bgcolor="#eeeeee"> 
			<td><a name="interval">Search for gene data</a></td>
			<td class="keywords">cmd=search<br>
			 cmd=sch</td>
			<td class="keywords">
				<strong>gene</strong> <BR>
				<strong>refseq</strong> <BR>
				<strong>genbankid</strong> <BR>
			
			</td>
			<td>The name or ID of the gene or protein to be searched, choose one and only one type of identifier</td>
		 </tr>
		 <tr bgcolor="#eeeeee"> 
			<td>&nbsp;</td>
			<td>&nbsp;</td>
			<td class="keywords">alias </td>
			<td>[0 or 1] Whether to search for possible aliases (only when a gene name is supplied), optional, disabled by default (0)</td>
		 </tr>
		 <tr bgcolor="#eeeeee"> 
			<td>&nbsp;</td>
			<td>&nbsp;</td>
			<td class="keywords">species</td>
			<td>Optional, only human, mouse and rat are supported now. Default is mouse if not specified.</td>
		 </tr>
		 <tr bgcolor="#eeeeee"> 
			<td>&nbsp;</td>
			<td>&nbsp;</td>
			<td class="keywords">tissue</td>
			<td>Optional, if not specified, display traits from all available tissues. For tissue abbreviations, please check <A target="_blank" Href="http://www.genenetwork.org/webqtl/main.py?cmd=help&topic=tissue" class="keywords">http://www.genenetwork.org/webqtl/main.py?cmd=help&topic=tissue</A></td>
		 </tr>
		 <tr bgcolor="#eeeeee"> 
			<td>&nbsp;</td>
			<td>&nbsp;</td>
			<td class="keywords">format</td>
			<td>[html or text], Optional, if not specified, return result in HTML format</td>
		 </tr>
		 <tr bgcolor="#eeeeee"> 
			<td colspan="4" >
			
Example 1 : <a target="_blank" href="http://www.genenetwork.org/webqtl/main.py?cmd=sch&gene=ddx25" class="keywords">
				http://www.genenetwork.org/webqtl/main.py?cmd=sch&gene=ddx25</a><BR>
			Example 2 : <a target="_blank" href="http://www.genenetwork.org/webqtl/main.py?cmd=sch&refseq=NM_133594&species=rat" class="keywords">
				http://www.genenetwork.org/webqtl/main.py?cmd=sch&refseq=NM_133594&species=rat</a><BR>
			Example 3 : <a target="_blank" href="http://www.genenetwork.org/webqtl/main.py?cmd=sch&genbankid=AW047046" class="keywords">
				http://www.genenetwork.org/webqtl/main.py?cmd=sch&genbankid=AW047046</a><BR>
			Example 4 : <a target="_blank" href="http://www.genenetwork.org/webqtl/main.py?cmd=sch&gene=GRTH&alias=1&species=rat&tissue=fat" class="keywords">
				http://www.genenetwork.org/webqtl/main.py?cmd=sch&gene=GRTH&alias=1&species=rat&tissue=fat</a><BR>
			Example 5 : <a target="_blank" href="http://www.genenetwork.org/webqtl/main.py?cmd=sch&gene=Grin2b&tissue=hip&format=text" class="keywords">
				http://www.genenetwork.org/webqtl/main.py?cmd=sch&gene=Grin2b&tissue=hip&format=text</a>
			</td>
		 </tr>
		 <tr bgcolor="#eeeeee"> 
			<td colspan="4">&nbsp;</td>
		 </tr>
		 
		</table>
		</TD>
		</TR>
		</TABLE>

		</TD>
	</TR>
		
	<TR>
		<TD colspan=2 bgColor=#eeeeee>
		<P class="title">Notes</P>
<ol>
<li><a name="Note1"></a>A value is required for the <span class="keywords">probeset</span> but not for <span class="keywords">probe</span> keyword. With only <span class="keywords">probeset</span>, commands will retrieve the average for the probeset from different average method. With <span class="keywords">probeset</span> and <span class="keywords">probe=all</span>, commands will retrieve data for all probes of that probeset; with <span class="keywords">probe</span>, data for a single probe. If the specified <span class="keywords">probeset</span> and <span class="keywords">probe</span> are inconsistent, no data is returned.</li>
		  
<li><a name="Note2"></a>The Affymetrix cell location code is six characters. The first three characters are the column or x location, and the last three the row or y location; for example, <span class="keywords">237198</span>. If the column has less than three digits, it should be preceded by one or two <span class="keywords">X</span> characters to make three characters. Similarly, the row should be prefixed with one or two <span class="keywords">Y</span> characters, if necessary; for example, <span class="keywords">X24YY9</span> or <span class="keywords">XX8Y34</span>.</li>
		  
<li><a name="Note3"></a>If <span class="keywords">chr</span> is omitted, command will retrieve information for all chromosomes.</li>
		  
<li><a name="Note4"></a>If <span class="keywords">format</span> is omitted, row formatting is the default.</li>
		
</ol>
		</TD>
		
		</TR></TABLE>
		</TD>
	</TR>
	<TR>
		<TD align=center bgColor=#ddddff class="solidBorder">
		<!--Start of footer-->
		<TABLE width="90%">
		<script language='JavaScript' src='/javascript/footer.js'></script>
		</TABLE>
		<!--End of footer-->

		</TD>
	</TR>
</TABLE>
<!-- /Footer -->
<script language="JavaScript" src="javascript/menu_new.js"></script>
<script language="JavaScript" src="javascript/menu_items.js"></script>
<script language="JavaScript" src="javascript/menu_tpl.js"></script>
<script language="JavaScript">
	<!--//
	new menu (MENU_ITEMS, MENU_POS);
	//-->
</script>
<script src="http://www.google-analytics.com/urchin.js" type="text/javascript">
</script>
<script type="text/javascript">
_uacct = "UA-3782271-1";
urchinTracker();
</script>
</BODY>
</HTML>
