<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.0 Transitional//EN">
<HTML><HEAD><TITLE>Status and investigators to contact</TITLE>
<META http-equiv=Content-Type content="text/html; charset=iso-8859-1">
<LINK REL="stylesheet" TYPE="text/css" HREF='css/general.css'>
<LINK REL="stylesheet" TYPE="text/css" HREF='css/menu.css'>
<SCRIPT SRC="javascript/webqtl.js"></SCRIPT>

</HEAD>
<BODY  bottommargin="2" leftmargin="2" rightmargin="2" topmargin="2" text=#000000 bgColor=#ffffff>
<TABLE cellSpacing=5 cellPadding=4 width="100%" border=0>
	<TBODY>
	<TR>
		<script language="JavaScript" src="/javascript/header.js"></script>
	</TR>
	<TR>
		<TD bgColor=#eeeeee class="solidBorder">
		<Table width= "100%" cellSpacing=0 cellPadding=5><TR>
	<!-- Body Start from Here -->
		<TD vAlign=top width="100%" colspan=2 bgColor=#eeeeee>
			<P class="title">Status and Investigators to Contact on Data Use and Publication <A HREF="/webqtl/main.py?FormID=editHtml"><img src="images/modify.gif" alt="modify this page" border= 0 valign="middle"></A></P>


MOUSE BXD: The University of Western Australia data sets (Thymus, Spleen, Peripheral Blood Leucocytes)

Status: Currently these are unpublished and private data source with usage restrictions.

Please contact Dr. Grant Morahan regarding access and use of these data.

<Blockquote>
Data that are accessible via GeneNetwork belong to several research groups listed below. Some of the data sets are still actively being generated and analyzed. The scientists who are generating these data have often agreed to remove password protection and let the research community view, share, and analyze data. Although they are willing and enthusiastic about sharing these data, they have not relinquished interest or ownership. If you are planning to use results and data extracted from GeneNetwork in publication, we request that you contact the data owners prior to submission.
</Blockquote><P></P>

<Blockquote>
<A HREF="dbdoc/BR_U_1203_H2.html" target="_blank">MOUSE BXD: Thymus, Spleen, Peripheral Blood Leucocytes</A> data sets:
<P><B>Status</B>: These are unpublished and private data source with usage restrictions. Error-checking and refinement of this data set is still in progress. Data were first entered November 2008. 

<P><B>References and Contact</B>:  For access to data prior to publication, please contact Grant Morahan (gem at waimr. uwa. edu. au) regarding use of these data sets on a collaborative basis.

<P>These expression data sets are being generated by investigators at The Western Australian Institute for Medical Research and The University of Western Australia (Grant Morahan, Munish Mehta, Quang Nguyen, James Jooste, and Violet Peeva). Samples are generated by Quang Nguyen and James Jooste. Arrays are all processed by Quang Nguyen. 
</Blockquote><P></P>


<Blockquote>
<A HREF="dbdoc/BR_U_1203_H2.html" target="_blank">MOUSE BXD: UTHSC Brain U74Av2</A> data sets:
<P><B>Status</B>: DEPRECATED, PUBLISHED, and OPEN data source, NO usage restrictions as of March 1, 2005. Even the most recent release of these data is now obsolete. We recommend using more recent Brain data sets from the University of Colorado Denver (Tabakoff and colleagues) or from INIA. 

<P><B>References</B>: Chesler EJ, Lu L, Shou S, Qu Y, Gu J, Wang J, Hsu HC, Mountz JD, Baldwin N, Langston MA, Threadgill DW, Manly KF, Williams RW (2005) Genetic dissection of gene expression reveals polygenic networks modulating brain structure and function. Nature Genetics 37:233-242

<P>The forebrain and midbrain expression data were generated using The William and Dorothy Dunavant Endowment to RWW and published in early 2005 (Chesler et al., 2005). Arrays were processed at <a href="http://www.genomeexplorations.com" target="_blank" class="fs14">Genome Explorations</A> (Divyen Patel) or by Dr. Tom Sutter and colleagues at the University of Memphis. Data are now available on the GeneNetwork simply by clicking on the Information Page associated with the many different versions (transforms). Please contact <A class="fs14" HREF="mailto:rwilliam@nb.utmem.edu">Rob Williams </A> if you have any questions on the use of these open data. We now consider these data to be somewhat archaic and of relatively poor quality compared to recent data sets such as the INIA BXD brain data and the UCHSC BXD brain data.
</Blockquote><P></P>

<Blockquote>
<A HREF="dbdoc/HC_U_0304_R.html" target="_blank">MOUSE BXD: GNF-Groningen Stem Cell U74Av2</A> data sets:
<P><B>Status</B>: PUBLISHED and OPEN data source, NO usage restrictions as of March 1, 2005. Error-checked and complete as of March 1, 2005. There are no plans to modify or expand this data set.

<P><B>References</B>: Bystrykh L, Weersing E, Dontje B, Sutton S, Pletcher MT, Wiltshire T, Su AI, Vellenga E, Wang J, Manly KF, Lu L, Chesler EJ, Alberts R, Jansen RC, Williams RW, Cooke M, de Haan G (2005) Uncovering regulatory pathways affecting hematopoietic stem cell function using “genetical genomics” Nature Genetics 37:225-232

<P>These hematopoietic stem cell expression data were generated by collaborators at The University of Groningen (Gerald de Haan) and The Genomics Institute of the Novartis Research Foundation (Mike Cooke). Flow-sorted cells were generated in Holland. RNA samples and arrays were processed at GNF in La Jolla, CA. This team incorporated all of their data into the GeneNetwork before publication. The GNF-Groningen stem cell data are now available on the NCBI GEO site using the accession identifier GSE2031. If you have questions on the use of these data, please contact <A class="fs14" HREF="mailto:g.de.haan@med.rug.nl">Gerald de Haan </A> and <A class="fs14" HREF="mailto:mcooke@gnf.org">Michael Cooke </A>.
</Blockquote><P></P>


<Blockquote>
<A HREF="dbdoc/CB_M_0305_P.html" target="_blank">MOUSE BXD: SJUT M430 Cerebellum</A> data sets: 
<P><B>Status</B>: Unpublished and open data source with usage restrictions only on large-scale and global analysis. This data set is now complete as of 2005.
 
<P>These cerebellar expression data are being generated by a consortium of investigators at St. Jude Children's Research Hospital (Clayton Naeve, Tom Curran, Peter McKinnon, Jim Morgan, Rich Smeyne) and at UTHSC (Dan Goldowitz, Lu Lu, Kristin Hamre, and Rob Williams). Samples are generated at UTHSC by Lu Lu and colleagues. Arrays are all processed at SJCRH by Clayton Naeve and colleagues. Please contact <A class="fs14" HREF="mailto:rwilliam@nb.utmem.edu">Rob Williams</A>, or <A class="fs14" HREF="mailto:dgold@nb.utmem.edu">Dan Goldowitz </A> regarding use of these data sets in publications or projects.
</Blockquote><P></P>


<Blockquote>
<A HREF="dbdoc/Treg_R_1006.html" target="_blank">MOUSE BXD: Helmholtz Centre for Infection Research</A> data sets: 
<P><B>Status</B>: Currently these are unpublished and private data source with usage restrictions. 
<P>Please contact<A class="fs14" HREF="mailto:Klaus.Schughart@helmholtz-hzi.de"> Dr. Klaus Schughart</A> regarding access and use of these data.
<P>November 17, 2006.
</Blockquote><P></P>
	

<Blockquote>
4. <A HREF="dbdoc/IBR_M_0405_P.html" target="_blank">MOUSE BXD: INIA Brain mRNA M430</A> data sets:

<P><B>Status</B>: PUBLISHED and OPEN data source, with no usage restrictions. Error-checked and complete as of March 1, 2005. There are no plans to modify or expand this data set. Error-checked and nearly complete as of Sept 1, 2005. 

<P class="subtitle">&nbsp;&nbsp;&nbsp;&nbsp;INIA data access:</P>
<Blockquote>

<P>Normalized data are available for this INIA data set at</P>
<Blockquote>
<OR>

<LI>Jan 2006, PDNN normalization (17 Mb file with strain means): <A HREF="ftp://atlas.utmem.edu/Public/Mouse_bxd/INIA_M_0106_PDNN.txt">ftp://atlas.utmem.edu/Public/Mouse_bxd/INIA_M_0106_PDNN.txt </A>

<LI>Jan 2006, RMA normalization (17 Mb file with strain means): <A HREF="ftp://atlas.utmem.edu/Public/Mouse_bxd/INIA_M_0106_RMA.txt">ftp://atlas.utmem.edu/Public/Mouse_bxd/INIA_M_0106_RMA.txt </A>

<LI>June 2006, QTL results from RMA normalized data (5.7 Mb, no strain means): <A HREF="ftp://atlas.utmem.edu/Public/Mouse_bxd/INIA_M_0606_RMA.txt">ftp://atlas.utmem.edu/Public/Mouse_bxd/INIA_M_0606_RMA.txt </A>

<LI>All data in ZIP format: <A HREF="ftp://atlas.utmem.edu/Public/Mouse_bxd/INIA_mRNA_data_sets.zip">ftp://atlas.utmem.edu/Public/Mouse_bxd/INIA_mRNA_data_sets.zip </A>
</OR>

</Blockquote>
</Blockquote>



<P><B>References</B>: Peirce JL, Lu L, Li H, Wang J, Manly KF, Hitzemann RJ, Belknap JK, Rosen GD, Goodwin S, Sutter TR, Williams RW (2006) How replicable are mRNA expression QTLs. Mammalian Genome 17:643-642

<P>These forebrain and midbrain expression sets were generated with continued support from NIAAA-INIA. Arrays were processed at the University of Memphis by Thomas Sutter and Shirlean Goodwin. These data are openly avaiable at all levels (CEL files, etc). Please contact <A class="fs14" HREF="mailto:rwilliam@nb.utmem.edu">Robert W. Williams </A> for access to orginal data.
</Blockquote><P></P>


<Blockquote>
5. <A HREF="dbdoc/SA_M2_0405_PC.html" target="_blank">MOUSE BXD: HBP/Rosen Striatum</A> data sets:

<P><B>Status</B>: Unpublished and open data source with usage restrictions only on large-scale and global analysis. Error-checked but still incomplete. We plan to significantly expand the size of this data set in 2005-2006.

<P>The striatal expresssion data sets are being generated by Glenn D. Rosen, Robert W. Williams, and colleagues with continued support from an NIH Human Brain Project award. Tissue and arrays are processed at <a href="https://www.bidmcgenomics.org/" target="_blank" class="fs14">Beth Israel Deaconess Medical Center</A> by Glenn Rosen and colleagues. Please contact <A class="fs14" HREF="mailto:grosen@bidmc.harvard.edu">Glenn Rosen</A> regarding extensive use of this data set in publications or projects.
</Blockquote><P></P>


<Blockquote>
<A HREF="dbdoc/BR_U_1203_H2.html" target="_blank">MOUSE LXS Illumina Hippocampus </A> data sets:

<P><B>Status</B>: Unpublished and open data source with usage restrictions only on large-scale and global analysis. Error-checked and complete.

<P><B>References</B>: Lu Lu et al. RSA July and CTC May 2007

<P>This data set was initially entered in GeneNetwork, October 2006. Please contact <A class="fs14" HREF="mailto:lulu@nb.utmem.edu">Dr. Lu Lu</A> if you have any questions on the use of these open data. </Blockquote><P></P>


<Blockquote>
<A HREF="dbdoc/MA_M_0704_R.html" target="_blank">MOUSE AKXD: NCI Mammary tumor mRNA M43 </A> data sets:

<P><B>Status</B>: Unpublished and open data source with usage restrictions only on large-scale and global analysis. Error-checked and complete.

<P>The mammary tumor expresssion data sets have been generated by Kent Hunter and colleagues with support from NCI Laboratory of Population Genetics. Tissue and arrays were processed at NCI. Please contact <A HREF="<A HREF="mailto:hunterk@mail.nih.gov" class="fs14">Kent Hunter</A> regarding extensive use of this data set in publications or projects.
</Blockquote><P></P>


<Blockquote>
<A HREF="dbdoc/BDF2_M_0805_P.html" target="_blank">B6D2F2</A> database:

<P><B>Status</B>: Unpublished but submitted; an open data source with usage restrictions only on large-scale and global analysis. Error-checked and complete as of Sept 1, 2005.

<P><B>References</B>: 105.	Peirce JL, Lu L, Li H, Wang J, Manly KF, Hitzemann RJ, Belknap JK, Rosen GD, Goodwin S, Sutter TR, Williams RW (2006) How replicable are mRNA expression QTLs. Mammalian Genome 17:643-642

<P>All of the OHSU/VA B6D2F2 Brain mRNA M430AB data sets have been generated by Robert Hitzemann and John Belknap at The Oregon Health Sciences University in Portland. For contact and citations and other information on these data sets please review the INFO pages and contact Drs. <A  class="fs14" HREF="mailto:belknajo@ohsu.edu"> Belknap </A> or <A class="fs14"  HREF="mailto:hitzeman@ohsu.edu"> Hitzemann </A> regarding use of this data set in publications or projects. 

</Blockquote><P></P>



<Blockquote>
<A HREF="dbdoc/LVF2_M_0704_R.html" target="_empty">B6BTBRF2-ob Mouse Liver and Metabolic Trait</A> data sets:

<P><B>Status</B>: Published open data source with no usage restrictions. Error-checked and complete. 
The F2 data set used in the manuscript is available at <A HREF="http://www.ncbi.nlm.nih.gov/geo/query/acc.cgi">GEO</A> under the accession number "GSE3330".


<P><B>References</B>: Lan H, Chen M, Byers JE, Yandell BS, Stapleton DS, Mata CM, Mui TK, Flowers MT, Schueler KL, Manly KF, Williams RW, Kendziorski, CM, Attie AD (2006) Combined expression trait correlations and expression quantitative trait locus mapping. PLoS Genetics 2:51-61

<P>The liver expression data were generated by Dr. Alan Attie and colleagues at the University of Wisconsin. All original Affymetrix data files were generated by Attie and colleagues. Data sets in The GeneNetwork were processed as described in the Information pages. Data sets were opened to the public by Dr. Alan Attie and colleagues, August 15, 2005. Please contact Alan Attie <attie at biochem.wisc.edu> for help in the use of these data.
</Blockquote><P></P>



<Blockquote>
<A HREF="dbdoc/LV_G_0704_R.html" target="_blank">UNC Agilent Liver</A> data sets:

<P><B>Status</B>: PUBLISHED and OPEN data source, NO usage restrictions as of February 1, 2007. Error-checked and complete as of March 1, 2005. There are no plans to modify or expand this data set.

<P><B>References</B>: Gatti D, Maki A, Chesler EJ, Kosyk O, Kirova R, Lu L, Manly KF, Qu Y, Williams RW, Perkins A, Langston ME, Threadgill DW, Rusyn I (2007) Genome-level analysis of genetic regulation of liver gene expression networks. Hepatology in press

<P>The BXD liver expression data sets were generated by Ivan Rusyn, David Threadgill, and colleagues with support from an NIEHS Toxicogenomics award. Arrays and the final data sets in The GeneNetwork were generated at the <a href="https://www.bidmcgenomics.org/" target="_blank" class="fs14">UNC Array Core</A>. Please contact <A class="fs14" HREF="mailto:iir@unc.edu">Ivan Rusyn</A> to obtain original data files or for help in the use of these data.
</Blockquote><P></P>


<Blockquote>
10. <A HREF="dbdoc/HC_M2_1005_P.html" target="_blank">Hippocampus Consortium</A> BXD and CXB data sets:

<P><B>Status</B>: Unpublished and open data source for BXD, CXB, and diverse inbred strains of mice with a usage restrictions on large-scale and global analysis. Error-checked and essentially complete.

<P>The hippocampus expresssion data sets were generated by a consortium of investigator with support from a large number of funding agencies. Tissue and arrays are processed at the University of Memphis by Thomas Sutter and Shirlean Goodwin.  Please contact <A class="fs14" HREF="mailto:rwilliam@nb.utmem.edu.edu">Robert W. Williams</A> regarding extensive use of this data set in publications or projects.
</Blockquote><P></P>



<Blockquote>
<A HREF="dbdoc/EYE_M2_1105_R.html" target="_blank">Hamilton Eye Institute Mouse Eye</A> data sets:

<P><B>Status</B>: Unpublished and open data source with usage restrictions only on large-scale and global analysis. Error-checked and essentially complete by September 2006.

<P>This eye expresssion data set were generated by Robert W. Williams, E. E. Geisert, L. Lu, and W. Gu with support solely from Barrett G. Haik. Tissue and arrays were processed at the VA Medical Center, Memphis, by Weikuan Gu and Yan Jiao.  Please contact <A class="fs14" HREF="mailto:rwilliam@nb.utmem.edu.edu">Robert W. Williams</A> regarding extensive use of this data set in publications or projects.



<P><a href="dbdoc/barley_SCRI.html">Barley seedling leaf and embryo 22K GeneChip data sets from SCRI</a>:<br>
  <strong>Status:</strong> unpublished. The data sets will become OPEN with no usage restrictions as soon as publication is accepted (expected by mid 2007). The data sets will also be available from the ArrayExpress: accessions E-TABM-111 and E-TABM-112. Please contact <a href="mailto:adruka@scri.sari.ac.uk">Arnis Druka</a> if you are interested in using these data sets before they are released.
<P> The barley expression data sets were funded by the BBSRC grant SCR/910/04 to Michael Kearsey (University of Birmingham, UK) and Robbie Waugh (SCRI, UK). Tissue and RNA isolation was performed by Arnis Druka at SCRI. Arrays were processed by Roger Wise (Iowa State University). Genetic linkage map was assembled by Arnis Druka by integrating updated RFLP-based mapping data provided by Andris Kleinhofs (Washington State University, Pullman) with SNP genotyping data (based on pilot barley OPA1) provided by Timothy Close (University of California, Riverside).  



<P><A HREF="dbdoc/BXDGeno.html" target="_blank">Genotypes Files</A>:
</Blockquote>
<Blockquote>
  <P> All microsatellite and SNP marker genotype data files for mouse genetic reference populations (AXB/BXA, BXD, LXS, CXB, BXH) are public. Genotype files for the B6D2F2 and B6BTBRF2 are public. The rat HXB/BXH genotypes are public. BayxSha Arabidopsis genotype files are also public. Any of the genotype files is available upon request to R. W. Williams or original data providers. 

<P>The majority of mouse genotypes in use after May 2005 are SNPs that were genotyped by the Jonathan Flint, Richard Mott (Wellcome Trust, Oxford), and Robert Williams.  DNA samples from more than ~480 strains of mice were genotyped at 15,360 SNPs at Illumina in early 2005. The entire set is referred to as the Wellcome-CTC SNP data set. The appropriate URL citation for these data is currently <A HREF="http://www.well.ox.ac.uk/mouse/INBREDS/" target="_blank" class="fs14">www.well.ox.ac.uk/mouse/INBREDS/</A>.

<P>The specific mouse genotype files used by WebQTL incorporated both SNPs and microsatellite have been substantially modified and error-checked and will not correspond precisely to the original Wellcome-CTC SNP data set files. All of these new markers are public and can be used. Please contact <A class="fs14" HREF="mailto:rwilliam@nb.utmem.edu">Rob Williams</A> if you would like the specific files used in any of the mouse genetic reference populations.
			</Blockquote><P></P>

<Blockquote>
<A HREF="dbdoc/BXDPublish.html" target="_blank">Phenotypes</A> databases:

<P>Mouse phenotype databases were generated primarily by extracting trait values from the literature. All of the phenotype databases (BXD, AXB, CXB, BXH, LXS, LGXSM, HXB/BXH, and BayxSha) are curated by <A class="fs14" HREF="mailto:cheslerej@ornl.gov">Elissa Chesler</A> and Robert W. Williams. In several cases, these databases include extensive and still unpublished traits. Please contact <A class="fs14" HREF="mailto:cheslerej@ornl.gov">Elissa Chesler </A> or <A class="fs14" HREF="mailto:rwilliam@nb.utmem.edu">Rob Williams</A> regarding new phenotypes you would like entered into any of the databases or regarding appropriate use of the entire database.
</Blockquote><P></P>
			

<Blockquote>
<B>Genomic and Array Annotation</B> databases:

<P>The GeneNetwork relies on custom and public databases for Affymetrix, Illumina, and Agilent array platforms. In the case of the mouse Affymetrix U74Av2 and M430 arrays, and the rat RAE230A array, we have extensively annotated probes and probe set data. Our files are manually curated and will NOT correspond 1-to-1 with any other publically available annotation of these particular Affymetrix platforms. In the case of mouse, we have data on the positions of a large number of SNPs. These data were contributed by a number of colleagues and integrated by Rob Williams, Rob Crowell, and colleagues. Please contact  <A class="fs14" HREF="mailto:rwilliam@nb.utmem.edu">Rob Williams</A> if you would like access to parts of this data set. Detailed SNP data have now been placed on WebQTL maps and we thank Celera for providing early access to these data in 2003. 
</Blockquote><P></P>

<P class="subtitle">&nbsp;&nbsp;&nbsp;&nbsp;Information about this text file:</P> 
<Blockquote>
  <P>This text file originally generated by RWW, March 2004. Updated by RWW, Nov 12, 2004; Dec 4, 2004; EJC, Aug 29, 2005; RWW, Sept 4, 2005; Nov 5, 2005; Jan 26, 2007; AD, Jan 28, 2007.
</P>
</Blockquote><P></P>

			<P></P>
		</TD>
		</TR></TABLE>
		</TD>
	</TR>
	<TR>
		<TD align=center bgColor=#ddddff class="solidBorder">
		<!--Start of footer-->
		<TABLE width="90%">
		<script language='JavaScript' src='/javascript/footer.js'></script>
		</TABLE>
		<!--End of footer-->
		</TD>
	</TR>
</TABLE>
<!-- /Footer -->
<script language="JavaScript" src="/javascript/menu_new.js"></script>
<script language="JavaScript" src="/javascript/menu_items.js"></script>
<script language="JavaScript" src="/javascript/menu_tpl.js"></script>
<script language="JavaScript">
	<!--//
	new menu (MENU_ITEMS, MENU_POS);
	//-->
</script>
<script src="http://www.google-analytics.com/urchin.js" type="text/javascript">
</script>
<script type="text/javascript">
_uacct = "UA-3782271-1";
urchinTracker();
</script>
</BODY>
</HTML>
