<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.0 Transitional//EN">
<HTML><HEAD><TITLE>GeneNetwork News</TITLE>
<META http-equiv=Content-Type content="text/html; charset=iso-8859-1">
<LINK REL="stylesheet" TYPE="text/css" HREF='/css/general.css'>
<LINK REL="stylesheet" TYPE="text/css" HREF='/css/menu.css'>
<SCRIPT SRC="javascript/webqtl.js"></SCRIPT>
<SCRIPT SRC="javascript/dhtml.js"></SCRIPT>
</HEAD>
<BODY  bottommargin="2" leftmargin="2" rightmargin="2" topmargin="2" text=#000000 bgColor=#ffffff>
<TABLE cellSpacing=5 cellPadding=4 width="100%" border=0>
	<TBODY>
	<TR>
		<script language="JavaScript" src="/javascript/header.js"></script>
	</TR>
	<TR>
		<TD bgColor=#eeeeee class="solidBorder">
		<Table width= "100%" cellSpacing=0 cellPadding=5><TR>
		<!-- Body Start from Here -->
		<TD valign="top" height="200" width="100%" bgcolor="#eeeeee">

<P class="title">Latest Updates and News from Genenetwork  <A HREF="/webqtl/main.py?FormID=editHtml"><img src="/images/modify.gif" alt="modify this page" border= 0 valign="middle"></A></P>
<Blockquote>
<STRONG>2012-5-4: Second Developmental Studies of the Genetics of Gene Expression in Brain have been entered into GeneNetwork (BIDMC/UTHSC Dev Neocortex P3 and P14 ILMv6.2 (Nov11) RankInv). </STRONG> Dr. Glenn Rosen and colleagues have contributed data on gene expression across sets of 32 BXD strains for the neocortex and striatum at two stages of development (postnatal days 3 and 14). They used the Illumina Mouse Genome 6 version 2 array. These data are matched by previous data sets for the adult neocortex and striatum. These data are now publicly available but users are requested to contact Glenn D. Rosen regarding the status of these new data. (Implemented by G Rosen, RW Williams and A Centeno).
</Blockquote>

<Blockquote>

<STRONG>2012-1-20: Mouse SNPs from dbSNP have been added to GeneNetwork.</STRONG> 10 million mouse SNPs from dbSNP (build 128) have been added to Variant Browser. They could be searched by name (e.g. rs31192936) (Implemented by Xiaodong Zhou and Ning Liu).
</Blockquote>

<Blockquote>
<STRONG>2012-1-20: Literature correlation has been update to 2011 version.</STRONG> Dr. Ramin Homayouni and Dr. Lijing Xu kindly provide the 2011 version of mouse gene-gene literature correlation matrix to GeneNetwork. (Implemented by Xiaodong Zhou).
</Blockquote>

<Blockquote>
<STRONG>2011-12-16: Expression data set for EPFL/LISP BXD Muscle Affy Mouse Gene 1.0 ST (Dec11) RMA ** has been entered in GeneNetwork.</STRONG> Laboratory of Integrative and Systems Physiology (<a href="http://auwerx-lab.epfl.ch/" target=_blank>LISP</a>). This data set is not yet freely available for global analysis. This data set has not yet been used or described in any publication. Please contact Johan Auwerx or Evan Williams at evan.williams@epfl.ch regarding use of these data. (Implemented by J Auwerx, E Williams, LA Rose, RW Williams and A Centeno).
</Blockquote>

<BLOCKQUOTE>
<STRONG>2011-9-29: We have added liver gene expression data for many strains of mice from </STRONG><a href="http://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE16780" target=_blank>GEO series GSE16780</a>. These data were generated by Dr. Jake Lusis and colleagues at UCLA and are currently listed as a BXD data set, although the study actually includes many other strains (see "GSE16780 UCLA Hybrid MDP Liver Affy HT M430A (Sep11) RMA". Since adding the data we have discovered errors in strain assignment that affect a majority of the conventional inbred strains and several RI strains. For this reason, these data should still be considered provisional.  For complete information please refer to <a href="http://www.ncbi.nlm.nih.gov/pubmed/20054062" target=_blank>A high-resolution association mapping panel for the dissection of complex traits in mice. Genome Res 2010 Feb;20(2):281-90. PMID: 20054062</a> (Implemented by <a href="http://www.ncbi.nlm.nih.gov/pubmed?term=Bennett%20BJ[Author]" target=_blank>Bennett BJ</a>, <a href="http://www.ncbi.nlm.nih.gov/pubmed?term=Ghazalpour%20A[Author]" target=_blank>Ghazalpour A</a>. Data entered on 9-29-11 by A. Centeno under Accession number:<a href="http://www.genenetwork.org/webqtl/main.py?FormID=sharinginfo&GN_AccessionId=373">GN373</a>).
</BLOCKQUOTE>

<BLOCKQUOTE>
<STRONG>2011-9-17: </STRONG>The QTLminer feature was put into GeneNetwork production server. (Implemented by Rudi Alberts, Lei Yan, Ning Liu, Xiaodong Zhou)
</BLOCKQUOTE>

<BLOCKQUOTE>
<STRONG>2011-7-20: </STRONG>GeneNetwork was moved to <a href="http://aws.amazon.com/">Amazon Cloud (EC2)</a>. (Implemented by Lei Yan)
</BLOCKQUOTE>

<BLOCKQUOTE>
<STRONG>2011-7-12: </STRONG>A SourceForge site for GeneNetwork was built: <a href="https://sourceforge.net/projects/genenetwork/">https://sourceforge.net/projects/genenetwork/</a>. (Implemented by Lei Yan, Robert Williams)
</BLOCKQUOTE>

<BLOCKQUOTE>
<STRONG>2011-7-11: </STRONG>A new account on our SVN server for GeneNetwork sharing was created. Anybody can get the latest version of GN source codes by checking out (this account cannot commit).<br>
URL: http://tyche.uthsc.edu/repos/gn/<br>
Username: gndownload<br>
Password: gndownload<br>
(Implemented by Lei Yan)
</BLOCKQUOTE>

<BLOCKQUOTE>
<STRONG>2011-7-5: </STRONG><b>Harvard Brain Tissue Resource Center/ Merck Research Laboratories</b>. This study aims at identifying functional variation in the human genome (especially as it pertains to brain expressed RNAs) and elucidate its relationship to disease and drug response. The ~800 individuals in this dataset are composed of approximately 400 Alzheimers disease (AD) cases, 230 Huntington's Disease and 170 controls matched for age, gender, and post mortem interval (PMI).  The tissue specimens for this study were provided by Harvard Brain Tissue Resource Center (HBTRC).<p>  
Three brain regions (cerebellum, visual cortex, and dorsolateral prefrontal cortex) from the same individuals were profiled on a custom-made Agilent 44K microarray of 39,280 DNA probes uniquely targeting 37,585 known and predicted genes, including splice variants, miRNAs and high-confidence non-coding RNA sequences. The individuals were genotyped on two different platforms, the Illumina HumanHap650Y array and a custom Perlegen 300K array (a focused panel for detection of singleton SNPs). Clinical outcomes available include age at onset, age at death, Braak scores (AD), Vonsattel scores (HD), Regional brain enlargement/atrophy.<p>
<b>Acknowledgements</b>. The Harvard Brain dataset was contributed by Merck Pharmaceutical through the Sage Bionetworks Repository. The tissues were provided by Harvard Brain Tissue Resource Center which is supported in part by PHS grant R24 MH068855 (<a href="http://www.brainbank.mclean.org/" target="_blank">http://www.brainbank.mclean.org/</a>). Investigators: Francine Benes/ Eric Schadt. (Implemented by Megan Mulligan, Rob Williams and Arthur Centeno).</p>
</BLOCKQUOTE>

<BLOCKQUOTE>
<STRONG>2011-6-16: </STRONG><b>CANDLE Study expression data entered in GN</b>. The primary goal of the <A HREF="http://www.candlestudy.org/php/partner.php" target="_blank">CANDLE study</a> is to study factors that affect brain development in young children. To this end, the current study will test specific hypotheses regarding factors that may negatively influence cognitive development in children.
For information on genomic and genetic studies related to CANDLE, please contact: Drs. Ronald M. Adkins (radkins1 at uthsc.edu) and Julia Krushkal (jkrushka at uthsc.edu). This data set is currently confidential. (Implemented by Khyobeni Mozhui, Rob Williams and Arthur Centeno).
</BLOCKQUOTE>

<BLOCKQUOTE>
<STRONG>2011-6-6: </STRONG><B>Confidential Phenotype Trait Feature</B>: We added one new feature of confidential phenotype trait to GeneNetwork. (Implemented by Xiaodong Zhou).
</BLOCKQUOTE>

<BLOCKQUOTE>
<STRONG>2011-6-6: </STRONG><B>SNP INDEL Variant Browser updated</B>: We have greatly expanded and improved the SNP INDEL variant browser that is built into GeneNetwork. This resource enables users to rapidly review both known and confidently imputed sequence variants in the mouse genome. The data set includes over 65 million SNPs that are largely taken from sequencing efforts of David Adams and colleagues at the Sanger Institute and our own team at UT. The imputation of SNPs to other strains was carried out by Eleazar Eskin, Nick Furlotte, and colleagues at UCLA. (Implemented by Ning Liu and Xiaodong Zhou).
</BLOCKQUOTE>

<BLOCKQUOTE>
<STRONG>2011-6-6: </STRONG><B>Major Overhaul of Trait Data pages</B>: The Trait Data and Analysis page has been redesigned to reduce complexity and visual clutter. Functions have not been changed, but may have been moved (Implemented by Zachary Sloan, Xiaodong Zhou, and Rob Williams).
</BLOCKQUOTE>


<BLOCKQUOTE>
<STRONG>2011-6-6: </STRONG><B>Upgraded GeneNetwork Hardware</B>: We are converting GeneNetwork to MySQL master-slave replication with faster solid-state hard drives to improve performance.  (Implemented by Lei Yan).
</BLOCKQUOTE>


<BLOCKQUOTE>
<STRONG>2011-3-21: </STRONG><B>Sample blocking</B>: The user can now block individual samples/strains in the Trait Data and Analysis page by typing either an individual index number or a range (ex: 1,2,3,10-20). This feature was created to eliminate the need for a user to manually replace each sample's value with 'x'. (Implemented by Zachary Sloan).
</BLOCKQUOTE>

<BLOCKQUOTE>
<STRONG>2011-3-18: </STRONG><B>User login status</B>: The user login status is shown in all dynamically generated pages by making use of session mechanism through entire GN system. (Implemented by Xiaodong Zhou).
</BLOCKQUOTE>

<BLOCKQUOTE>
<STRONG>2011-2-23: </STRONG><B>More space for your Trait Collections</B>: We have greatly expanded the number of traits, transcripts, genes, and markers be added to your collections. The current limit is now 3,000; up from 100 in the previous version. This improvement was achieved by storing collection information using a different and more secure method (session control rather than cookies) (Implemented by Xiaodong Zhou).
</BLOCKQUOTE>

<BLOCKQUOTE>
<STRONG>2011-2-9: </STRONG><B>INIA Amygdala BLA Affy MoGene 1.0 ST (Nov10)</B>: <P>This is a final error-checked release of an amygdala gene expression data set generated by Khyobeni Mozhui, Lu Lu, and Robert W. Williams and colleagues with funding support from the NIH NIAAA. The basolateral complex of the amygdala of untreated young adult mice was profiled using the Affymetrix GeneChip <A HREF="http://www.affymetrix.com/browse/products.jsp?productId=131476&categoryId=35765#1_1">Mouse Gene 1.0 ST</A> array that contains approximately 34,728 probe sets that target approximately 29,000 well defined transcripts (RefSeq mRNA isoforms) and essentially all known protein coding genes in mouse. This array is an "exon style" array with multiple probes in all known exons of each gene (an average of about 27 per gene) and is an abridged version of the Affymetrix Exon 1.0 ST array. However, it also does contain some probes that target non-coding RNAs and even miRNA precursors (search "ncrna").<br>
This data set is not yet freely available for global analysis. This data set has not yet been used or described in any publication. Please contact Robert W. Williams at rwilliams@uthsc.edu regarding use of these data.
</BLOCKQUOTE>

<BLOCKQUOTE>
<STRONG>2011-2-9: </STRONG><B>INIA Hypothalamus Affy MoGene 1.0 ST (Nov10)</B>: <P>These hypothalamic gene expression data were generated by Khyobeni Mozhui, Lu Lu, and Robert W. Williams and colleagues with funding support from NIAAA. The data set includes samples from 50 strains, including 46 BXDs, both parental strains, and reciprocal F1 hybrids. Expression data were generated using the Affymetrix Mouse Gene 1.0 ST exon-style microarray (multiple probes in all known exons) by Lorne Rose in the UTHSC Molecular Resources Center (MRC), Memphis TN. The table below provides a summary of cases, sex, and age. Hypothalamic tissue was dissected by K. Mozhui (description to follow) with special attention to time of day (every sample has time stamp). RNA was extracted by K. Mozhui. All other processing steps by the UTHSC MRC by L. Rose. Data were processed by Arthur Centeno.<br>
This data set is not yet freely available for global analysis. This data set has not yet been used or described in any publication. Please contact Robert W. Williams at rwilliams@uthsc.edu regarding use of these data.
</BLOCKQUOTE>

<BLOCKQUOTE>
<STRONG>2011-2-7: </STRONG><B>GenEx BXD EtOH Liver Affy M430 2.0 (Jan11) RMA **</B>: Data set generated with support of NIAAA by Drs. Robert Rooney (Genome Explorations Inc.), Divyen Patel (Genome Explorations Inc.), and Kristin Hamre (UTHSC). All animals were on standard chow and water ad lib. Both the saline control group and the ethanol=treated group were given solutions via intragastric gavage with controls getting saline and the alcohol-treated group getting 6g/kg of ethanol. Ethanol-treated mice were generally comatose for 4-6 hrs but were responsive and moving by the next morning. Tissue was collected at 24 hours after the initial infusion.<br>
This data set is not yet freely available for global analysis. This data set has not yet been used or described in any publication. Please contact Robert W. Williams at rwilliams@uthsc.edu regarding use of these data.
</BLOCKQUOTE>

<BLOCKQUOTE>
<STRONG>2011-2-3: </STRONG><B>OHSU HS-CC Striatum ILM6v1 (Feb11) RankInv</B>: The current study focused on the extent genetic diversity within a species (Mus musculus) affects gene
co-expression network structure. To examine this issue, we have created a new mouse resource, a heterogeneous
stock (HS) formed from the same eight inbred strains that have been used to create the collaborative cross (CC). The
eight inbred strains capture > 90% of the genetic diversity available within the species. For contrast with the HS-CC, a
C57BL/6J (B6) Ã— DBA/2J (D2) F2 intercross and the HS4, derived from crossing the B6, D2, BALB/cJ and LP/J strains,
were used. Brain (striatum) gene expression data were obtained using the Illumina Mouse WG 6.1 array, and the data
sets were interrogated using a weighted gene co-expression network analysis (WGCNA).<br>
Read full article: <a href="dbdoc/GN304_Iancu_Hitzemann.pdf" target="_blank">Genetic diversity and striatal gene networks: focus on the heterogeneous stock-collaborative cross (HS-CC) mouse</a>.
</BLOCKQUOTE>

<BLOCKQUOTE>
<STRONG>2011-1-26: </STRONG><B>Auwerx Lab BXD Phenotype Data</B>: We have entered the first large-scale metabolic, cardiovascular, and clinical chemistry data sets (n = 143 phenotypes) for the BXD strains of mice. Data are averages for males and females separately, for as many as 43 strains. Data were generated using animals born at UTHSC (Memphis) and phenotyped in Strasbourg, France in 2008 using <A HREF="http://empress.har.mrc.ac.uk/about.html">EMPReSS Slim</A> EUMORPHIA standard operating protocols. Blood pressure phenotypes were included in a PLoS Genetics paper in <A HREF="http://www.plosgenetics.org/article/info%3Adoi%2F10.1371%2Fjournal.pgen.1000591">2009</A>, but we now provide the complete phenotypes from this cohort of animals. (Phenotyping by Hana Koutinkova, Johan Auwerx and colleagues; data processing by H Koutnikova, RW Williams, EG Williams, and Xiaodong Zhou).

</BLOCKQUOTE>

<BLOCKQUOTE>
<STRONG>2011-1-26: </STRONG>Expression data for the prefrontal cortex of the BXD strains have added into GN, mice from each genotype received 4 weekly cycles of chronic intermittent ethanol (CIE) vapor exposure (EtOH group) or air exposure (CTL group) in inhalation chambers. The general study design was generated by Dr. Michael Miles and colleagues. All data sets are currently being tested. Contact Dr. Miles at VCU Medical Center for access (Implemented by M Miles, RW Williams and A Centeno).
<DIR>
<OL>
<LI>VCU BXD PFC CIE Air M430 2.0 (Jan11) RMA **
<LI>VCU BXD PFC CIE EtOH M430 2.0 (Jan11) RMA **
<LI>VCU BXD PFC EtOH vs CIE Air M430 2.0 (Jan11) Sscore **
</OL>
</DIR>
</BLOCKQUOTE>

<BLOCKQUOTE>
<STRONG>2011-1-14: </STRONG>NCSU Expression data set for Drosophila melanogaster have been added to GeneNetwork. For more information about The Drosophila Genetic Reference Panel <A HREF="http://service004.hpc.ncsu.edu/mackay/Good_Mackay_site/DBRP.html" target="_blank">click here</A>. (Julien F. Ayroles, Trudy F. C. Mackay).
<DIR>
<OL>
<LI>NCSU Drosophila Whole Body (Jan11) RMA
</OL>
</DIR>
</BLOCKQUOTE>

<BLOCKQUOTE>
<STRONG>2011-1-11: </STRONG>Genome Explorations Expression data sets for Liver have been added to GeneNetwork. (Implemented by B Rooney, K Hamre, RW Williams and A Centeno).
<DIR>
<OL>
<LI>GenEx BXD Sal Liver Affy M430 2.0 (Jan11) RMA Females ** 
<LI>GenEx BXD Sal Liver Affy M430 2.0 (Jan11) RMA Males **
<LI>GenEx BXD Sal Liver Affy M430 2.0 (Jan11) RMA Both Sexes **
<LI>GenEx BXD EtOH Liver Affy M430 2.0 (Jan11) RMA Females **
<LI>GenEx BXD EtOH Liver Affy M430 2.0 (Jan11) RMA Males **
<LI>GenEx BXD EtOH Liver Affy M430 2.0 (Jan11) RMA Both Sexes **
</OL>
</DIR>
</BLOCKQUOTE>

<BLOCKQUOTE>
<STRONG>2010-12-22:</STRONG><B> Phenotype traits</B> can be searched by LRS. For instance, search by LRS=(23 46) or LRS=(9 99 Chr4 122 155). (Implemented by Xiaodong Zhou).
</BLOCKQUOTE>

<BLOCKQUOTE>
<STRONG>2010-12-20: </STRONG>The final quality controlled release of a spleen gene expression data set generated by a DOD-funded consortium (Byrne, Kotb, Williams, and Lu) has been entered in GeneNetwork.
<DIR>
<OL>
<LI>UTHSC Affy MoGene 1.0 ST Spleen (Dec10) RMA 
</OL>
</DIR>
</BLOCKQUOTE>

<BLOCKQUOTE>
<STRONG>2010-12-10:</STRONG>
We built a new <a href="http://richard.uthsc.edu/" target="_blank">Cluster</a> system. Four nodes were deployed. Every node is Dell PowerEdge R815, has 48 cores, 64G RAM, 1.3G RAM/core. CPU is AMD 1.86G Hz. The headmaster connects one Dell PowerVault MD1000 (15*2TB hard drives, RAID5).
And <a href="http://richard.uthsc.edu:8080/" target="_blank">Galaxy</a> system was installed on the Cluster system.
(Implemented by Lei Yan).
</BLOCKQUOTE>


<BLOCKQUOTE>
<STRONG>2010-10-29:</STRONG><B> First Developmental Studies of the Genetics of Gene Expression in Brain</B> have been entered into GeneNetwork. Dr. Glenn Rosen and colleagues have contributed data on gene expression across sets of 32 BXD strains for the neocortex and striatum at two stages of development (postnatal days 3 and 14). They used the Illumina Mouse Genome 6 version 2 array. These data are matched by previous data sets for the adult neocortex and striatum. These data are open but users are requested to contact Glenn D. Rosen regarding the status of these new data. (Implemented by G Rosen, RW Williams and A Centeno).
</BLOCKQUOTE>

<BLOCKQUOTE>
<STRONG>2010-10-29:</STRONG><B> Large RNA-seq data for BXD Whole Brain</B> is being added to our GeneNetwork version of the UCSC Genome Browser. We are still loading these data and eventually will display RNA-seq data for over 30 strains of mice. Gene level summary of these data will be entered into GeneNetwork for quantitative analysis later this year. Please contact RW Williams regarding the status of these new data. (Implemented by David Li, Lu Lu, Xusheng Wang, Lei Yan, and RW Williams).
</BLOCKQUOTE>

<BLOCKQUOTE>
<STRONG>2010-9-23:</STRONG> Genotype data for mouse strains BXD101, BXD102, and BXD103 have been added to GeneNetwork. These data were extracted from a large scale regenotyping of the BXD strains done using the Mouse Diversity array from Affymetrix that was designed by Pardo and Churchill. (Implemented by Ning Liu and Xiaodong Zhou).
</BLOCKQUOTE>

<BLOCKQUOTE>
<STRONG>2010-9-20:</STRONG> Five groups of gene expression data for the hippocampus of BXD strains of mice have been entered in GeneNetwork. NON = No stress and no saline control injection; NOS = No restraint stress and given only saline injections prior to sacrifice; NOE = No restraint stress and given an ethanol injection prior to sacrifice; RSS = short restraint stress (1 episode) followed by a saline injection; and finally, RSE = Restraint stress followed by an ethanol injection. (Implemented by Lu Lu, RW Williams and A Centeno).
</BLOCKQUOTE>

<BLOCKQUOTE>
<STRONG>2010-9-13:</STRONG>  There is now a good <A HREF="http://en.wikipedia.org/wiki/Genenetwork" class="normalsize"  target="_blank">Wikipedia</A> entry for GeneNetwork. Please check, correct, and improve. (Implemented by RW Williams).
</BLOCKQUOTE>

<BLOCKQUOTE>
<STRONG>2010-9-13:</STRONG> GeneNetwork source code is now licenses using the <A HREF="http://www.gnu.org/licenses/agpl.html">Affero General Public License version 3</A>. A SourceForge site will be set in the next few months. In the interim, please contact us directly for code. (Implemented by Xiaodong Zhou and RW Williams).
</BLOCKQUOTE>

<BLOCKQUOTE>
<STRONG>2010-08-02:</STRONG> We have installed a <A HREF="http://ucscbrowser.genenetwork.org/" class="normalsize"  target="_blank">GeneNetwork UCSC Genome Browser</A> mirror site that displays a set of 4.58 million SNPs that distinguish strain DBA/2J from the reference strain, C57BL/6J. Other DNA sequence and RNA-seq data sets will be added over the next several months. (Implemented by Lei Yan and Xusheng Wang).
</BLOCKQUOTE>


<BLOCKQUOTE>
<STRONG>2010-08-02:</STRONG> <A HREF="http://www.genenetwork.org/webqtl/main.py?FormID=tissueCorrelation" class="normalsize"  target="_empty">Tissue Correlation and Expression Level Services</A>: We have added a new web interface that allows you to directly evaluate differences in gene expression across 30 different tissue types. 

<P>The current version is mainly meant for testing. The interface still needs work.We have a long list of improvements in the works, but please send us your ideas. To test drive the tissue correlation feature you currently need to enter a set of mouse GeneID numbers. For example <I><A HREF="http://www.ncbi.nlm.nih.gov/gene/11820">App</A></I> is  NCBI Enbrez Gene ID 11820.  <I><A HREF="http://www.ncbi.nlm.nih.gov/gene/23821">Bace</A></I> is Gene ID 23821. If you enter these two numbers in the interface and then click your heels twice. You should get back a simple matrix of values that lists both Pearson and Spearman correlations based on a comparison of expression in 25 tissue types. Click on the correlation values and this will pop up two scatterplots (Pearson and Spearman types).

<P>At the same time, this tool provides expression estimates for both genes, where a value of 8 in the Pearson plot represents the mean across all tissues and each unit represents a two-fold difference (log2 expression; Spearman rank values are just that--rank out of 25). To access this new feature select <B>Search -> Tissue Correlation</B>. All of the expression data are taken from on C57BL/6J litter mates studied using the Illumina Mouse 6 2.0 array (Implemented by Ning Liu, data from Lu Lu, RW Williams, and Xusheng Wang).

</BLOCKQUOTE>

<BLOCKQUOTE>
<STRONG>2010-08-02: </STRONG>The javascript that controls the sort menu on top of collection page has been improved. The old version doesn't work properly in IE browser. The new one works well in IE, Firefox, Safari. (Implemented by Ning Liu).
</BLOCKQUOTE>



<BLOCKQUOTE>
<STRONG>2010-07-02: </STRONG>Microarray annotations have been improved and will soon be more consistent across platforms. We now synchronize the gene level annotation of probes and probe sets. When gene level attributes such as gene symbols, alias, name, and other identifiers, are changed, the change is applied to all other probes and probe sets with the same Gene Id. (Implemented by Xiaodong Zhou).
</BLOCKQUOTE>
<BLOCKQUOTE>
<STRONG>2010-07-02: </STRONG>We have added Homologene identifiers to help in comparative analysis. (Implemented by Xiaodong Zhou)
</BLOCKQUOTE>


<BLOCKQUOTE>
<STRONG>2010-07-02: </STRONG>Improved Search Page that loads much more rapidly from even slow connections. We replaced the AJAX version of the Search Page with a javascript version. (Implemented by Ning Liu).
</BLOCKQUOTE>


<BLOCKQUOTE>
<STRONG>2010-07-02: </STRONG>Improved icons and GUI for selecting and analyzing GeneNetwork data sets. The use of icons enable fast recognition of functions and is also more compatible with touch screen interfaces. There are three types of icons:

<OL>
<LI>Selection tools with a grey background
<LI>GeneNetwork analysis tools with blue background
<LI>External resources analysis tools with a clear background
</OL

Here is what a small collection window should look like. 


<P><DIR><IMG src="/images/upload/New Interface.png" valign="top" width="685" height="609"></DIR>

This GUI was implemented by Zach Sloan.

</BLOCKQUOTE>


<BLOCKQUOTE>
<STRONG>2010-07-02: </STRONG>We have improved interface for the partial correlations, Now both the zero order and higher order correlations use identical sample sizes for more direct comparisons. We have also added several checking procedures to help you avoid undesirable results. (Implemented by Xiaodong Zhou).
</BLOCKQUOTE>
<!--ARTHUR STARTS EDITING HERE-->
<!--Code below to Show hide Contact information -->
<HR>
<a href="#" onclick="colapse('answer1')">See More News</a><br>
<span id="answer1" style="display: none; return: false;">
<!--OLD NEWS START HERE-->
<BLOCKQUOTE>
<STRONG>2010-06-29: </STRONG>A new <a href="http://rhodes.uthsc.edu/">GeneNetwork Wiki (MediaWiki)</a> was built, and the old GeneNetwork Wiki (TWiki) was frozen for security issues. (Implemented by Lei Yan).
</BLOCKQUOTE>

<BLOCKQUOTE>
<STRONG>2010-06-29: </STRONG> <a href="https://www.ornl.gov/bsec_conferences/2010/Day_1/Williams_BSEC_May2010.pd">PDF</A> and tour of GeneNetwork presented at the Oak Ridge National Laboratory that is useful as a primer and good source of images for your own work or presentations. (Implemented by R Williams).
</BLOCKQUOTE>

<BLOCKQUOTE>
<STRONG>2010-06-18: </STRONG><B>Mouse brain and eye RNA-seq data</B> displays on the UCSC Genome Browser have been upgraded by Xusheng Wang for C57BL/6J and DBA/2J strains of mice. 

<OL>
<LI><A HREF="http://genome.ucsc.edu/cgi-bin/hgTracks?hgS_doOtherUser=submit&hgS_otherUserName=Williamslab&hgS_otherUserSessionName=RNAseq_Brain_Log">brain</A>
<LI><A HREF="http://genome.ucsc.edu/cgi-bin/hgTracks?hgS_doOtherUser=submit&hgS_otherUserName=Williamslab&hgS_otherUserSessionName=eye_RNAseq_log2">eye</A>
<LI><A HREF="http://genome.ucsc.edu/cgi-bin/hgTracks?hgS_doOtherUser=submit&hgS_otherUserName=Williamslab&hgS_otherUserSessionName=Brain_eye_log2">brain and eye combined</A>

</OL>

Sequence tag counts are plotted on a Log2 Y axis over a range from 2 to 10 (n = 2^2 to 2^10). (Implemented by Xusheng Wang, David Li, and colleagues).
</BLOCKQUOTE>


<BLOCKQUOTE>
<STRONG>2010-05-18: </STRONG><A HREF="http://genome.cshlp.org/content/19/6/1133.long">Oxford Heterogeneous Stock Expression Data Sets for Hippocampus, Lung, and Liver</A> have been added to GeneNetwork. These are large (n > 250), powerful, and well structured data sets from a single cohort of highly recombinant outcrossed mice that have a genetic architecture similar to that of human populations. All tissue was taken from non-inbred mice generated by systematically outcrossing a stock of eight inbred strains by Robert Hitzemann and colleagues (A/J, AKR/J, BALBc/J, CBA/J, C3H/HeJ, C57BL/6J, DBA/2J, and LP/J). Please refer to the publication by Huang and colleagues (<A HREF="http://genome.cshlp.org/content/19/6/1133.long" class="fs14"  target="_empty">2008</A>). 

<P>The Oxford group (Jonathan Flint, Richard Mott, and colleagues) maintains a complete <A HREF="http://mus.well.ox.ac.uk/mouse/HS/#eqtls" class="fs14"  target="_empty">data repository</A> and key summary data that can be accessed at 

<P>All correlation and network graphing functions are currently implemented in GeneNetwork. You can generate correlation matrices, graphs, and compute partial correlations. However, QTL mapping has not yet been implemented for these three data sets. Traits have been mapped by the Oxford group and QTLs can be viewed using <A HREF="http://mus.well.ox.ac.uk/cgi/gscan/wwwqtl.cgi" class="fs14"  target="_empty">GSCANDB</A>. We are currently working with Dr. Richard Mott and colleagues to include a real-time R Happy package in GeneNetwork that will enable users to map the original trait data and derivatives. (Implemented by Arthur Centeno, Rob Williams, and Jonathan Flint).

<P><DIR><IMG src="/images/upload/HSmice.jpg" valign="top" width="600" height="272"></DIR>

</BLOCKQUOTE>

<P>
<P>


<BLOCKQUOTE>
<STRONG>2010-05-18: </STRONG>Partial Correlations can now be calculated in GeneNetwork. We have integrated <A HREF="http://www.r-project.org/">R Project</A> code to compute up to third order partial correlations. This makes it possible to compute the correlation between X and Y traits while controlling for the possible confounding effects of three other variables, for example the age of cases, experimental groups, or even a genetic marker. To use the Partial Correlation feature, please add X, Y, and the possible confounder traits or markers into a Collection. Then just look for the Partial Correlation icon and text. (Implemented by Xiaodong Zhou).

<P><DIR><IMG src="/images/upload/PartialR_Icon.png" valign="top"></DIR>

</DIR>
</BLOCKQUOTE>
<BLOCKQUOTE>
<STRONG>2010-05-01: </STRONG><B>Initial RNA-seq data</B> for brains and eyes of C57BL/6J and DBA/2J strains. We are now experimenting with RNA-seq expression data generated using SOLiD short (50 nt) reads. Links to initial RNA-seq data are provided for <A HREF="http://genome.ucsc.edu/cgi-bin/hgTracks?hgS_doOtherUser=submit&hgS_otherUserName=Williamslab&hgS_otherUserSessionName=RNAseq_Brain_Log">brain</A> and <A HREF="http://genome.ucsc.edu/cgi-bin/hgTracks?hgS_doOtherUser=submit&hgS_otherUserName=Williamslab&hgS_otherUserSessionName=eye_RNAseq_log2">eye</A>. Data are plotted on a Log2 Y axis over a range from 2 to 10 (2^2 to 2^10 tags). (Implemented by Xusheng Wang).
</BLOCKQUOTE>

<BLOCKQUOTE>
<STRONG>2010-04-08: </STRONG>We have added new options for scatterplots. Font size, marker type and size, and colors can now be changed. As before, we output both types of correlations. (Implemented by Zachary Sloan).
</BLOCKQUOTE>

<P><DIR><IMG src="/images/upload/APOE_Scatter.png" valign="top" width="900" height="410"></DIR>
<DIR><P><SMALL><B>Legend:</B> Examples of scatterplots of <I>Apoe</I> and <I>Aplp1</I> expression in the BXD hippocampus using both Pearson (left) and Spearman rank (right) correlations (50% of original size).</SMALL></P>
</DIR>

<BLOCKQUOTE>
<STRONG>2010-03-31: </STRONG>First RNA-seq data available on the <A HREF="http://genome.ucsc.edu/cgi-bin/hgTracks?hgS_doOtherUser=submit&hgS_otherUserName=Williamslab&hgS_otherUserSessionName=RNAseq" class="fs14"  target="_empty">UCSC Genome Browser</A>. The RNA-seq data were generated using whole brain mRNA (polyadenylated) from a total of 12 animals: 4 C57BL/6J young adults (2 males and 2 females), 4 DBA2/J young adults (also 2M and 2F), 2 B6D2F1 (1M and 1F), and 2 D2B6F1 (1M and 1F).
(Implemented by Xusheng Wang, Lu Lu, William Taylor, RWW, and David Li)
</BLOCKQUOTE>

<BLOCKQUOTE>
<STRONG>2010-03-31: </STRONG>Added search "tip text" to the main search page. (Implemented by Lei Yan)
</BLOCKQUOTE>

<BLOCKQUOTE>
<STRONG>2010-03-23: </STRONG>Rank correlation plots have been computed and implemented from Correlation Table output pages and Correlation Matrix output plots. We now routinely output both types of correlations for diagnostic purposes. Users can change the plot size and labeling. More output options are  in progress. (Implemented by Zachary Sloan)
</BLOCKQUOTE>

<BLOCKQUOTE>
<STRONG>2010-03-11: </STRONG>Both dynamic and static web pages now include a separate header and footer that administrators can edit from the manager interface. (Implemented by Lei Yan)
</BLOCKQUOTE>

<BLOCKQUOTE>
<STRONG>2010-03-08: </STRONG>The interval mapping function for Mouse Diversity Panel (MDP) group has been fixed. (Implemented by Xiaodong Zhou)
</BLOCKQUOTE>

<BLOCKQUOTE>
<STRONG>2010-01-08: </STRONG>The Network Graph page has been improved. Several new graph algorithm options are provided. We added one function called "Lock Graph Structure" that allows the user to hold the position of all nodes and the length of all edges constant, letting him/her easily compare between different correlation types. In the image, the nodes and lines between them are "hot linked" now so that user can easily check each trait and correlation. User can also export the raw data (traits and correlations) in different format as input to other network graph software such as <a href="http://www.cytoscape.org/">Cytoscape</a>. We also fixed a bug where the user was unable to correctly select an edge's color. (Implemented by Zachary Sloan and Xiaodong Zhou)
</BLOCKQUOTE>

<BLOCKQUOTE>
<STRONG>2009-12-04: </STRONG>Search page has been redesigned. The new design provides more help to new users. (Implemented by RW Williams)
</BLOCKQUOTE>

<BLOCKQUOTE>
<STRONG>2009-11-20: </STRONG>TouchGraph's Navigator software has an interesting applet that lets you see the relationship between GeneNetwork and other sites on the web. Click <a href="http://www.touchgraph.com/TGGoogleBrowser.php?start=genenetwork.org&signed=false">here</a> to check it out.
</BLOCKQUOTE>


<BLOCKQUOTE>
<STRONG>2009-11-12: </STRONG>Swiss GeneNetwork site is on-line as of November 16. This mirror site is in the Laboratory for Integrative Systems Biology at the EPFL (Prof Johan Auwerx and colleagues) in Lausanne (Implementation by Evan Williams, Lei Yan, and Johan Auwerx.)
</BLOCKQUOTE>

<BLOCKQUOTE>
<STRONG>2009-11-6: </STRONG>Fly Toxicogenomic data sets are expected to be available on the GN production web site in early December 2009. We are currently experimenting with two data sets generated by Douglas Ruden and colleagues: 
<DIR>
<OL>
<LI><I>UAB Whole body D.m. mRNA control (Oct09) RMA</I> 
<LI><I>UAB Whole body D.m. mRNA lead acetate (pbAc) (Oct09) RMA</I>
</OL>
</DIR>
Raw data were provided by Dr. Grier Page and can be found on the <a href="http://alexandria.uthsc.edu:84/" target="_blank">GeneNetwork development server</a>.  (Implementation by Arthur Centeno, Robert W. Williams, Doug Ruden, Grier Page, and Xiaodong Zhou.)
</BLOCKQUOTE>

<BLOCKQUOTE>
<STRONG>2009-10-27: </STRONG>Fixed links to <a href="http://syndb.cbi.pku.edu.cn/sdb_text.php">PKU SynDB</a>. (The API interface methods was changed from GET to POST.) (Implementation by Lei Yan)
</BLOCKQUOTE>

<BLOCKQUOTE>
<STRONG>2009-09-24: A </STRONG><A HREF="http://artemis.uthsc.edu/">GeneNetwork Archive/Time Machine</A> server has be set up. This systems allows users to work with  older versions of GeneNetwork that correspond well with specific publications. Each version is designated by a time stamp (2009-03-04, for the March 4, 2009 version). (Implementation by Lei Yan)
</BLOCKQUOTE>

<BLOCKQUOTE>
<STRONG>2009-09-21: </STRONG>We have added 150 behavioral traits that are related to anxiety levels in the BXD  strains of mice into GeneNetwork. Mice fall into five different treatment groups: baseline (BASE), treated with a saline control injection (no stress, only saline; NOS), treated with an ethanol (no stress, only ethanol; NOE), treated to mild restraint stress followed by a saline injection (RSS), treated to mild restraint stress followed by an ethanol injection (RSE). Each of these five sets of 30 behavioral assays is matched to a corresponding hippocampal gene expression data set (Illumina). Please search for "Melloni" in the BXD Phenotypes database ANY or ALL fields. (Implementation by Xiaodong Zhou, Melloni Cook, and Lu Lu.)
</BLOCKQUOTE>

<BLOCKQUOTE>
<STRONG>2009-09-09: </STRONG>Unit testing of all GeneNetwork servers and mirrors has been implemented. This system monitors the performance of six different GN functions using Get commands. Success is evaluated by comparing the length of the returned html document with expectation. When a system fails, an error message is sent to the Sys Admin.  (Implementation by Lei Yan)
</BLOCKQUOTE>

<BLOCKQUOTE>
<STRONG>2009-09-09: </STRONG> Barley gene expression data has being integrated into GeneNetwork. All data are from Matthew Moscou, Roger Wise and Nick Lauter at the Iowa State University. Data is currently public with the correspondent <a href="http://www.genenetwork.org/dbdoc/B1LI0809R.html" class="fs14">acknowlegment of data use and disclaimer</a>, and those interested in obtaining full access should contact moscou at iastate.edu (Implementation by Matthew Moscou, Robert W. Williams, and Arthur Centeno)
<P><DIR><IMG src="http://www.genenetwork.org/images/upload/barley_QSM.png" valign="top" width="420" height="167"></DIR>
</BLOCKQUOTE>

<BLOCKQUOTE>
<STRONG>2009-09-09: </STRONG> We have added a simple web interface used by data managers to delete phenotype trait. (Implementation by Xiaodong Zhou)
</BLOCKQUOTE>


<BLOCKQUOTE>
<STRONG>2009-08-24: </STRONG><B><A HREF="http://faculty.chass.ncsu.edu/garson/PA765/partialr.htm">PARTIAL CORRELATIONS</A></B> have been implemented in GeneNetwork on the <A HREF="http://alexandria.uthsc.edu:85">beta test site</A> and will soon be move to our production site. You can now compute correlations between a "reference trait" and any other large set of traits while controlling for variability associated with up to three other factors. For example, it is possible to compute the correlation between the expression of the gene formin 2 (<I>Fmn2</I>) and all other 45,000 probe sets in the BXD hippocampus data set while controlling for the effects of the genotype at the location of the <I>Fmn2</I> gene itself. This is done by using the SNP rs6375522 that is located in <I>Fmn2</I>as a controlled cofactor. In this particular situation the partial correlation then provides you with a measure of the association between differences in formin 2 expression and all other transcripts in the absence of genetic variation in formin 2 itself. Partial correlations are a great way to study correlations between genes with control for sometimes unwanted sources of variability--linkage with neighboring genes being one important unwanted source. Partial correlations can also be used to remove effects of sex, age, batch effect, etc. Users can upload their own "correction" factors that can be used to calculate partial correlations. (Implementation by David Crowell)
</BLOCKQUOTE>

<BLOCKQUOTE>
<STRONG>2009-08-20: </STRONG><B><A HREF="http://www.genenetwork.org/webqtl/snpBrowser.py">SNP BROWSER UPGRADE</A></B>. We have reworked and renamed the SNP Browser module of GeneNetwork. The new name is <B>Variant Browser</B>, a change we made because we are adding data on indels (about 140,000 so far). All indel data is from our own comparative whole-genome sequence analysis of indels between C57BL/6J and DBA/2J. The figure below, for example, shows four indels in the mouse <I>Hc</I> gene. The Variant Browser is currently only useful for the analysis of mouse genomes, but we hope to provide useful links and tools from other species in the next few years. (Implementation by Evan G. Williams)
</BLOCKQUOTE>


<P><DIR><IMG src="http://www.genenetwork.org/images/upload/IndelsinHc.png" valign="top" width="422" height="252"></DIR>
<DIR><P><SMALL><B>Legend:</B> Example of the new Variant Browser illustrating results of an indel search for hemolytic complement (<I>Hc</I>).</SMALL></P>
</DIR>


<BLOCKQUOTE>
<STRONG>2009-08-12: </STRONG><B>A massive human brain expression data set (363 cases) for the study of Late-Onset Alzheimer's Disease (LOAD)</B> with age-matched elderly control subjects has been added to GeneNetwork. This cortical expression data set is taken from the work of Dr. Amanda Myers and colleagues (see <A HREF="http://www.ncbi.nlm.nih.gov/projects/geo/query/acc.cgi?acc=GSE15222" class="fs14"  target="_empty">GEO GSE15222</A>, <A HREF="http://labs.med.miami.edu/myers" class="fs14"  target="_empty">Webster et al., 2009</A>). (Implementation by Robert W. Williams and Arthur Centeno)
</BLOCKQUOTE>

<BLOCKQUOTE>
<STRONG>2009-08-10: </STRONG> Service upgrade: We have developed a program that monitors free space on GeneNetwork servers. When hard drive space gets too tight, email notification is sent to system administrators. (Implementation by Lei Yan)
</BLOCKQUOTE>

<BLOCKQUOTE>
<STRONG>2009-08-10: </STRONG> The <B>QTL Heat Map and the QTL Cluster Map</B> have been significantly improved. The QTL heat map now allows the user to reorder traits with much greater flexiblity, using all of the sorting functions that are built into the Trait Collection window (sort by position, sort by database, sort by symbol, sort by LRS, etc.). By default, the order of traits in the Traits Collection window is used by the QTL heat map. In addition, the QTL heat maps can be clustered, as usual, by trait correlations--the Cluster Map function developed originall by Elissa Chesler. We have made some minor changes to the heat map display that allow figures to be oriented horizontally. (Implementation by Xiaodong Zhou.)
</BLOCKQUOTE>

<BLOCKQUOTE>
<STRONG>2009-07-29: </STRONG> We have fixed the "Set to Default" function so that it works properly with any species on GeneNetWork.  (Implementation by Xiaodong Zhou)
</BLOCKQUOTE>

<BLOCKQUOTE>
<STRONG>2009-07-27: </STRONG> All mouse genome position data on GeneNetwork have been updated from mouse assembly mm8 to assembly mm9. This has affected the position of the genes, probes, and SNPs. The mm8 assembly remains on the <a href="http://gn.genetics.ucla.edu">mirrors</a>, temporarily. (Implementation by Evan Williams)
</BLOCKQUOTE>

<BLOCKQUOTE>
<STRONG>2009-07-27: </STRONG> We have fixed the Principal Component Analysis (PCA) function to handle as many traits as computationally feasible given the sample size and numbers of cases. This improved PCA code is used when computing correlation matrices.  (Implementation by Xiaodong Zhou)
</BLOCKQUOTE>

<BLOCKQUOTE>
<STRONG>2009-07-26: </STRONG><B>Human brain expression data</B> in patients with Alzheimer's disease and age-matched elderly control subjects. This cortical expression data set is taken from <A HREF="http://www.ncbi.nlm.nih.gov/projects/geo/query/acc.cgi?acc=GSE5281" class="fs14"  target="_empty">GEO GSE5281</A> (<A HREF="http://physiolgenomics.physiology.org/cgi/reprint/28/3/311" class="fs14"  target="_empty">Liang et al. 2006</A>). Samples were laser-captured from cortical layer 3 (except the hippocampus) and run on the Affymetrix U133 Plus 2.0 array. We renormalized the data to an average expression of 8 units on a log2 scale. Two versions of the data have been entered in GeneNetwork: one consisting of 157 of 161 arrays; the second consisting of what we regard as the best 102 arrays (those with mean correlations of better than 0.88 with all other arrays). Case IDs have the following code structure: Brain Region, GEO ID, Sex, Age, Disease Status. E119615M63N is a sample of the entorhinal cortex of case GSM119615, a male 63 year old normal case. The tissue codes are E = enorhinal cortex, H = hippocampus pyramidal layer, MT = medial temporal cortex, PC = porterior cingulate cortex, SP = supeior frontal cortex,  V = primary visual cortex. GeneNetwork does not allow sophisticated display of the data, but you can perform correlation analyses of any of the 56,000 probe sets. For example expression of the <I>APP</I> transcript is higher in the AD cases and correlates well with many other AD related genes. At least 7.5% of cases are assigned incorrectly by sex (see <A HREF="http://www.genenetwork.org/dbdoc/GSE5281_F_RMA0709.html" class="fs14"  target="_empty">INFO</A> file) (Implementation by Robert W. Williams and Arthur Centeno)
</BLOCKQUOTE>


<BLOCKQUOTE><STRONG>2009-07-24: </STRONG>We have added annotation for the Affymetrix U133 Plus 2.0 human microarray (GPL570). Annotation file download from NetAffx July 2009. Minor reannotation by Arthur Centeno prior to entry into GeneNetwork. This annotation file will be used with GSE5281 (steph-affy-human-433773), a data set of gene expression in six brain regions from normal and Alzheimer's disease patients. </BLOCKQUOTE>

<BLOCKQUOTE><STRONG>2009-07-22: </STRONG> We have finished the first phase of sequencing the genome of DBA/2J at UTHSC (SOLiD) and UCLA (Solexa) and have extracted a set of 2.8 million SNPs with comparatively high quality scores that differ between DBA/2J and the C57BL/6J. Many of these SNPs are novel. Some overlap with Celera and Perlegen data. All SNPs have been added to the GeneNetwork SNP browser using the laboratory identifier symbols "MRS1xxxxxxx" label. (Implementation by Evan Williams and Xusheng Wang) </BLOCKQUOTE>

<BLOCKQUOTE><STRONG>2009-07-21: </STRONG> Imputed alleles has been added for 6 million SNPs and a panel of 74 strains in the GeneNetwork SNP browser. Data were generated by the Genome Dynamics group at the Jackson Laboratory using the Perlegen data set as a reference. The GN SNP browser tables are now more significantly larger in terms of both mouse strain and SNP coverage. The figure below illustrates results for <I>Zim3</I> in the GN SNP browser. (Implementation by Evan Williams) </BLOCKQUOTE>

<P><DIR><IMG src="http://www.genenetwork.org/images/upload/SNPBrowserUpdate.png" valign="top" width="450" height="242"></DIR>
<DIR><P><SMALL><B>Legend:</B> Example of the updated SNP browser; SNP search results for <I>Zim3</I></SMALL></P>
</DIR>

<BLOCKQUOTE><STRONG>2009-07-20: </STRONG>SymAtlas links in GeneNetwork has been replaced by BioGPS links. These new links provide summary data on expression of genes in 50 or more tissues in seveal species. BioGPS also provides data on expression QTLs. We thank Dr. Andrew Su and colleagues at the GNF. (Implemented by Xiaodong Zhou)</BLOCKQUOTE>

<BLOCKQUOTE>
<STRONG>2009-06-24: </STRONG>
GeneNetwork is integrated with the <A HREF="http://motif.memphis.edu/gcat/">Gene-set Cohesion Analysis Tool (GCAT)</A>. GCAT determines the functional coherence of gene sets by performing latent semantic analysis of Medline abstracts. To try this new function, select a set of genes/transcripts in one of the GeneNetwork Correlation Tables and then click on the GCAT button (upper left).  (Implementation by Xiaodong Zhou, Lijing Xu, Ramin Homayouni and colleagues at the University of Memphis)
</BLOCKQUOTE>


<P>
<DIR><IMG src="http://www.genenetwork.org/images/upload/GCAT.png" valign="top" width="360" height="365">
</DIR>
<DIR><P><SMALL><B>Legend:</B> GCAT output graph of genes associated with Comt </SMALL></P>
</DIR>



<BLOCKQUOTE><STRONG>2009-06-20: </STRONG>Expression data for the ventral tegmental area (VTA) of the midbrain has been added to the BXD panel of mouse strains. Three related data sets--saline control, acute ethanol, and an ethanol vs saline contrast S scores--were generated by Dr. Michael Miles and colleagues. All data sets are currently being tested. Contact Dr. Miles at VCU Medical Center for access  (Implemented by Arthur Centeno, Xiaodong Zhou, and Michael Miles)</BLOCKQUOTE>


<BLOCKQUOTE><STRONG>2009-06-14: </STRONG>We have upgraded the Network Graph (Association Network) functions in GN. You can now change color and thickness of lines (edges) that connect nodes (usually transcripts or phenotypes). Each node of these graphs is hyperlinked to the underlying data. (Implemented by Lei Yan)</BLOCKQUOTE>

<P>
<DIR><IMG src="http://www.genenetwork.org/images/upload/NetworkGraph2009.png" valign="top" width="656" height="376">
</DIR>

<DIR><P><SMALL><B>Legend:</B> Example of the type of network graph that can be generated using both expression and phenotype data sets. </SMALL></P>
</DIR>

<BLOCKQUOTE><STRONG>2009-06-05: </STRONG>Upgraded the annotation to the Affymetrix Rat RAE 230 and RAE 230 2.0 microarrays. We have also incorporated the probe sequence data for the newer array. Annotation for the rat Exon 1.0 ST array is also in progress. (Implemented by Xioadong Zhou and RWW)</BLOCKQUOTE>

<BLOCKQUOTE><STRONG>2009-04-28: </STRONG> The edit HTML function was changed from cgi technique to mod_python. Our goal is to eventually change all the few functions in GN that still use CGI technique to mod_python (Implementation by Xiaodong Zhou)
</BLOCKQUOTE>

<BLOCKQUOTE><STRONG>2009-04-28: </STRONG> In each probeset info file, one link was added (Accession number: GNxxx) so that user can download the raw data of this particular dataset.  (Implementation by Xiaodong Zhou)
</BLOCKQUOTE>

<BLOCKQUOTE><STRONG>2009-04-24: </STRONG>Information on polymorphisms in micro RNA targets taken from the  <A HREF="http://compbio.uthsc.edu/miRSNP/">PolymiRTS Database</A> have been added to interval map tables in GeneNetwork. (Implementation by Lei Yan)
</BLOCKQUOTE>

<BLOCKQUOTE><STRONG>2009-04-15: </STRONG><B>Microarray Annotation Files</B>. We have begun to make all of the manually curated array annotation files used in GeneNetwork available <a href="http://www.genenetwork.org/share/annotations" class="fs14" target="_blank">here</a>. (Implementation by Lei Yan, Arthur Centeno, Xusheng Wang, and Xiaodong Zhou)
</BLOCKQUOTE>


<BLOCKQUOTE><STRONG>2009-04-15: </STRONG><B>GN genotype data improvement</B>. We have begun to revise all genotype database tables and data sets in anticipation of adding large mouse, rat, and human genotype data sets in the next few months. We have reconciled discrepancies between physical and genetic maps for all species and developed a new method to produce special files used for interval mapping (so-called "geno" files). (Implementation by Xiaodong Zhou and Rob Williams)
</BLOCKQUOTE>

<BLOCKQUOTE><STRONG>2009-04-15: </STRONG><B>GeneNetwork Database Improvement</B>. All datga (phenotype, genotype, and expression data) in GeneNetwork have previously been entered into a single massive table (data table) in fully normal form. To improve performance we have now split quantitative data for individuals, cases, and strains into four related tables: 1. PublishData (classic phenotype data), 2. GenoData (genotypes),  3. ProbeSetData (gene expression data), and 4. ProbeData (individual probe expression data). The reason to do this is that these data types are used and updated in different ways. An original table that held all error term data (the SE table) has also been separated in the same way. Our intent is to improve SQL query performance and reduce recovery time. (Implemented by Xiaodong Zhou)
</BLOCKQUOTE>

<BLOCKQUOTE><STRONG>2009-04-07: </STRONG><B>Usage <A HREF="http://lily.uthsc.edu/awstats/awstats.pl?config=lily.utmem.edu" class="fs14"  target="_empty">Statistics</a> for Lily</B>, one of the main GeneNetwork server are now available. GeneNetwork runs on a set of servers and we occasionally rotate the production server and staging server (currently named <A HREF="http://lily.uthsc.edu/awstats/awstats.pl?config=lily.utmem.edu" class="fs14"  target="_empty">Statistics</a>Lily and <A HREF="http://proust.uthsc.edu/awstats/awstats.pl?config=proust.utmem.edu" class="fs14"  target="_empty">Statistics</a>Proust). (Implemented by Lei Yan)</BLOCKQUOTE>


<BLOCKQUOTE><STRONG>2009-02-20: </STRONG>We are making some GUI interface improvements to the <A HREF="http://www.genenetwork.org/dbResults.html">Genome Graph</A> interface. (Implemented by Lei Yan)
</BLOCKQUOTE>


<BLOCKQUOTE><STRONG>2009-02-20: </STRONG><B>Trait Correlation Upgrade</B>. The most important innovation from the user's point-of-view is a fast method to compute correlations among transcripts that reduces the response latency about 20-fold--from 90 seconds to under 5 seconds using standard Affymetrix and Illumina data sets consisting of 45000 probes or probe sets. Of even greater importance, this improvement means that GN can now handle massive Affymetrix Exon 1.0 ST data sets that have over 1.1 million probe sets in about 60 seconds. The new method exploits a set of text files that are external to the database (essentially a <A HREF="http://en.wikipedia.org/wiki/Materialized_view" class="fs14"  target="_empty">materialized view</A>), parallel computing technique (<A HREF="http://www.parallelpython.com/" class="fs14"  target="_empty">Parallel Python</A>) and optimized SQL queries. (Implementation by Xiaodong Zhou and David Crowell).
</BLOCKQUOTE>

<BLOCKQUOTE><STRONG>2009-02-18: </STRONG><B>Data Security Upgrade</B>. New data security system has been active since Feb 18, 2009. You only have access to the confidential dataset to which you are assigned. If you have any questions, please contact Dr. Robert Williams or Xiaodong Zhou. (Implementation by Xiaodong Zhou and Hongqiang Li).
</BLOCKQUOTE>

<BLOCKQUOTE><STRONG>2009-2-13: </STRONG>Add link to <A HREF="http://ontologicaldiscovery.org/index.php">Ontological Discovery Environment</A> (Implemented by Lei Yan)
</BLOCKQUOTE>

<BLOCKQUOTE><STRONG>2009-01-20: </STRONG><B>Human and Rat Trait Correlation pages</B> have been upgraded to included data on tissue correlations and literature correlations. (Implementation by Xiaodong Zhou).
</BLOCKQUOTE>

<BLOCKQUOTE><STRONG>2009-01-16: </STRONG><B>Advanced Search GUI</B> The GUI for advanced search function has been developed. This <A HREF="/index3.html">GUI</A> is much more user friendly than the old text interface (Implementation by Lei Yan).
</BLOCKQUOTE>

<BLOCKQUOTE><STRONG>2009-01-08: </STRONG><B>Link to WebGestalt 2.0</B> The links in GN to WebGestalt have been switched to its new edition <A HREF="http://bioinfo.vanderbilt.edu/wg2/">WebGestalt 2.0</A> (Implementation by Lei Yan).
</BLOCKQUOTE>

<BLOCKQUOTE><STRONG>2008-12-30: </STRONG>The UML class diagrams for GN python code have been finished. The inheritance among classes and dependency among modules are also represented by UML diagrams. (Implemented by Xiaodong) 
</BLOCKQUOTE>

<BLOCKQUOTE><STRONG>2008-12-30: </STRONG>The UML database diagram for GN database has been finished (Implemented by Hongqiang) 
</BLOCKQUOTE>

<BLOCKQUOTE><STRONG>2008-12-23: </STRONG><B>Heritability</B> values have been added to the BXD Eye HEIMED mRNA data set. These data are shown in the Basic Statistics page for this particular data set. We hope to routinely add this type of information for new array data as they are added to GeneNetwork. (Implementation by Lei Yan and Xusheng Wang).
</BLOCKQUOTE>

<BLOCKQUOTE><STRONG>2008-12-19: </STRONG><A HREF="http://alexandria.uthsc.edu:82/dbdownload/download.jsp" class="fs14"  target="_empty">Array Annotation Query, Download, and Annotate </A> module has been added to GeneNetwork. (Implementation by Lei Yan and Xusheng Wang).
</BLOCKQUOTE>


<BLOCKQUOTE><STRONG>2008-12-18: </STRONG>New rat HXB/BXH gene expression data from Norbert Hubner and colleagues is being entered into GeneNetwork. The following four tissue types are being added: adrenal glands, soleus muscle, heart, and liver. These data are still experimental but will be opened in 2009.
</BLOCKQUOTE>


<BLOCKQUOTE><STRONG>2008-12-4: </STRONG><A HREF="http://www.genenetwork.org/glossary.html#tissueCorrelation">Tissue Correlation</A> scores for the great majority of genes have been added to GeneNetwork. A high tissue correlation between two genes indicates that they tend to be expressed together across a set of diverse tissues and organs. Tissue Correlations can be computed from most Trait Data and Analysis pages by selecting <I>Tissue Correlation, Pearson's r</I> or <I>Tissue Correlation, Spearman's rho</I> using the <B>Trait Correlation</B> analysis tool.  (Implementation by Xiaodong Zhou and Xusheng Wang) 
</BLOCKQUOTE>



<BLOCKQUOTE><STRONG>2008-12-4: </STRONG>University of Western Australia GN mirror site is up and running in the laboratory of Dr. Grant Morahan. This system built and maintained by Munish Mehta. Munish also has an active code development program with novel tools and resources. 
</BLOCKQUOTE>

<BLOCKQUOTE><STRONG>2008-12-3: </STRONG>The JAX Mouse Diversity Genotyping array is being used to regenotype the BXD strains. This array was designed by Fernando Pard-Manuel de Villena and Gary Churchill. The array generates as many as 625,000 SNPs and 900,000 invariant genomic probes (one SNP about every 4.3 kb, see JAX Notes, Winter 2008, No 512). The genotypes be integrated into GeneNetwork in the next three months. (Implementation for GN by Lu Lu, R. Williams, and Hongqiang Li).
</BLOCKQUOTE>


<BLOCKQUOTE><STRONG>2008-11-28: </STRONG>Gene-set Cohesion Analysis Tool has been integrated into GN. <A HREF="http://motif.memphis.edu/gcat/">GCAT Home</A> (Implemented by Xiaodong Zhou)
</BLOCKQUOTE>

<BLOCKQUOTE><STRONG>2008-11-10: </STRONG>One new server <A HREF="http://lily.uthsc.edu/">Lily</A> has been installed and configured as GN main server.  (Implemented by Lei Yan)
</BLOCKQUOTE>

<BLOCKQUOTE><STRONG>2008-10-20: </STRONG>First human expression data set for lymphoblasts (Epstein-Bar virus immortalized B-cells) from the CEPH panel (large Mormon families) have been integrated into GeneNetwork. The data are original from a study by Monks and colleagues (<A HREF="http://www.pubmedcentral.nih.gov/articlerender.fcgi?tool=pubmed&pubmedid=15514893" class="fs14"  target="_empty">2004</A>). <B>Please note</B>: These data can currently be used to study patterns of coexpression among transcripts, but we have not yet implemented mapping algorithms and do not expect to make interval mapping available until summer 2009. We thank Stephanie Santorico for providing her data and help.  (Implemented by Hongqiang Li, Stephanie Monks Santorico, Xusheng Wang, and Arthur Centeno.)
</BLOCKQUOTE>

<BLOCKQUOTE><STRONG>2008-10-15: </STRONG>We have uploaded four LXS expression data sets for the mouse hippocampus (NOS, NOE, RSS, and RSE; n = 33 strains per data set generated using Illumina Mouse WG-6.1 Beadarray). The four sets are part of a study on the effects of ethanol and stress on brain gene expression. Please see the INFO files associated with these data sets for more background, for example, the "no restraint-saline" condition (<A HREF="http://www.genenetwork.org/dbdoc/Illum_LXS_Hipp_NOS_1008.html" class="fs14"  target="_empty">NOS</A>).  (Implemented by Lu Lu and Arthur Centeno.)
</BLOCKQUOTE>

<BLOCKQUOTE><STRONG>2008-10-05: </STRONG>Major new BXD behavioral and drug-related phenotypes for 62 strains have been integrated into GeneNetwork by a large research consortium funded by the National Institute on Drug Abuse and the National Institute for Alcohol Abuse and Alcoholism. A total of 242 phenotypes were measured in collaboration with the Systems Genetics Group at the Oak Ridge National Laboratory. These records can be searched by selecting the BXD Phenotypes data set and entering the search string "Chesler." <B>Please request permission</B> from Drs. EJ Chesler and D Goldowitz to use these data. Primary phenotypic data on individual mice are available in the <A HREF="http://mouse.ornl.gov"  class="fs14"  target="_empty">Mouse Track System</A>. (Implemented by Elissa Chesler (cheslerej@ornl.gov), Philip VM, Ansah TA, Blaha CD, Cook MN, Hamre KM, Lariviere WR, Matthews DB, Mittleman G, Goldowitz D)
</BLOCKQUOTE>

<BLOCKQUOTE><STRONG>2008-10-04: </STRONG><A HREF="http://gn.genetics.ucla.edu/" class="fs14"  target="_empty">UCLA Department of Genetics GeneNetwork mirror site</A> is now online. Data in this site is synchronized periodically (once per month) with the Tennessee production site. (Implemented by Evan G. Williams and Kev Adler.) 
</BLOCKQUOTE>

<BLOCKQUOTE><STRONG>2008-10-05: </STRONG><A HREF="ADD_LINK_HERE"  class="fs14"  target="_empty">An XML schema</A> for expression genetic data sets of the type used by GeneNetwork has been generated by Ilze Druka and colleagues. This schema is based in part on the GeneNetwork database.  (Implemented by Ilze Druka)
</BLOCKQUOTE>

<BLOCKQUOTE><STRONG>2008-9-15: </STRONG><B>GN code improvement by cleaning up "namespace pollution"</B> Every GN python module used to import other module by "from SomeModule import *" so that the class, function and variable of the imported module can be used without the imported module name as prefix. Since one typical GN python module imports over ten modules, and GN python modules always import other GN python modules like chain (B import A, C import B, D import C ...), the namespace is seriously polluted. It not only makes the source code very difficult to read (for anything not defined in current module, programmer may have look up dozens of imported modules to find where it is defined), but also makes the dependency among modules unclear, so the change to one module always has "unpredictable" effects to other modules. As GN software keeps growing, the namespace pollution problem keeps building up and make it harder and harder to maintain the software and develop new features. A lot of effort has been made to change all GN python module to import module by "import SomeModule", and add the module's name as prefix to the the class, function and variable of the imported module (thousands of places). This work not only greatly improves the source code readability, but also make the dependency among GN python modules clear, hence much easier for maintenance and future development.  (Implemented by Xiaodong Zhou)
</BLOCKQUOTE>

<BLOCKQUOTE><STRONG>2008-07-29: </STRONG>GeneNetwork <A HREF="http://rhodes.uthsc.edu/bin/view/GeneNetwork"  class="fs14"  target="_empty">TWiki</A> site has been moved to a new server.  (Implemented by Kev Adler)
<BR>GeneNetwork <A HREF="http://nb.uthsc.edu:8080"  class="fs14"  target="_empty">Roundup</A> bug tracking site has been moved to a new server. (Implemented by Hongqiang Li)
<BR>
</BLOCKQUOTE>

<BLOCKQUOTE><STRONG>2008-07-30: </STRONG>Massive UCLA microarray data sets for four F2 populations and four tissues have been integrated into GeneNetwork. Data are now available for up to 300 animals for each of these tissues: 
<OL>
<LI>Adipose tissue (see <A HREF="http://www.ncbi.nlm.nih.gov/pubmed/17641228?" class="fs14"  target="_empty">2007</A>)
<LI>Brain (see <A HREF="http://www.ncbi.nlm.nih.gov/pubmed/16539680?" class="fs14" target="_empty">2006</A>)
<LI>Liver (see <A HREF="http://www.ncbi.nlm.nih.gov/pubmed/17641228?" class="fs14" target="_empty">2007</A>)
<LI>Muscle (see <A HREF="http://www.ncbi.nlm.nih.gov/pubmed/20439777" class="fs14" target="_empty">2010</A> and <A HREF="http://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE12795">GEO series GSE12795</A>)
</OL>
Data sets were generated by Jake Lusis, Eric Schadt, and colleagues and include one of the first expression genetics study published in <A HREF="http://www.ncbi.nlm.nih.gov/pubmed/12646919" class="fs14"  target="_empty">2003</A> (Mouse BDF2 UCLA). Papers that describe several of these new data sets have been published (Yang et al. <A HREF="http://genome.cshlp.org/cgi/content/abstract/16/8/995" class="fs14"  target="_empty">2006</A>; Schadt et al., <A HREF="http://biology.plosjournals.org/perlserv/?request=get-document&doi=10.1371/journal.pbio.0060107" class="fs14"  target="_empty">2008</A>) and several massive data sets are open access (Mouse BHF2-Apoe UCLA, from GEO GSE2814, GSE3086, GSE3087, and GSE3088). Several other data sets are still being analyzed. For early access to still unpublished data sets (e.g., BH/HB F2 UCLA and CastB6/B6CastF2 UCLA) please contact <A HREF="http://www.genetics.ucla.edu/labs/lusis/" class="fs14"  target="_empty">Dr. Aldons Jake Lusis</A> and colleagues. (Implemented by Evan Williams)
<BR>
</BLOCKQUOTE>

<BLOCKQUOTE><STRONG>2008-08-29: </STRONG>EYE data: The Mouse HEIMED whole eye gene expression data set has been extended to 99 strains, including 27 common inbred strains. With support from Dr. Barrett Haik, Eldon Geisert, Lu Lu and colleagues we have added data for 15 new strains of mice. Arrays were processed at the UTHSC by Weikuan Gu and colleagues. This data set is open for use without a password.  (Implemented by RWW, Daniel Ciobanu, Lu Lu, and Arthur Centeno)
<BR>
</BLOCKQUOTE>


<BLOCKQUOTE><STRONG>2008-07-15: </STRONG>The GeneNetwork MySQL configurations have been optimized, resulting in much faster and more reliable service. (Implemented by Kev Adler)
<BR>
</BLOCKQUOTE>

<BLOCKQUOTE><STRONG>2008-06-20: </STRONG>GeneNetwork is now part of the NIH NCRR Biomedical Informatics Research Network (BIRN) with a fully configured BIRN equipment rack. Our thanks to BIRN-Coordinating Center and to the IT staff at UTHCS (Implemented by Bao Nguyen, Mark James, James Martin, Billy Hatcher, and Kev Adler)
<BR>
</BLOCKQUOTE>


<BLOCKQUOTE><STRONG>2008-05-20: </STRONG>Neocortex: With support from the High Q Foundation, we have added a matched neocortex (cerebral cortex) gene expression data set for 73 strains of mice to accompany the striatum data set highlighted below. The neocortex data set includes estimates of expression for 20 common inbred strains, 52 BXD strains, B6D2F1 hybrids generated using the Illumina Mouse-6 v1.1 Sentrix array. Samples were generated by Glenn Rosen, Lu Lu, and colleagues. Arrays were processed at the UTHSC. This data set is open for use without a password.  (Implemented by HS, RWW, Lu Lu, and Arthur Centeno)
<BR>
</BLOCKQUOTE>

<BLOCKQUOTE><STRONG>2008-05-20: </STRONG>Striatum: With support from the High Q Foundation, we have added a striatum (caudate-putamen) gene expression data set for 75 strains of mice. The data set includes estimates of expression for 20 common inbred strains, 54 BXD strains, B6D2F1 hybrids generated using the Illumina Mouse-6 v1.1 Sentrix array. Samples were generated by Glenn Rosen, Lu Lu, and colleagues. Arrays were processed at the UTHSC. This data set is open for use without a password.  (Implemented by HS, RWW, Lu Lu, and Arthur Centeno)
<BR>
</BLOCKQUOTE>

<BLOCKQUOTE><STRONG>2008-04-20: </STRONG>Lung:  We have added lung gene expression data set for 57 strains of mice. The data set includes estimates of expression for 8 common inbred strains, 47 BXD strains, and reciprocal F1 hybrids (B6D2F1 and D2B6F1) generated using the M430 2.0 Affymetrix array. Samples were generated by Lu Lu and colleagues. Arrays were processed by Yan Jiao and Weikuan Gu at the Memphis VA. This data set is still provisional and not available without a password. If you would like early access, please contact Prof. Klaus Schughart (Helmholtz Centre for Infection Research, Braunschweig, Germany) at kls@helmholtz-hzi.de. (Implemented by HS, RWW, Lu Lu, and Arthur Centeno)
<BR>
</BLOCKQUOTE>


<BLOCKQUOTE><STRONG>2008-04-10: </STRONG>Nucleus Accumbens: <A Href="http://www.brainchip.vcu.edu/" class="fs14"  target="_empty">Dr. Michael Miles</A> and colleagues have added gene expression data for the nucleus accumbens of the BXD strains into GN. The nucleus accumbens is an important part of the brain involved in emotional state and reward. It is also critically involved in drug abuse and alcoholism. Three complementary data sets have been submitted: expression in nucleus accumbens following a saline control injection, expression in nucleus accumbens following an injection of ethanol, and a data set that highlights the difference in expression between the two conditions (saline and ethanol). Each data set includes estimates of expression for 35 BXD strains, as well as C57BL/6J and DBA/2J. Samples were generated and processed by Miles and colleagues. The saline control data set is available without a password, but the two other data sets are still available only with a password. If you would like early access, please contact Prof. Michael F. Miles (Virginia Commonwealth University) at mfmiles@vcu.edu. (Implemented by MM and Arthur Centeno.)
<BR>
</BLOCKQUOTE>


<BLOCKQUOTE><STRONG>2008-04-10: </STRONG>Prefrontal Cortex: <A Href="http://www.brainchip.vcu.edu/" class="fs14"  target="_empty">Dr. Michael Miles</A> and colleagues have added gene expression data for the <A HREF="http://www.genenetwork.org/dbdoc/VCUSal_1006_R.html" class="fs14"  target="_empty">prefrontal cortex</A> of the BXD strains into GN. The prefrontal cortex (PFC, or prelimbic neocortex) is an important part of the brain involved in emotional state and reward. It is also critically involved in drug abuse and alcoholism. Three complementary data sets have been submitted: expression in PFC following a saline control injection, expression in nucleus accumbens following an injection of ethanol, and a data set that highlights the difference in expression between the two conditions (saline and ethanol). Each data set includes estimates of expression for 35 BXD strains, as well as C57BL/6J and DBA/2J. Samples were generated and processed by Miles and colleagues. The saline control data set is available without a password, but the two other data sets are still available only with a password. If you would like early access, please contact Prof. Michael F. Miles (Virginia Commonwealth University) at mfmiles@vcu.edu. (Implemented by MM and Arthur Centeno.)
<BR>
</BLOCKQUOTE>


<BLOCKQUOTE><STRONG>2008-03-26: </STRONG> Dr. Fan Zhang has left the GN software development group. He has spent the past five months substantially reworking the architecture of GN hardware, rewriting the code to increase is portability, and in improving security. Thanks Fan for your many contributions and good luck back home in China.
<BR>
</BLOCKQUOTE>


<BLOCKQUOTE><STRONG>2008-03-17: </STRONG> Two Affymetrix Mouse EXON ST 1.0 array data sets for the hippocampus (n = 84 strains) and striatum (n = 48 strains), including a variety of common inbred strains and numerous BXD strains have being integrated this week into GeneNetwork (see Mouse--BXD--Hippocampus and Striatum). These data were generated with the support of Affymetrix, Dr. David Kulp, and the High Q Foundation. Annotation on these two large data sets is in progress. (Implemented by Manjunatha Jagalur, Arthur Centeno, Xusheng Wang, Lu Lu, and Hongqiang Li).
<BR>
</BLOCKQUOTE>

<BLOCKQUOTE><STRONG>2008-03-14: </STRONG> GeneNetwork site has been moved to two new Dell PowerEdge 1950  and 2950 8-core servers to provide more capacity and higher performance. (Implemented by Fan Zhang).
<BR>
</BLOCKQUOTE>

<BLOCKQUOTE><STRONG>2008-03-2: </STRONG> GeneNetwork site usage is now being studied using Google Analytics. We hope this will allow us to improve usage and performance for GN clients.  (Implemented by Fan Zhang).
<BR>
</BLOCKQUOTE>

<BLOCKQUOTE><STRONG>2008-02-17: </STRONG> We have finished synchronizing the updated GN code between the UTHSC GN main site and the <A HREF="http://genenetwork.helmholtz-hzi.de/" class="fs14"  target="_empty">HZI GeneNetwork</A> mirror. This is still a manual process that needs to be done once every few months. (Implemented by Fan Zhang and Rudi Albert).
<BR>

</BLOCKQUOTE>
<BLOCKQUOTE><STRONG>2008-02-14: </STRONG> We have implemented a system that improves the synchronization of  the main production database and with the set of servers that are part of the cluster. These servers are referred to by our development team as GN server "bundles". (Implemented by Fan Zhang and Kev Adler).
<BR>
</BLOCKQUOTE>

<BLOCKQUOTE><STRONG>2008-01-20: </STRONG>  GeneNetwork consists of a cluster of servers that have identical software code and identical databases. The creation of the individual servers (bundles) that are part of the GN cluster is quite complicated. Each bundle consists of an application stack (Linux CENT OS 5, MySQL 5.0, Apache HTTP server, Mod_Python, and many code libraries and scripts). Components of the bundles ideally work well regardless of the physical hardware and network configuration. We have now rewritten and annotated the GN code and reconfigured bundles to make them as independent as possible from their particular network situation. We refer to this effort as "de-hardcoding" or refactoring GN. GN bundles are now easy to be deployed with minimum manual work. The following tasks still need to be done: configuration of IP addresses, specification of absolute directory structure as DocumentRoot for Apache, recompilation of certain third-party Python libraries according to hardware infrastructure (i386 vs X64_86).(Implemented by Fan Zhang). The number of bundle servers has been increased to six. 
<BR>

<A HREF="http://webqtl.utmem.edu:81/bin/view/GeneNetwork/GNBundleSetup"  class="fs14"  target="_empty">GN Bundle Setup</A>
<BR>
<A HREF="http://webqtl.utmem.edu:81/bin/view/GeneNetwork/GNDeHardCoding"  class="fs14"  target="_empty">De-hardcoding Procedures</A>
</BLOCKQUOTE>
 
<BLOCKQUOTE><STRONG>2008-01-12: </STRONG>Probe level data has been added into the <A HREF="dbdoc/BR_M2_1106_R.html" class="fs14"  target="_empty">UCHSC BXD Whole Brain M430 2.0 November 2006 RMA</A> data set. Affymetrix data sets in GeneNetwork always include the higher level "probe set" data, but in the case of this important whole brain data set from The University of Colorado we have also entered all of the individual probe level data. This means that users can drill down to examine the expression of individual 25-mer probes. To access the individual probe data please click on the PROBE TOOL button on the Trait Data and Analysis page of GN. Then click on any individual probe; the left hand column. (Implemented by Arthur Centeno, Hongqiang Li, and Daniel Ciobanu.)  
<BR>
</BLOCKQUOTE>

<BLOCKQUOTE><STRONG>2008-01-08: </STRONG>We have added a large <A HREF="dbdoc/UTHSC_1107_RankInv.html" class="fs14"  target="_empty">Mouse Striatum Expression </A> data set into GeneNetwork. This data set includes replicate estimates of expression in the dorsal striatum for 75 strains of mice (54 BXDs and 21 common strains). All data were generated using the new Illumina Mouse 6.1 bead array with support from the High Q Foundation and the NIH. A matched neocortical expression data set will be uploaded into GN in the next few months. (Implemented by Glenn Rosen, Lu Lu, Arthur Centeno, Hongqiang Li, and Rob Williams.)  
<BR>
</BLOCKQUOTE>

<P>
<DIR><IMG src="/images/upload/HQF_Drd1a_Striatum.gif" valign="top" width="532" height="250">

</DIR>

<DIR><P><SMALL><B>Legend:</B> Expression of the dopamine D1a receptor mRNA in the striatum from the new Illumina data set. Each bar provides expression values (log2 transformed values) for a single strain. </SMALL></P>
</DIR>

<BLOCKQUOTE><STRONG>2007-12-08: </STRONG>John Stuart and colleagues have uploaded a <A HREF="dbdoc/UTHSC_1107_RankInv.html" class="fs14"  target="_empty">Spleen M430 2.0 Expression </A> data set into GeneNetwork for the CXB recombinant inbred strains and the two parents of this RI set 9C57BL/6By and BALB/cBy). This data set includes replicate estimates of expression for the whole spleen. (Implemented by Lu Lu, Arthur Centeno, Hongqiang Li, and Rob Williams.)  
<BR>
</BLOCKQUOTE>

<BLOCKQUOTE><STRONG>2007-10-20: </STRONG>In progress: GeneNetwork developers are working closely with the Biomedical Informatics Research Network to map GN data sets and metadata into <A HREF="http://xwiki.nbirn.net:8080/xwiki/bin/view/BIRN-OTF/SemiAutomated+Term+to+BIRNLex+Mapping" class="fs14"  target="_empty">OWL/RDF</A> concepts.  (Implemented by Rob Williams, Fan Zhang, and Bill Bug.)  
<BR>
</BLOCKQUOTE>

<BLOCKQUOTE><STRONG>2007-10-19: </STRONG>GeneNetwork codebase has been more throughly documented/annotated over the past several weeks by Kev Adler and Fan Zhang. Many important changes have been made to abstract the code and its calls to external resources (Implemented by Fan Zhang.)  
<BR>
</BLOCKQUOTE>

<BLOCKQUOTE><STRONG>2007-10-16: </STRONG> The performance of MySQL database as the data vault has been increased dramatically. The consumed time for running a standard correlation mapping has been decreased from 3-4 minutes down to less than 2 seconds. (Implemented by Fan Zhang.) <BR>
</BLOCKQUOTE>


<BLOCKQUOTE><STRONG>2007-10-16: </STRONG>A small Java application has been written to handle early steps in the processing and error-checking of microarray data sets that heading into GeneNetwork. The program is still a beta version and is available upon request from Hongqiang Li. It is called ArrayPipeliner. (Implemented by Hongqiang Li.)  
<BR>
</BLOCKQUOTE>


<BLOCKQUOTE><STRONG>2007-10-12: </STRONG>GeneNetwork has been moved to a new cluster. The GN MySQL relational database is now running on a Dell 2950. Our apologies for inconvenience associated with this move. (Implemented by Fan Zhang and Hongqiang Li.)  
<BR>
</BLOCKQUOTE>


<BLOCKQUOTE><STRONG>2007-10-24: </STRONG>A preliminary version <A HREF="http://webqtl.utmem.edu:81/bin/view/GeneNetwork/UploadingDatasetsIntoMySQL" class="fs14"  target="_empty">Standard Operating Procedures</A> for entering new microarray data sets into GeneNetwork are now available on the GeneNetwork TWiki pages (Implemented by Arthur Centeno, Hongqiang Li, and Rob Wiliams.)  
<BR>
</BLOCKQUOTE>


<BLOCKQUOTE><STRONG>2007-10-02: </STRONG>The Illumina mouse microarray annotation file has been substantially updated and extended by Xusheng Wang, Hongqiang Li, and Rob Williams. The new annotation file covers four variant arrays: Mouse 6, Mouse 6.1, Mouse 8, and Mouse 8.1.  This annotation file is used by the Mouse LXS Hippocampus data (Mouse 6) and by three new Mouse BXD Hematopoietic cell data sets (Mouse 6.1) generated by Gerald de Haan's and colleagues. The Illumina Annotation files are available for users of GeneNetwork, but are not yet freely as a complete text file download. If you need the new annotation file rather than that provide by Illumina, please contact Robert W. Williams.  (Implemented by Xusheng Wang, Hongqiang Li, and Rob Wiliams.)  
<BR>
</BLOCKQUOTE>



<BLOCKQUOTE>
<STRONG>2007-07-27: </STRONG> The <A HREF="http://www.cs.utk.edu/~library/TechReports/Thesis/heinrich.pdf">SGO Literature Correlation</A> data generated by the <A HREF="https://shad.cs.utk.edu/sgo/sgo.html"  class="fs14"  target="_empty">Semantic Gene Organizer (SGO)</A> team has been updated. SGO is a patented algorithm based on <A HREF="http://en.wikipedia.org/wiki/Latent_semantic_analysis"  class="fs14"  target="_empty"><I>latent semantic analysis</I></A> that provides a score of the terminological overlap between any two genes based on PubMed records and abstracts (M Berry, Kevin Heinrich, R Homayouni, and colleagues). These scores range from 0 to 1 (cosine similarity) and can be used like correlation coefficients, although all values are positive. GeneNetwork provides the Literature Correlations along with expression level correlations in most Correlation Table output pages. Roughly 5% of literature correlations have an r value greater than 0.6. The Literature Correlation covers roughly 75% of known genes in mouse, rat, and human.

<P>The new 2007 dataset covers 21,903 genes and was derived from a corpus consisting of 242,365 MEDLINE abstracts.  (The previous literature correlation dataset dated back to 2005 and was derived from a set of 108,367 abstracts covering 13,129 genes.) 

<P>We have also added a new <b>SGO Literature Correlation</b> feature to the <b>Trait Data and Analysis Form</b>.  You can now select any gene and find the top genes (100 to 2000) with which it shares common terms and context based on the literature. To try this feature select one of the <B>Calculate:</B> SGO Literature Correlation options in the <B>Trait Correlations</B> part of the Trait Data and Analysis form. Normally GeneNetwork computes correlations based on expression and then provides the literature correlation as a secondary data type. This new feature reverses the situation; now the literature correlation is primary and the expression data are given as a secondary data type. This is a quick and powerful way to determine whether expression data support particular correlations extracted from the literature.  (Implemented by Lijing Xu, <a href="http://whatmind.com/bio/">Nick Furlotte</a>, <a href="http://cas.memphis.edu/binf/Ramin_Homayouni.htm">Ramin Homayouni</a>.)  
<BR>
</BLOCKQUOTE>

<P>
<DIR><IMG src="/images/upload/LitCorSearch.png" valign="top" width="410" height="63">
</DIR>

<DIR><P><SMALL><B>Legend:</B> The Literature Correlation feature can be accessed from the Trait Correlation tool.</SMALL></P>
</DIR>



<BLOCKQUOTE>
<STRONG>2007-07-26: </STRONG>Affymetrix <A HREF="dbdoc/Striatum_Exon_0707.html"  class="fs14"  target="_empty">Mouse Exon 1.0 ST</A> data are now available for the first time in GeneNetwork for the striatum of 30 BXD strains and 20 standard inbred strains. These data were generated with the support of the <A HREF="http://www.highqfoundation.org/" class="fs14" target="_empty">High Q Foundation</A>. (Implemented by Hongqiang Li, Arthur Centeno, Manjunatha Jagalur.)  
<BR>
</BLOCKQUOTE>

<P>
<DIR><IMG src="/images/upload/ExonStriatumSearch.png" valign="top">
</DIR>

<DIR><P><SMALL><B>Legend:</B> Access to the Mouse Exon 1.0 ST Affymetrix expression data set from the <A HREF="dbdoc/Striatum_Exon_0707.html  class="fs14" target="_empty">High Q Foundation</A>.  </SMALL></P>
</DIR>


<BLOCKQUOTE><STRONG>2007-06-21: </STRONG>A European GeneNetwork <A HREF="http://genenetwork.helmholtz-hzi.de/"  class="fs14" target="_empty">mirror site http://genesys.helmholtz-hzi.de</A> has been established at <A HREF="http://www.helmholtz-hzi.de/en/" class="fs14" target="_empty">The Helmholtz Zentrum f&uuml;r Infektionsforschung</A> in Braunschweig, Germany. A <A HREF="http://genesys.helmholtz-hzi.de/search.html" target="_empty">GeNeSys private site</A> will operate as part of the German Network for Systems Genetics (<A HREF="http://www.helmholtz-hzi.de/en/research_groups/experimental_mouse_genetics/experimental_mousegenetics/projects/genesys/" class="fs14" target="_empty">GeNeSys</A>). (Implemented by Evan G. Williams and Klaus Schughart, <A HREF="http://www.helmholtz-hzi.de/en/research_groups/experimental_mouse_genetics/experimental_mousegenetics/" class="fs14" target="_empty">Experimental Mouse Genetics</A>.)  
<BR>

</BLOCKQUOTE>



<BLOCKQUOTE><STRONG>2007-04-26: </STRONG>Zhaohui Sun, one of our lead programmers is moving to Dallas with his family. As a quick review of these recent News items will show, Zhaohui has made many important contributions to GeneNetwork over the past year. Thanks Zhaohui. We will miss you. 
<BR>
</BLOCKQUOTE>

<BLOCKQUOTE><STRONG>2007-04-26: </STRONG>The <A HREF="http://www.genenetwork.org/webqtl/snpBrowser.py"  class="fs14" target="_empty">SNP Browser</A> that is integrated into GN has been further improved. It is now possible to filter SNPs by domain (exon, intron, etc.) and by function (e.g., mis-sense, silent). The updated version also includes a wider variety of strains of mice. (Implemented by Zhaohui Sun.)  
<BR>
</BLOCKQUOTE>

<BLOCKQUOTE><STRONG>2007-04-25: </STRONG>GeneNetwork's <A HREF="http://www.genenetwork.org/CGIDoc.html" class="fs14" target="_empty">Scriptable Interface</A> is now being extended at the request of Dr. Graeme Wistow of the National Eye Institute NEIBank so that external users and bioinformatics site developers can directly access the "best" data for particular genes and transcripts from specific data sets. 

<P>For example, it is now possible to link directly to the <B>Basic Statistics</B> data page for the expression of the rhodopsin gene in the eyes of BXD mouse strains using a string that has this form:

<DIR>
"http://www.genenetwork.org/webqtl/main.py?FormID=showBest&gene=Rho&database=Eye_M2_0906_R" (no line break)
<BR>

<BR>where "gene=Rho" can be replaced by "refseq=NM_145383" to retrieve data using the RefSeq identifier (do not use the decimal point and digit that may follow, such as "refseq=NM_145383.1"), or can be replaced by "geneid=212541" to retrieve data using the Entrez Gene identifier. 

<BR><BR>Although not recommended, the string "&searchAlias=1" can be added at the end of the command to retrieve data using the alias of a proper gene symbol when the original does not work. Thus "gene=RP4" will resolve to "gene=Rho" only if you add "&searchAlias=1" at the end of the command.
</DIR>

<P>When a particular gene is associated with more than one probe or probe set, our code selects <B>the single probe or probe set</B> with the highest average expression in that particular data set. For example, there are four probe sets that traget different parts of rhodopsin mRNAs. Using the Affymetrix M430 2.0 array, probe set 1425172_at has the highest expression in the eyes of BXD strains (the probes target the last two exons and the proximal 3' UTR).

<P>To choose a particular data set one needs to know the appropriate code, such as Eye-M2_0906_R in the example above. These codes can currently be obtained from:

<DIR>
"http://www.genenetwork.org/webqtl/main.py?cmd=help&parse=machine"
</DIR>

<P>This implementation will be moved from our beta site to the production GN site in early May, 2007. (Implemented by Hongqiang Li.)  


<BR>
</BLOCKQUOTE>


<BLOCKQUOTE><STRONG>2007-03-23: </STRONG>We have annotated Illumina's Sentrix Mouse-6 1.0 microarray BeadArray platform (see Mouse LXS Hippocampus data sets). We have added or corrected gene assignments, symbols, and gene descriptions for almost all of the 46,166 Illumina probes. We added many data types not provided by Illumina and the MEEBO consortium in their original annotation files (http://www.microarray.org/data/download/MEEBO_Data.txt), including updated Entrez Gene IDs, gene ontology categories, human orthologs, OMIM identifiers, and the precise locations of the 50-mer probe sequences on the most recent mouse genome assembly (Feb 2006, mm8). Some helpful metrics on this annotation: For 46166 probes on the Mouse 6 array platform (including control probes) we have identified 36687 NCBI Entrez Gene IDs; 26180 matched human Gene IDs; 23899 matched rat Gene IDs; 26882 NCBI HomoloGene IDs; and 12790 OMIM IDs. Position data for the 50-mer Illumina Mouse-6 array were BLAT aligned to the latest mouse genome assembly by Hongqiang Li. Many of the probes and alignments have been error-checked to a limited extent by RWW. Annotation is still continuing and we will be adding new data over the next several months. (Curation by Robert W. Williams, implemented into GeneNetwork by Hongqiang Li.)  
<BR>
</BLOCKQUOTE>

<BLOCKQUOTE><STRONG>2007-03-02: </STRONG>The new production site of Genenetwork has been built from the code base maintained in Subversion. Subversion has completely been incorporated to our software development practice. Previously we have built a development / demo site - <a href="http://web2qtl.utmem.edu" class="fs14" target="_empty"> http://web2qtl.utmem.edu</a> from Subversion, which will be our official beta site from now on. The previous beta site has been merged into our current production (main) site. The users can still access the old main and the old beta site through the archive: <a href="http://www.genenetwork.org:82/search.html" class="fs14">http://www.genenetwork.org:82/search.html</a>. This archive site still uses the position data from the older assembly of the mouse genome (UCSC mm6) so it allows the users to retrieve old results. (Implemented by Zhaohui Sun.)  
<BR>
</BLOCKQUOTE>

<BLOCKQUOTE><STRONG>2007-02-21: </STRONG>The <A HREF="http://www.genenetwork.org/webqtl/snpBrowser.py"  class="fs14" target="_empty">SNP Browser</A> that is integrated into GN has been updated to include all Perlegen/NIEHS SNPs, all Celera SNPs, Wellcome Trust-CTC SNPs, and the <A HREF="http://www.broad.mit.edu/personal/claire/MouseHapMap/Inbred.htm"  class="fs14" target="_empty">Mouse Haplotype Map SNPs</A>. We will continue to improve the annotation and strain of coverage  of these SNPs, but you will find that you can already download +/- 100 nt around the SNP or automatically BLAT SNPs to the UCSC Genome Browser. It is also now possible to search for SNPs by some of their identifiers. We have intentionally left in duplicate SNPs to allow ou to evaluate data consistency. The "Gap" value will be 0 for redundant SNPs. (Implemented by Zhaohui Sun.)  
<BR>
</BLOCKQUOTE>


<BLOCKQUOTE><STRONG>2006-11-28: </STRONG>We finished converting all mouse genomic and genetic data sets to the latest mouse genome sequence assembly (mm8). Prior to this day all mouse sequence data has used position data from the March 2005 mm6 assembly (equivalent to NCBI Build 34). Currently we are using all position data of mm8, equivalent to NCBI Mouse Build 36 and Ensembl Mus musculus version 40.36a, released in February 2006. (Implemented by Zhaohui Sun, Evan Williams and Rob Williams.)  
<BR>
</BLOCKQUOTE>

<BLOCKQUOTE><STRONG>2006-10-30: </STRONG>LXS Hippocampus Illumina Mouse-6 Sentrix array data (77 strains of mice represented by male and female pools, 43,514 probe sequences) have been uploaded into Beta site for testing.  (Implemented by Lu Lu and Hongqiang Li.)  
<BR/>
</BLOCKQUOTE>


<BLOCKQUOTE><STRONG>2006-09-16: </STRONG>The code base for GeneNetwork, WebQTL, and many other code modules used by GN have been successfully moved into a source code revision management system called Subversion.  This will greatly improve the GN software team's ability to maintain the code, develop new applications, and to collaborate with other research groups to expand GN.  This work provides a solid foundation for the scalability and portability of genenetwork. As part of this effort, we have set up a new beta site that is currently the official "product" of the Subversion tree at <a href="http://web2qtl.utmem.edu" class="fs14"> http://web2qtl.utmem.edu</a>.  (Implemented by Zhaohui Sun and Hongqiang Li.)  
<BR/>
</BLOCKQUOTE>


<BLOCKQUOTE><STRONG>2006-08-21: </STRONG>GeneNetwork TWiki Code and Hardward documentation site has been set up at <a href="http://webqtl.utmem.edu:81/bin/view/GeneNetwork/WebHome" class="fs14">
http://webqtl.utmem.edu:81/bin/view/GeneNetwork/WebHome</a>.  (Implemented by Stephen Pitts and Zhaohui Sun, with help from Bill Bug, and Hongqiang Li.)  
<BR/>
</BLOCKQUOTE>

<BLOCKQUOTE><STRONG>2006-08-20: </STRONG>Members of the Kidney Consortium have contributed a very large gene expression data for 68 strains of mice (adults of both sexes), including 53 BXD strains to GeneNetwork. The August 2006 and July 2006 data sets are preliminary and will be subject to change.  Due to strong sex differences and imperfect representation of all strains by arrays from both sexes, we recommend using the sex-corrected data sets. The Kidney Consortium is lead by Erwin Bottinger at Mt. Sinai School of Medicine. All array data we processed by Kremena Star (MSSM), and all samples were generated by Lu Lu and colleagues at UTHSC. Data have been normalized by Hongqiang Li and R Williams and Kremena Star. (Implemented by Kremena Star, Lu Lu, Hongqiang Li, Russell Chesney, Robert Williams, and Erwin Bottinger.)  
<BR>
</BLOCKQUOTE>


<BLOCKQUOTE><STRONG>2006-08-16: </STRONG>We soon will convert all mouse genomic and genetic data sets to the latest mouse genome sequence assembly. Through August 2006 all mouse sequence data has used position data from the March 2005 mm6 assembly (equivalent to NCBI Build 34). Starting September 2006 we will convert all position data to mm8, equivalent to NCBI Mouse Build 36 and Ensembl Mus musculus version 40.36a, released in February 2006. (Implemented by Evan Williams and Rob Williams.)  
<BR>
</BLOCKQUOTE>


<BLOCKQUOTE><STRONG>2006-07-12: </STRONG>Barley gene expression data and new SNP genotypes are being integrated into the GeneNetwork beta site for testing. All data are from Arnis Druka at the Scottish Crop Research Institute. Data are currently password protected, and those interested in obtaining access should contact Arnis.Druka@scri.ac.uk  (Implemented by Jintao Wang, Arnis Druka, and Rob Williams.)  
<BR>
</BLOCKQUOTE>


<P>
<DIR><IMG src="/images/upload/Barley_Search.png" valign="top">
</DIR>

<DIR><P><SMALL><B>Legend:</B> Access to the new Barley Affymetrix expression data set from the <A HREF="http://www.scri.sari.ac.uk/"_empty">Scottish Crop Research Institute</A>.  </SMALL></P>
</DIR>


</BLOCKQUOTE>




<BLOCKQUOTE><STRONG>2006-07-11: </STRONG>Jintao Wang, the lead programmer for GeneNetwork, WebQTL, GenomeGraph, and GeneWiki has accepted a position at Federal Express. We are really sorry to see Jintao leave later in July,  but wish him all the best at this exciting new position in one of the world's greatest companies. (Implemented by Jintao Wang ;-)  
<BR>
</BLOCKQUOTE>

<BLOCKQUOTE><STRONG>2006-07-10: </STRONG>The Mouse Phenome Database (MPD) and several other large data sets are being integrated into the GeneNetwork's <B>Mouse Diversity Panel</B>. To access these new data sets please select "MOUSE-GROUP-Mouse Diversity Panel". The Mouse Diversity Panel will eventually includes the MPD, additional strain data sets extracted from the published literature, the Wellcome Trust-CTC SNP collection, and several large gene expression data sets, including those for whole brain, hippocampus, cerebellum, and eye. (Implemented by Jintao Wang and Evan G. Williams.)
<BR>
</BLOCKQUOTE>

<P>
<DIR><IMG src="/images/upload/MDP_Search.png" valign="top">
</DIR>

<DIR><P><SMALL><B>Legend:</B> Access to the new Mouse Diversity Panel</A> data sets.  </SMALL></P>
</DIR>
</BLOCKQUOTE>


<P>
<DIR><IMG src="/images/upload/WBCs_MDP_female.gif" valign="top">
</DIR>

<DIR><P><SMALL><B>Legend:</B> Bar chart of white blood cell counts across 43 strains of mice taken from the Mouse Diversity Panel. Virutally all of the phenotype data are provided from the <A HREF="http://www.jax.org/phenome" class="fs14" target="_empty">Mouse Phenome Project</A>.  </SMALL></P>
</DIR>
</BLOCKQUOTE>


<BLOCKQUOTE><STRONG>2006-06-23: </STRONG>The <A HREF="http://www.genenetwork.org/webqtl/main.py?FormID=geneWiki" class="fs14" target="_empty">GeneWiki</A> annotation and open note making system has been upgraded and now has an independent search page (see the <B>Search</B> menu, pop-down). We expect to make additional changes to the interface over the next few weeks. (Implemented by Jintao Wang.)
<BR>
</BLOCKQUOTE>

<BLOCKQUOTE><STRONG>2006-06-23: </STRONG>Three new versions of the Hippocampus Consortium expression data set for adult BXD strains (June06 in MAS5, RMA, and PDNN versions). These data sets exclude several marginal arrays and correct for one incorrectly labeled strain in older data sets. (Implemented by Hongqiang Li.)
<BR>
</BLOCKQUOTE>


<BLOCKQUOTE><STRONG>2006-06-12: </STRONG>GenomeGraph is being rewritten to exploit a Scalable Vector Graphics (SVG) interface that allows zooming and other advanced GUI features. Visit the beta version of <A HREF="http://www.genenetwork.org/dbResults2.html"  class="fs14" target="_empty">GenomeGraph</A> for a test drive. You will need an SVG plugin for your machine. We do not know of an effective universal SVG plug-in for Macintosh Intel machines. Therefore, if you have a Macintosh with an Intel processor you will currently need to force Safari to open using the Rosetta emulation mode. It is easy to do this. Just follow these <A HREF="http://docs.info.apple.com/article.html?artnum=303120"  class="fs14" target="_empty"> directions</A>. (Implemented by Jintao Wang.) Running Safari using Rosetta will slow things down a bit, so consider this a temporary solution. (Implemented by Jintao Wang.)
<BR>
</BLOCKQUOTE>


<BLOCKQUOTE><STRONG>2006-06-09: </STRONG>Powerful new search method (RIF=your-text-here) that exploits the new Gene "Reference into Function" (GeneRIF) taken from NCBI. A search string such as rif=autism or RIF=Autism will find all genes/transcripts/proteins that are known to be associated with autism based on the GeneRIF entries from NCBI (n = 125 hits using one of the Affymetrix M430 expression data sets). (Implemented by Jintao Wang.)
<BR>
</BLOCKQUOTE>

<BLOCKQUOTE><STRONG>2006-06-08: </STRONG>GeneWiki has been upgraded to include the current set of NCBI GeneRIF entries. These GeneRIFs provide a summary of information about genes. You can search of data in the GeneRIFs using this simple command in the ANY or ALL fields "RIF=text_string" or "rif=text_string".  For example: RIF=schizophrenia will generate a list of all genes with schizophrenia listed anywhere in their list of GeneRifs. We encourage all users to enter their own comments and notes in the GeneWiki to supplement and extend the GeneRIF. (Implemented by Jintao Wang.)
<BR>
</BLOCKQUOTE>

<BLOCKQUOTE><STRONG>2006-06-05: </STRONG>A new assembly of the mouse genome (mm8) is now being integrated into GeneNetwork databases and the web site. Please note that many data sets still rely on the mm6 assembly. (Implemented by Jintao Wang.)
<BR>
</BLOCKQUOTE>


<BLOCKQUOTE><STRONG>2006-05-18: </STRONG>We are experimenting on the Beta site with Scalable Vector Graphics (SVG) displays of scatter plots generated from the Correlations Results tables. SVG allows you to modify the display size and the area of graphs using control clicks. You will need an SVG plug-in for your browser and hardware. SVG works fine with most Intel and Macintosh computers. However, if you have a Macintosh with an Intel processor you will find that the SVG version of the GenomeGraph does not work unless you force Safari to open using Rosetta. It is easy to do this. Just follow these <A HREF="http://docs.info.apple.com/article.html?artnum=303120"  class="fs14" target="_empty"> directions</A>. (Implemented by Jintao Wang.) Running Safari using Rosetta will slow it down somewhat, so consider this a temporary solution.
<BR>
</BLOCKQUOTE>


<BLOCKQUOTE><STRONG>2006-05-11: </STRONG>New and final "Eye M430v2 (Apr06) RMA" database has been added to GN beta and production web site. This data set includes data for 71 strains including 55 BXD strains, C57BL/6J, DBA/2J, reciprocal F1 hybrids and 12 other strains of mice. The <B>Info</B> file is still incomplete. Data generated by Weikuan Gu, Eldon Geisert, Yan Jian, Lu Lu, and Rob Williams with support from Barrett Haik and the Hamilton Eye Institute. (Implemented by Hongqiang Li, Yanhua Qu, and Jintao Wang.)
<BR>
</BLOCKQUOTE>

<BLOCKQUOTE><STRONG>2006-04-26: </STRONG>New and final Mouse BXD Eye mRNA expression database is being added to the Beta site using a new quality control procedure. The data are still being error corrected as of April 29, 2006. This data set includes 57 BXD strains, C57BL/6J, DBA/2J, F1 hybrids and 12 other strains of mice. Data generated by Weikuan Gu, Eldon Geisert, Yan Jian, Lu Lu, and Rob Williams with support from Barrett Haik and the Hamilton Eye Institute. (Implemented by Hongqiang Li, Yanhua Qu, and Jintao Wang.)
<BR>
</BLOCKQUOTE>


<BLOCKQUOTE><STRONG>2006-04-26: </STRONG> Search menus are now being updated so that they provide a complete list of available databases in hierarchical pull-down menus. (Implemented by Jintao Wang.)
<BR>
</BLOCKQUOTE>


<BLOCKQUOTE><STRONG>2006-04-18: </STRONG> We have converted our python code to utilize <A HREF="http://www.modpython.org/" target="_blank" class="fs14">Mod_python</A>. Mod_python is an Apache code module that embeds a Python interpreter within the server and that will often run many times faster than a traditional Common Gateway Interface (CGI). Mod_python will not help much for those processes (e.g., interval mapping or correlation tables) that take a long time to compute. But for fast processes, such as generating AJAX menus, opening data-editing page, it helps substantially. (Implemented by Jintao Wang.)
<BR>
</BLOCKQUOTE>

<BLOCKQUOTE><STRONG>2006-03-31: A </STRONG><B>Correlation Results Tables</B> now includes a feature to add multiple columns of correlations. This makes it possible to quickly identify well and poorly conserved correlations across data sets and tissues. You may need to use a newer browser to exploit this new feature. (See News item of March 14th; Implemented by Jintao Wang.)
<BR>
</BLOCKQUOTE>

<BLOCKQUOTE><STRONG>2006-03-16: </STRONG><A HREF="http://www.ncbi.nlm.nih.gov/entrez/linkout/index.html" target="_blank" class="fs14">NCBI Entrez Gene LinkOut</A> established. LinkOut is a service of NCBI and Entrez that allows you to link directly from PubMed and other Entrez databases to a wide range of information and services beyond the Entrez system.  NCBI pages now link from mouse and rat genes to GeneNetwork expression data sets. (Implemented by Hongqiang Li.)
<BR>
</BLOCKQUOTE>

<BLOCKQUOTE><STRONG>2006-03-14: </STRONG><A HREF="http://www.stjudebgem.org/web/view/probe/viewProbeDetails.php?id=383" target="_blank" class="fs14">GENSAT BGEM</A> link to GN established.  The Brain Gene Expression Map is a large library of in situ gene expression images of the embyronic, neonatal, and adult mouse. It includes data for over 3000 genes. (Implemented by Tom Curran and the BGEM group at St Jude Children's Research Hospital.)
<BR>
</BLOCKQUOTE>

<BLOCKQUOTE><STRONG>2006-03-15: A </STRONG><B>Correlation Results Tables</B> are now implemented using AJAX code that allows rapid resorting of the top 100, 200, or 500 traits. You will now see small UP and DOWN sort arrows in the column heads.  You may need to use a newer browser to exploit this new feature. Being able to resort tables is useful when you would like to filter a list of traits by expression value (usually from high to low) or by position. AJAX is a programming method that makes web pages more responsive and dynamic. (Implemented by Jintao Wang.)
<BR>
</BLOCKQUOTE>


<BLOCKQUOTE><STRONG>2006-01-20: GeneNetwork's </STRONG><B>MySQL</B> relational database has been moved to a dual dual-core AMD Opteron 280 computer system assembled by Monarch Computer for improved performance. This system has halved the time required to compute correlation tables from about 100 seconds down to 40 seconds. (Implemented by Jintao Wang.)
<BR>
</BLOCKQUOTE>

<BLOCKQUOTE><STRONG>2005-12-19: </STRONG>A short <A HREF="http://www.ampainsoc.org/pub/bulletin/sum05/web1.htm" target="_blank" class="fs14">Review</A> of GeneNetwork by William R. Lariviere on the American Pain Society web site.
<BR>
</BLOCKQUOTE>



<BLOCKQUOTE><STRONG>2006-01-03: A </STRONG><B>GeneWiki</B>  system is being implemented. GeneWiki (also known as Gene Notes) allows any user of GN to add notes to the GN database. You can add annotations for genes of interest. All annotation is public. For example, RWW has added annotations on expression patterns of genes in different brain regions using taken from the Allen Brain Atlas and GENSAT. Our first GeneWiki implementation does not conform to all WIKI  standards, and it may be more appropriate to consider GeneWiki as a simple system for adding notes on genes. We hope to load GeneWiki with many of the NCBI GeneRifs. (Implemented in progress by Jintao Wang.)
<BR>
</BLOCKQUOTE>


<BLOCKQUOTE><STRONG>2005-12-19: An </STRONG><A HREF="http://www.genenetwork.org/search3.html" target="_blank" class="fs14">AJAX</A> implementation of the Search Page is now being tested on the beta site. There should be almost no noticeable difference if you are using a current version of common web browsers (Explorer, Firefox, Safari). Please contact us if you have any problems. (Implemented by Jintao Wang.)
<BR>
</BLOCKQUOTE>

<BLOCKQUOTE><STRONG>2005-12-15: </STRONG><A HREF="http://www.genenetwork.org/cgi-bin/beta/main.py?FormID=geneWiki&symbol=Etv1" target="_blank" class="fs14">GeneWiki</A> feature added to GeneNetwork. You can add short annoations to the GN database that related to genes using an interface we have borrowed from the NCBI Gene Reference into Function (GeneRIF). To read all annotations provided by all users please click on the <B>Annotations</B> button (or GeneRIF button). All annotations are open and public. Annotations should ideally be of use to the research community. Here is an example of a recent annotation entered for the mouse <I>Etv1</I> gene:  "Amygdala and hippocampal CA1 and subiculum expression signature, highly specific neocortical layer 5 expression signature, cerebellar granule cell expression signature (data from Allen Brain Atlas, ABA)." 

<P>When adding a note, if possible please provide a PubMed ID number or a web address (URL). You can use the Annotations feature to find groups of genes that belong to interesting functional categories. We are currently using this feature to define sets of "expression signatures" for different parts of the mouse brain, for example genes and transcripts with highly selective expression in the dentate gyrus of the hippocampal formation. (Implemented by Rob Williams and Jintao Wang.)
<BR>
</BLOCKQUOTE>

<BLOCKQUOTE><STRONG>2005-12-02: </STRONG><A HREF="http://www.genenetwork.org/searchHelp.html" target="_blank" class="fs14">New Advanced Search</A> function now allows users to search for either cis-acting or trans-acting QTLs across entire expression data sets. The general fomat is "TransLRS=(Low_LRS_limit, High_LRS_limit, Mb_buffer)". This syntax can be combined in the ALL field with other conditions, such as the chromosome location of the QTL and the expression level of the trait. For a better explanation please see the <A HREF="http://www.genenetwork.org/searchHelp.html#cistrans_query" target="_blank" class="fs14"> Advanced Search</A> page. (Implemented by Jintao Wang.)
<BR>
</BLOCKQUOTE>

<BLOCKQUOTE><STRONG>2005-11-21: </STRONG><A HREF="http://www.genenetwork.org/schema_query_probeset.xml" target="_blank" class="fs14">Demonstration XML Schema</A> for mouse data sets has been published for the use of the Biomedical Informatics Research Network (BIRN). For readability, please review the source code version of this page. This is an initial demonstration/proof-of-principle. (Implemented by Hongqiang Li.)
<BR>
</BLOCKQUOTE>

<BLOCKQUOTE><STRONG>2005-11-15: </STRONG><B>Basic Statistics</B> pages have been improved to handle larger data sets and to provide better graphic output. (Implemented by Jintao Wang and Rob Williams)
<BR>
</BLOCKQUOTE>

<BLOCKQUOTE><STRONG>2005-11-14: </STRONG><A HREF="http://www.genenetwork.org/literatureCorrelation.html" target="_blank" class="fs14">Literature Correlations </A>  gene data set by Ramin Homayouni, Michael Berry and colleagues has been updated. The literature correlations are positive values between o and 1 that summarize the pair-wise similarity of genes (or transcripts) on the basis of the known literature using the methods described on the <A HREF="https://shad.cs.utk.edu/sgo/sgo.html" target="_blank" class="fs14">Semantic Gene Organizer </A> site.  (Implemented by Ramin Homayouni, Lai Wei, Kevin Heinrich, and Jintao Wang.)
<BR>
</BLOCKQUOTE>

<BLOCKQUOTE><STRONG>2005-11-01: </STRONG><A HREF="http://www.genenetwork.org/webqtl/main.py?FormID=editHtml" target="_blank" class="fs14">New Affymetrix M430v2 Eye Data Set for 63 strains of mice</A>  (C57BL/6J, DBA/2J, their reciprocal F1 hybrids, 47 BXD recombinant inbred strains, and 12 diverse inbred strains) have been entered on the beta site by the UTHSC Hamilton Eye Institute. Expression data for whole eye is available from <B>Species</B> = Mouse,  <B>Group</B> = BXD, and <B>Type</B> = Eeye. The Information (INFO) file that accompanies this M430 data set is still provisional. Use of these data in publications is currently limited to members of the HEIMED consortium pending addition of more data, publication, and formal release, but if you would like permission to make selected use of data please contact Robert W. Williams, UTHSC. (Implemented by Lu Lu, Yan Jiao, Yanhua Qu with support of the Hamilton Eye Institute.)
<BR>
</BLOCKQUOTE>


<BLOCKQUOTE><STRONG>2005-10-24: </STRONG><A HREF="http://www.genenetwork.org/dbdoc/Hippocampus_M430_V2_PDNN_Sept05.html" target="_blank" class="fs14">New Affymetrix M430v2 Hippocampus Data Set for 96 strains of mice</A>  (65 BXD, 13 CXB, and 16 diverse inbred strains, B6D2F1 and D2B6F1) will be placed on the beta site by the Hippocampus Array Consortium at the end of October. Expression data for whole hippocampus will be available from <B>Species</B> = Mouse,  <B>Group</B> = BXD, and <B>Type</B> = Hippocampus. The Information (INFO) file that accompanies this M430 data set is still provisional. Use of these data in publications is currently limited to members of the consortium pending data addition, publication, and formal release, but if you would like permission to make selected use of data please contact Robert W. Williams, UTHSC. (Implemented by Lu Lu, Shirlean Goodwin, Yanhua Qu, Rob Williams, and members of the Hippocampus Consortium.)
<BR>
</BLOCKQUOTE>


<BLOCKQUOTE><STRONG>2005-10-10: </STRONG><A HREF="http://www.genenetwork.org/dbdoc/SA_M2_0905_P.html" target="_blank" class="fs14">New Affymetrix M430v2 Striatum Data Set for a B6D2F2 Intercross</A> has been placed on the beta test site by Robert Hitzemann and colleagues. Expression data for the striatum of 30 males and 30 females are available from <B>Species</B> = Mouse and  <B>Group</B> = BDF2-2005. The Information (INFO) file that accompanies the M430 data is still provisional. For use of these unpublished data please contact Robert Hitzemann, Department of Behavioral Neuroscience, Oregon Health & Science University. (Implemented by Yanhua Qu.)
</BLOCKQUOTE>


<BLOCKQUOTE><STRONG>2005-10-07: </STRONG><A HREF="http://www.genenetwork.org/searchHelp.html" target="_blank" class="fs14">Advanced Search</A> options have been improved. The main improvement involves combining Gene Ontology searches with other advanced search syntax. (Implemented by Hongqiang Li.)
</BLOCKQUOTE>


<BLOCKQUOTE><STRONG>2005-09-28: </STRONG> GeneNetwork Mouse <A HREF="http://www.genenetwork.org/cgi-bin/beta/snpBrowser.py" target="_blank" class="fs14">SNP Browser</A> has been upgraded with Perlegen/NIEHS data. The SNP Browser is a tool that is used in combination with the Interval Analyst to evaluate and rank genes and polymorphisms in intervals thought to be responsible for variation in traits. The SNP Browser includes all Celera Genomics mouse SNPs, all public mouse SNPs in dbSNP (as of August 2005), and all Perlegen-NIEHS SNPs (http://mouse.perlegen.com/mouse/download.html as of late Sept 26, 2005). We thank Paul Thomas and Richard Mural of Celera Genomics, Gary Churchill and Natalie Blade of the Jackson Laboratory, and the Perlegen/NIEHS sequencing consortium for help and access to data. (Implemented by Robert Crowell, Alex Williams, and Jintao Wang.)
<BR>
<BR>An example: To search for SNPs type in this string and then modify position as desired:
<BR>http://www.genenetwork.org/cgi-bin/beta/snpBrowser.py
</BLOCKQUOTE>


<BLOCKQUOTE><STRONG>2005-09-27: </STRONG><A HREF="http://www.geneontology.org/" target="_blank" class="fs14">Gene Ontology</A> searching is now possible. This search feature allows you to search for all genes/transcripts related to particular categories using the appropriate GO identifer.  For example, to extract all transcripts associated with "synaptic vesicle exocytosis" enter the string "GO:0016079" in the ANY field. To browse GO terms and classes link to <A HREF="http://www.godatabase.org/cgi-bin/amigo/go.cgi" target="_blank" class="fs14">AmiGo</A>. As of Sept 2005, the GO contains approximately 20,000 terms of which approximately 6300 GO terms can be associated with genes in one or more of the GeneNetwork databases. Approximately 700 high level GO terms will return well over 200 genes. Given the 500 transcript limit it is therefore useful to select lower level GO terms that will return 100 or fewer probe sets/transcripts/genes. (Implemented by Hongqiang Li.)
</BLOCKQUOTE>


<BLOCKQUOTE><STRONG>2005-09-20: </STRONG><A HREF="http://genome.ucsc.edu/" target="_blank" class="fs14">The UCSC Gene Browser</A> is now linked to GeneNetwork from the Gene Description and Page Index as a "Quick Link" for both mouse and rat genomes. (Implemented by Jintao Wang at UT and Fan Hsu at UCSC.)
</BLOCKQUOTE>


<BLOCKQUOTE><STRONG>2005-09-06: </STRONG><A HREF="http://www.genenetwork.org/phenotype_sop.html" target="_blank" class="fs14">Phenotype Data Entry SOP</A>. We are beginning to develop standard operating procedures (SOP) to allow colleagues to deposit new data sets into the GeneNetwork. Please review this initial Phenotype data entry SOP if you have traits that you would like added to either an existing or new mapping panel (Partially implemented by Rob Williams.)
</BLOCKQUOTE>


<BLOCKQUOTE><STRONG>2005-08-26: </STRONG><B>OHSU/VA B6D2F2 Brain mRNA 430 (Aug05) MAS5, RMA and PDNN </B> array data sets now are available. These data sets include M430 Set A and Set B arrays (Implemented by Yanhua Qu.)
</BLOCKQUOTE>

<BLOCKQUOTE><STRONG>2005-08-19: </STRONG><A HREF="http://www.genenetwork.org/dbResults.html" target="_blank" class="fs14">GenomeGraph</A> has been implemented for several large array data sets and can now be used for testing purposes. GenomeGraph is a new module of The GeneNetwork that is designed for the analysis of entire array data sets. (Implemented by Jintao Wang.)
</BLOCKQUOTE>

<BLOCKQUOTE><STRONG>2005-08-17:</STRONG> <A HREF="http://www.genenetwork.org/cgi-bin/beta/schema.py" target="_blank" class="fs14">Dynamic GeneNetwork Database Schema Description</A> allows database experts to review the data structure and fields used by the GeneNetwork MySQL relational database. We have just begun the textual annotation of the database tables and field. This new system will soon replace the current database "dump" available at http://www.genenetwork.org/schema.html (Implemented by Hongqiang Li.)</BLOCKQUOTE>


<BLOCKQUOTE><STRONG>2005-08-16:</STRONG> <B>Traits in the Selections Windows Now Sortable</B>. The Selection command is used to move trait data from one or more databases into a single <B>Selections window</B> (aka the "shopping cart") for common analysis. For example, users can put classical phenotypes such as body and brain weight in the same Selections window with transcripts for growth hormone receptor (Ghr), GH releasing hormone (Ghrh), and GHRH receptor (Ghrhr) in liver and brain. The new feature makes it possible to sort items in the Selections window by database, position, or name. Sorting is helpful is reviewing contents of the window and in reordering items prior to calculating correlation matrices. Please recall that all itmes in a Selections window must come from a single genetic reference population or panel, for example the AXB/BXA strains of mice, the BXH strains of rat, or from one of several intercrosses. (Implemented by Jintao Wang.)</BLOCKQUOTE>

<P>
<DIR><DIR><IMG src="/images/upload/sort_by.png" valign="top">
</DIR></DIR>

<BLOCKQUOTE><STRONG>2005-08-12:</STRONG> <A HREF="http://www.genenetwork.org/dbdoc/LVF2_M_0704_R.html" target="_empty" class="fs14">New Mouse Liver and Metabolic Trait Databases</A> have been released by Dr. Alan Attie and colleagues. While these data may be reviewed, their use is still are reserved until final publication. The primary database is an Affymetrix M430 survey of gene expression in the liver of 60 selected F2 mice (a B6 x BTBR F2-ob/ob cross) that includes data on approximately 45,000 probe sets. This array database is accompanied by 24 classical metabolic and blood chemistry traits. All F2 animals were genotyped a 194 microsatellite markers. (Implemented by Alan Attie and colleagues, Yanhua Qu, and Jintao Wang.)
</BLOCKQUOTE>


<BLOCKQUOTE><STRONG>2005-08-08: </STRONG> <B>The Interval Analyst</B> (IA) provides a tabular summary of known genes in a chromosomal interval with data on gene expression, gene size, SNPs number and density, and human homologs. The IA is still a beta site function but will be release to the public site in the next week. The IA table is automatically generated with each chromosome map. IA tables can be extensively customized and resorted. For the BXD and AXB/BXA mouse genetic reference panels, the IA also provides access to Celera SNPs, as well as public SNPs for a variety of sources. Clicking on the SNP number for a specific gene in the IA generates a SNP browser table (at present, only for mouse). The purpose of the IA is to allow users to rank-order genes in an interval that may be contributing to variability in phenotypes. (Implemented by Evan Williams, Robert Crowell, Alex Williams, and Rob Williams.)
</BLOCKQUOTE>


<BLOCKQUOTE><STRONG>2005-08-08: </STRONG> The design of <B>Chromosome and Whole Genome QTL Maps</B> has been signficantly improved and updated. These new physical QTL maps merge LRS or LOD functions with gene and SNP tracks and can be zoomed to the level of single genes and SNPs. Maps can be exported in 2X versions that are near publication quality. Below most maps you will now find a customizable <B>Interval Analyst</B> table that can be customized to help rank order candidate genes. Variants of these new maps have been introduced to handle all species and genetic reference populations. (Implemented by Robert Crowell, Alex Williams, Evan Williams, and Rob Williams; final integration by Jintao Wang.)
</BLOCKQUOTE>

<P>
<DIR><DIR><IMG src="/images/upload/physical_chr10_map.gif" valign="top">
</DIR>

<P><SMALL><B>Legend:</B> Sample of a new high resolution physical map. This map shows a locus that modulates the expression of the <I>Cart</I> transcript (cocaine and amphetamine regulated transcript) on distal Chr 10 in BXD mouse strains (brain tissue). The <B>Control Block</B>, top middle, permits users to customize the display and its resolution. Pink, blue, and beige horizontal bars above the map provide links to higher resolution maps (8x) or to the UCSC and ENSEMBL genome browsers. Statistical thresholds for linkage are marked by grey and pink horizontal lines and are based on 2,000 permutations. The Y-axis provides a scale for the plot of LRS or LOD scores that are plotted using a thicker blue line. The calculation of linkage statistics are based on a total of 147 useful markers that have been genotyped in all 89 BXD strains (The Wellcome-CTC Mouse Strain SNP database with added microsatellite markers). The far more digital look of the LRS function that traditional interval maps arises for the simple reason that locations of recombinations in this cross have been precisely defined and only a fewer regions exploit a true interval mapping approach (see News item of 2005-06-17 for additional detail). </SMALL></P>

<P><SMALL>The thinner red and green lines and the right Y-axis display the additive effect size; green for high alleles inherited from one parent (DBA/2J in this example), and red for high alleles from the other parent (C57BL/6J). The units are log2 expression differences where 0.2 is equivalent to a 2^0.2-fold difference. The large number of closely packed tick marks along the top of the map show locations of genes on Chr 10. Gene blocks are color coded by the average density of SNPs per gene using a rainbow color sequence with low density in the blue/green spectrum and high density in orange/red spectrum. The bright orange hash marks along the X-axis provide a graphic estimate of numbers of SNPs that are segregating in the BXD strains in any particular chromosomal region. A long interval from 30 Mb to 65 Mb is almost identical by descent between the two parental strains. </SMALL></P>

<P><SMALL>Many regions of these maps are responsive to a mouse click. For example, the name and size of any gene can be determined by simply placing the mouse cursor over its mark. The same applies to the significance thresholds and the SNP track. Below each of these maps is a complete list of known genes in the interval with numerous links to other data types, including information on expression, lists of known SNPs in each gene, and corresponding regions of the human genome. All physical map positions in mouse are based on the Mouse Build 34, mm6 (March 2005).</SMALL> 
</P>
</DIR>


<BLOCKQUOTE><STRONG>2005-08-02: </STRONG>An <B>Export Traits</B> function button has been added to the set of tools available in each <B>Selections Window</B> (the Selections window is known informally as the "shopping cart"). Export Traits now joins other tools such as <B>Cluster Tree</B>, <B>Network Graph</B>, and <B>Compare Correlates</B> at both the top and bottom or each Selection window.  Any set of traits in the Selection window can be easily exported, including conventional phenotypes, genotypes, and subsets of array data. The default output format is compatible with Microsoft Excel. (Implement by Jintao Wang.) 
</BLOCKQUOTE>

<BLOCKQUOTE><STRONG>2005-07-29: </STRONG> <B>Rat RAE 230A and Mouse (M430 and U74A) Affymetrix Probe Set Annotation Tables</B> have been significantly improved and realigned to rat and mouse genome assemblies. Information taken from the BLAT alignment data has been added to GeneNetwork data tables. Data types include the alignment score of concatenated probes, probe set specificty (usually the ratio of first hit score divided by second hit score), a position values of the 3' and 5' ends of the concatenated probe sequences. [Implemented by Senhua Yu (rat) and Yanhua Qu (mouse).]
</BLOCKQUOTE>

<BLOCKQUOTE><STRONG>2005-07-27: </STRONG> All <B>Mouse Genotype Databases </B> have now been fully updated using Wellcome-Illumina-CTC SNP data sets consisting of 13377 SNPs.  These SNPs have been integrated with the older microsatellite markers used through July 2005.  You can search for markers (see <A HREF="http://www.genenetwork.org/searchHelp.html" target="_empty" class="fs14">Advanced Search</A>) and treat genotypes as a standard "trait."  You can also align the sequence of any marker to the latest genome assembly to determine where a SNP or microsatellite is located. (Implemented by Jing Gu, Lu Lu, and Jintao Wang.)
</BLOCKQUOTE>

<BLOCKQUOTE><STRONG>2005-07-26: </STRONG><B>Complete Upgrade of the PUBLISHED PHENOTYPE Databases</B>. All PubMed abstracts were searched in June and July of 2005 for publications pertaining to BXD, AXB, CXB, or BXH mouse recombinant inbred strains. Means and standard errors were collected, reviewed, and extracted from these papers. Data were then entered manually in GeneNetwork tables by Emily English and Elissa Chesler. </BLOCKQUOTE>

<BLOCKQUOTE><STRONG>2005-07-26: </STRONG><B>Sorting Traits</B> by several different variables is now possible in the Search Results page. Select from seven different ways to sort lists as shown in the screen shot below.(Implemented by Jintao Wang)</BLOCKQUOTE>
<P>
<DIR><IMG src="/images/upload/sort_by.gif" valign="top">
</DIR>
<P>
<BLOCKQUOTE><STRONG>2005-07-26: </STRONG><A HREF="http://www.genenetwork.org/qtlreaper.html"  target="_empty" class="fs14">QTL Reaper</A> tutorial has been added to the GeneNetwork site. QTL Reaper is a command line program for high throughput mapping of array data sets. (Implemented by Evan Williams.)</BLOCKQUOTE>

<BLOCKQUOTE><STRONG>2005-07-25: </STRONG><B>An Error Detected and Corrected in SJUT Cerebellum</B> databases dated March 2005. Data for BXD23 mistakenly included a BXD14 sample. All three March 2005 databases (RMA, PDNN, MAS5) have now been corrected. Values for the two affected strains are changed relative to data in this database prior to July 25, 2005. (Implemented by Jing Gu, Rob Williams, and Yanhua Qu.)</BLOCKQUOTE>

<BLOCKQUOTE><STRONG>2005-07-22: </STRONG><B>Modified Linux Virtual Server</B> configuration to eliminate problems with client institution firewall restrictions on numbers of simultaneous connections. Our thanks to Dr. Michael Miles for his help diagnosing firewall problems for clients. (Implemented by Jintao Wang.)</BLOCKQUOTE>

<BLOCKQUOTE><STRONG>2005-07-21: </STRONG><B>Improved Advanced Search</B>. It is now possible to combine search strings to generate complex queries. For example, this combination <B>Mb=(Chr11 90 100) Mean=(12 20)</B> when entered into the lower ALL field will find transcripts that map to Chr 11 between 90 and 100 Mb that also have mean expression between 12 and 20 units. (Implemented by Jintao Wang.)</BLOCKQUOTE>


<BLOCKQUOTE><STRONG>2005-07-15: </STRONG>. GeneNetwork Mouse <A HREF="http://www.genenetwork.org/cgi-bin/beta2/snpBrowser.py?chr=5&start=4.088955&end=4.110455&padding=2.0&geneName=Cyp51&sort=B6_D2" target="_blank" class="fs14">SNP Browser</A> has been implemented. The SNP Browser is a tool that will eventually be used in combination with the Interval Analyst to evaluate and rank genes and polymorphisms in intervals thought to be responsible for variation in traits. The SNP Browser includes all Celera Genomics mouse SNPs, all public mouse SNPs in dbSNP (as of August 2005), and all Perlegen-NIEHS SNPs (http://mouse.perlegen.com/mouse/download.html as of late June 2005).  The SNP Browser is still at an early stage of development. We thank Paul Thomas and Richard Mural of Celera Genomics, Gary Churchill and Natalie Blade of the Jackson Laboratory, and the Perlegen/NIEHS sequencing consortium for help and access to data. (Implemented by Robert Crowell, Alex Williams, and Jintao Wang.)
<BR>
<BR>An example: To search for SNPs on Chr 5 from X to Y Mb:
<BR>http://www.genenetwork.org/cgi-bin/beta2/snpBrowser.py?chr=5&start=4.088955&end=4.110455
</BLOCKQUOTE>


<BLOCKQUOTE><STRONG>2005-07-15: </STRONG><B>Access to GeneNetwork Archive site</B>. The archive site provides access to old data sets and old genotype files that have now been superceded. We anticipate that it will be used mostly to verify old findings and to document changes in results. The Archive is now available from the main search page. (Implemented by Jintao Wang.)</BLOCKQUOTE>

<BLOCKQUOTE><STRONG>2005-07-13: </STRONG><B>LXS Genotypes Upgraded</B>. Genotypes for the large GRP of LXS strains has been greatly improved thanks to the Illumina-Wellcome-CTC SNP project. The original set of 330 markers has been replaced with a  set of 2659 informative markers. Download either the <A HREF="http://www.genenetwork.org/genotypes/LXS.geno" target="_blank" class="fs14">LXS</A> genotypes or <A HREF="http://www.genenetwork.org/genotypes/BXD.geno" target="_blank" class="fs14">BXD</A> genotypes used by WebQTL as text files.
(Implemented by Jing Gu and Jintao Wang.)</BLOCKQUOTE>

<BLOCKQUOTE><STRONG>2005-07-12: </STRONG><B>Search Page Upgraded</B>. Users now can change the default settings to those they most commonly use. Your browser must be configured to allow The GeneNetwork to retain a "cookie" on your computer. We have also added a new button labeled ADVANCED SEARCH that provides advice and syntax for searches. (Implemented by Jintao Wang.)</BLOCKQUOTE>

<BLOCKQUOTE><STRONG>2005-07-12: </STRONG><B>Pair-Scan Upgraded</B>. The pair scan now exploits the new Wellcome-Illumina high density genotype files. This result in more exhaustive searches for two-locus interactions. This is particulary true when single chromosome pairs are scanned by clicking on the initial DIRECT output graph. (Implemented by Jintao Wang.)</BLOCKQUOTE>

<BLOCKQUOTE><STRONG>2005-07-12: </STRONG><B>Updated Affymetrix M430 GeneChip Annotation Data</B>. We have realigned all M430 probes and probe set sequences onto the latest mouse assembly (Build 34 or mm6). This annotation is more complete than most other available M430 probe set annotation of which we are aware, including Affymetrix NetFX.  (Implemented by Yanhua Qu.)</BLOCKQUOTE>

<BLOCKQUOTE><STRONG>2005-06-17: </STRONG><B>New High Density Mapping Algorithm</B> that exploits the Wellcome-CTC SNP data has been implemented for the BXD mouse genetic reference populations on both public and beta sites. In the case of the BXD panel (BXD1 through BXD100), the merged SNP and microsatellite maps are based on a total of 7636 informative markers that differ between the parental strains, C57BL/6J (B) and DBA/2J (D). The locations of these makers are known on the latest assembly of the mouse genome (Build 34, mm6). The median distance between markers in this subset is 178,831 bp. The mean distance is 324,493 bp. There are only 26 intervals between markers that are longer than 5 Mb. No interval is greater than 10 Mb except on Chr X. These long intervals are essentially monomorphic between the parental strains. 

<P>The new algorithm exploits a selected subset of 3795 markers that includes all markers with unique strain distribution patterns (<A HREF="http://www.genenetwork.org/glossary.html#SDP" target="_blank" class="fs14">SDP</A>), as well as pairs of markers (the most proximal and most distal markers) for SDPs represented by two or more markers. This BXD genotype data set can be downloaded by ftp at <A HREF="ftp://atlas.utmem.edu/public/BXD_WebQTL_Genotypes_June05.txt" target="_blank" class="fs14">ftp://atlas.utmem.edu/public/BXD_WebQTL_Genotypes_June05.txt</A>.

<P>The mapping algorithm is a mixture of simple marker regression, linear interpolation, and standard Haley-Knott interval mapping. If two adjacent markers have identical SDPs they will have identical linkage statistics, as will the entire interval between the markers (assuming complete and error-free haplotype data for all strains). On a physical map the <A HREF="http://www.genenetwork.org/glossary.html#LRS" target="_blank" class="fs14">LRS</A> and the <A HREF="http://www.genenetwork.org/glossary.html#additive" target="_blank" class="fs14">additive effect</A> values will therefore be constant over this interval. Between neighboring markers that are separated by 1 cM or more we use a conventional interval mapping method (Haley-Knott) combined with a Haldane estimate of genetic distance. When the interval is less than 1 cM we simply interpolate linearly based on a physical scale between the markers. The result of this <B>mixture mapping algorithm</B> is a map of the trait that has an unusal profile that is particular striking on a physical (Mb) scale, with many plateaus, abrupt linear transitions between plateaus, and a few regions with the standard graceful curves typical of interval maps.</P>

<P>The same procedure will soon be implemented for other mouse GRPs, including AXB/BXA, CXB, BXH, and AKXD.

<P>For users that would like reference access to the old set of genotypes, we will set up an Archive site with the May 2005 microsatellite markers and maps.

<P>To download the combined SNP and microsatellite genotype file used in WebQTL please link to ftp://atlas.utmem.edu/public/ and look for Illumina_UT_BXD_May05.xls (entire data set) or BXD_WebQTL_Genotypes_June05.txt (extracted subset of markers used by WebQTL), or link to Dr. Richard Mott's <A HREF="http://www.well.ox.ac.uk/mouse/INBREDS/" target="_blank" class="fs14">Mouse Inbred Line Genotype</A> site for the original SNP data set. (Implemented by RW Williams, KF Manly, and JT Wang.)
</BLOCKQUOTE>

<BLOCKQUOTE><STRONG>2005-06-13: </STRONG><B>Rat HXB Fat Data Set</B> released on the www.genenetwork.org/search3.html test site (stabilized RMA transform). The Affymetrix RAE230A data files generated by Tim Aitman and colleagues were downloaded from the Array Express site. The set of 120+ arrays covers a total of 30 RI strains and complements a recent paper (H&uuml;bner et al., <A HREF="http://www.ncbi.nlm.nih.gov/entrez/query.fcgi?cmd=Retrieve&db=pubmed&dopt=Abstract&list_uids=15711544" target="_blank" class="fs14">2005</A>). Error checking is still in progress and this is a pre-release data set to use for test purposes. (Implemented by Senhua Yu, R. Williams, and Jintao Wang. More transforms are in progress.)</BLOCKQUOTE>

<BLOCKQUOTE><STRONG>2005-06-12: </STRONG><B>Moved GeneNetwork and Upgraded Utilities</B>. The GeneNetwork and the WebQTL module has been moved to a cluster of nine P4 single processor computers.  Eight of the nodes are devoted to the GeneNetwork application code while the ninth node runs the Linux virtual server. The MySQL database server currently runs on a separate Proliant dual processor node. The <A HREF="http://roundup.sourceforge.net/doc-0.8/" target="_blank" class="fs14">Roundup</A> issue tracking systems has been upgraded to v. 0.83 and is now available at http://www.genenetwork.org:8080/webqtl/. Analog has also been upgraded to v 6.0. (Implemented by Jintao Wang, with thanks again to Ari Berman.)</BLOCKQUOTE>

<BLOCKQUOTE><STRONG>2005-05-24: </STRONG><B><A HREF="http://www.well.ox.ac.uk/mouse/INBREDS/" target="_blank" class="fs14">Ultra-high Resolution Mouse SNP Genetic Maps</A></B> are now gradually replacing the previous generation of microsatellite maps. Until May 2005, all genetic maps of recombinant inbred strains of mice in WebQTL have relied heavily on a set of roughly 1500 microsatellite markers genotyped across all RI sets by the Informatics Center for Mouse Neurogenetics (Williams et al., 2001; Peirce, Lu et al, 2004). In collaboration with members of the CTC (Richard Mott, Jonathan Flint and colleagues), we have helped genotype a total of 480 strains using a panel of 13,377 SNPs. More than half of the SNPs are informative in most crosses. These SNPs have been combined with microsatellites to produce new consensus maps for BXD and other GRPs using the latest mouse genome assembly as a reference frame (Build 34 - mm6). In the case of the BXD GRP, a total of 88 strains were genotyped using the full set of SNPs of which 7482 are informative. The order of markers given in WebQTL is essentially the same as that given in Build 34. To reduce false positive errors when mapping using this ultradense map, we have eliminated most single genotypes that generate double-recombinant haplotypes. Double-recombinant haplotypes are most commonly produced by typing errors ("smoothed" genotypes). (Implemented by Lu Lu, Jing Gu, Jintao Wang, Ken Manly, and Rob Williams, with help from Jonathan Flint and Richard Mott). </BLOCKQUOTE>

<BLOCKQUOTE><STRONG>2005-05-23: </STRONG><B>Search Functions</B> have been upgraded. It is now possible to (1) find all transcripts whose genes map to a give chromosomal location; (2) all traits and transcripts that have a mean value within a particular range; (3) all traits that have a peak genome-wide linkage score (LRS score or p value) within a particular range. These new search functions are still being tested on the test site (http://www.genenetwork.org/search3.html). (Implemented by Jintao Wang).

<P> (1) To find transcripts by <B>chromosomal position</B> the search syntax needs to follow these rules:
<UL>
<LI>"Position in (ChrY 0.3 52.4)"  or "Position = (Chr1, 98 104)" [Note: No space between "Chr" and the number or letter of the chromosome. ]
 
<LI>"Pos in (ChrY 0.3 52.4)" or "Pos =(Chr1, 98 104)"  [don't enter the quotes.]

<LI>"Mb in (ChrY 0.3 52.4)" or "Mb = (Chr1, 98 104)"  [don't enter the quotes.]

</UL>

<P> (2) To find traits by <B>mean value</B>,  the search syntax needs to follow these rules:
<UL>
<LI>"Mean in (12.3, 12.4)"  or Mean=(12.3, 12.4)  [These strings will find those traits with a mean value from 12.3 and 12.4. Don't enter the quotes.]
 </UL>

<P> (3) To find traits by <B>LRS value</B> or <B>p value</B>, the search syntax needs to follow these rules:
<UL>

<LI>"LRS in (20, 30)"  or "LRS=(20, 30)"  [These strings will find traits with LRS values ranging from 20 to 30. This search depends on the existence of database of precomputed LRS values. If this database has not yet been set up for a particular data set, then the search will not return any records. Don't enter the quotes.]

<LI>"pvalue in (0.0001, 0.001)"  or "pvalue=(0.0001, 0.001)"  [These strings will find traits with p values ranging from 0.0001 to 0.001. This search depends on a database of precomputed values. If this database has not yet been set up for a particular data set, then the search will not return any records. Don't enter the quotes.]
 </UL> </BLOCKQUOTE>


<BLOCKQUOTE><STRONG>2005-05-13: </STRONG><B><A HREF="http://www.linuxvirtualserver.org/" target="_blank" class="fs14">Virtual Server</A></B> implementation of The GeneNetwork is being beta tested. The Linux Virtual Server (LVS) allows GeneNetwork to exploit a small clusters of servers to handle larger numbers of clients quicky. Performance is particularly critical during bioinformatics class projects when large numbers of students make nearly simultaneous requests.  (Implemented by Jintao Wang, Senhua Yu, and Ari Berman).
</BLOCKQUOTE>


<BLOCKQUOTE><STRONG>2005-05-12: </STRONG><B><A HREF="http://www.genome-explorations.com/" target="_blank" class="fs14">Genome Explorations Inc.</A></B> has been provided a license to run a copy of the GeneNetwork and WebQTL software as part of a Phase I Small Business Innovation Research (SBIR) grant from NIAAA. The TCP/IP address is <A HREF="http://132.192.47.88/" target="_blank" class="fs14">132.192.47.88</A>. The site currently contains three data sets (MAS5, RMA, and PDNN) generated at GE and UTHSC (subcontractor) using a total of 85 Affymetrix M430 2.0 arrays.  The first data release consists of 26 BXD strains, the two parental strains, C57BL/6J and DBA/2J, and ten other inbred strains of mice (A/J, 129S1/SvJ, AKR/J, BALB/cJ, BALB/cByJ, C3H/HeJ, CAST/Ei, KK/HIJ, LG/J, and NOD/J).  (Implemented by Jintao Wang, Yanhua Qu, Lu Lu, Roberrt Williams, Robert Rooney, and Divyen Patel).
</BLOCKQUOTE>


<BLOCKQUOTE><STRONG>2005-05-10: </STRONG><B><A HREF="http://www.genenetwork.org/dbResults.html" target="_blank" class="fs14">Whole Transcriptome Mapping Display</A></B>: We are testing an interface that displays a entire transcriptome QTL map for a tissue similar to figures 3A and 3B of Chesler and colleagues (2005). Note that one parameter can be used to modify the false discovery rate of the points that are plotted. Plots have been precomputed for more than 30 databases and transforms. (Implemented by Jintao Wang).
</BLOCKQUOTE>

<BLOCKQUOTE><STRONG>2005-05-04: </STRONG><B>New Mouse Genome Assembly (<A HREF="http://www.ncbi.nlm.nih.gov/genome/guide/mouse/contig/Build34.html" target="_blank" class="fs14">NCBI Build 34</A>, UCSC mm6)</B> released by NCBI (implemented by Deanna Church and colleagues).  Over the next several months all mouse genome megabase and nucleotide position data and links in the GeneNetwork (markers, probes, SNPs, genes) will be converted to this new assembly. BLAT searches initiated with WebQTL already exploit the most recent build. GeneNetwork users may find small discrepancies in gene and marker locations until all database tables are updated. 
</BLOCKQUOTE>


<BLOCKQUOTE><STRONG>2005-04-22: </STRONG><B>Arabidopsis Data Sets</B> released on the www.genenetwork.org/search3.html test site. The <B>Genotypes</B> and <B>Phenotypes</B> files for the Bay-0 x Shahdara cross data were all provided by Olivier Loudet. Please see the <A HREF="http://www.genenetwork.org/arabidopsisCross.html#BayXSha" target="_blank" class="fs14">Information</A> file. Implemented by O. Loudet, R. Williams, and Jintao Wang.
</BLOCKQUOTE>

<BLOCKQUOTE><STRONG>2005-04-21: </STRONG><B>Rat HXB Kidney Data Set</B> released on the www.genenetwork.org/search3.html test site (original RMA transforms). The Affymetrix RAE230A data files were provided by Norbert H&uuml;bner and colleagues. The set of 120+ arrays covers a total of 30 RI strains and complements a recent paper (H&uuml;bner et al., <A HREF="http://www.ncbi.nlm.nih.gov/entrez/query.fcgi?cmd=Retrieve&db=pubmed&dopt=Abstract&list_uids=15711544" target="_blank" class="fs14">2005</A>). Implemented by Senhua Yu, R. Williams, and Jintao Wang. More transforms are in progress (MAS5 added May 13, 2005).

</BLOCKQUOTE>

<BLOCKQUOTE><STRONG>2005-04-14: </STRONG><A HREF="http://www.genenetwork.org/dbdoc/BR_U_0405_SS.html" target="_blank" class="fs14"><B>New S-Score Transform</B></A> for the BXD Brain data set released on the www.genenetwork.org/search3.html test site. This data set complements existing MAS5, PDNN, RMA, dCHIP, and HWTIPM transforms. The Significance score method centers the expression of every probe set at 0. The signal values are therefore the strain deviations in Z score units from the grand mean based on 100 arrays.  The  <A HREF="http://www.brainchip.vcu.edu/expressionda.htm"  target="empty" class="fs14">S-score software</A> is described in Zhang et al. (<A HREF="http://www.ncbi.nlm.nih.gov/entrez/query.fcgi?cmd=Retrieve&db=pubmed&dopt=Abstract&list_uids=11902839" target="empty" class="fs14">2002</A>) and Kerns et al. (<A HREF="http://www.ncbi.nlm.nih.gov/entrez/query.fcgi?cmd=Retrieve&db=pubmed&dopt=Abstract&list_uids=14597311"  target="empty" class="fs14">2003</A>).
 </BLOCKQUOTE>

<BLOCKQUOTE><STRONG>2005-04-08: </STRONG><B>Expanded HBP/Rosen Striatum Data Sets</B> released on the www.genenetwork.org/search3.html test site (MAS5, RMA, and PDNN transforms). The new data set covers a total of 33 strains using 59 M430 2.0 arrays. A good demonstration of the improved performance of the expanded data set is <I>Kcnj9</I> (probe set 1450712_at_A), a known cis-QTL in multiple data sets. This trait generates a peak LRS score of 27.0 in the initial November 2004 data set (MAS5) and a peak LRS of 47.8 in the April 2005 data set (MAS5). The peak LRS is approximately 600 Kb proximal to the <I>Kcnj9</I> gene. The Heritability Weight Transform (HWT) data set will be added in the next several weeks. 
</BLOCKQUOTE>

<BLOCKQUOTE><STRONG>2005-04-04: </STRONG><B>Expanded INIA Brain Data Sets</B> released on the www.genenetwork.org/search3.html test site (MAS5, RMA, and PDNN transforms). Seventy-one new samples have been added, bringing the total to 105 arrays covering 42 BXD strains, both parents, and the F1 hybrid. A good demonstration of the improved performance of the expanded data set is <I>Kcnj9</I> (probe set 1450712_at_A), a known cis-QTL in multiple data sets. This trait generates a peak LRS score of 14 in the initial October 2004 data set (MAS5) and a peak LRS of 41.9 in the April 2005 data set (MAS5). The peak LRS is approximately 2000 Kb distal to the <I>Kcnj9</I> gene. We have also tested these data using probe set 1418908_at_A (<I>Pam</I>). This trait generates a peak LRS score of 52.8 in the initial October 2004 data (MAS5) and a peak LRS of 54.2 in the April 2005 (MAS5). The peak LRS is approximately 800 Kb distal to <I>Pam</I> gene. The Heritability Weight Transform (HWT) transform will be added in the next several weeks. 
</BLOCKQUOTE>

<BLOCKQUOTE><STRONG>2005-03-21: </STRONG><B>Expanded Cerebellum Data Sets</B> released on the www.genenetwork.org/search3.html test site (MAS5, RMA, and PDNN transforms). Fifty-four new samples have been added. We have tested these data using probe set 1418908_at_A (<I>Pam</I>), a known cis-QTL in multiple data sets. This trait generates a peak LRS score of 31.7 in the initial March 2003 data (MAS5), a peak LRS score of 32.3 in the October 04 data (MAS5), and a peak LRS of 52.2 in the March 2005 (MAS5). In the March 2005 data, the peak LRS is only 500 Kb from the 5' promoter region of the <I>Pam</I> gene. The abundantly expressed GABA alpha 6 receptor (<I>Gabra6</I>)  transcript  (1417121_at_A) is another good test case of a cis modulated trait in cerebellum. (Implemented by the GeneNetwork group and the Cerebellum Consortium). The Heritability Weight Transform (HWT) data set will be added in the next several weeks. 
</BLOCKQUOTE>


<BLOCKQUOTE><STRONG>2005-03-15: </STRONG><B>Cluster Trees</B> now compute and display up to 100 traits simultaneously. This makes it possible to select the top 100 covariates of a trait from a Correlation Results table and map all 100 as a hierarchically organized group. (Implementation by Jintao Wang).
</BLOCKQUOTE>


<BLOCKQUOTE><STRONG>2005-03-04: </STRONG><B><A HREF="http://www.genenetwork.org/literatureCorrelation.html"  target="empty" class="fs14">Literature Correlation</A></B> data set has been integrated into GeneNetwork <B>Correlation Results</B> output tables. This important new feature provides an estimate of the strength of relations between pairs of genes that is based on a textual analysis of PubMed abstracts (latent semantic index correlations). Values are based on a matrix of 16,000 gene-gene simlarity scores computed by Ramin Homayouni (UTHSC) and Michael Berry (UT Knoxville). This feature is still experimental, and GeneNetwork users should note that pairs of genes that are mentioned together in a small set of papers may have inappropriately high correlations. For more information on the algorithm please contact Ramin Homayouni. (Implementation by Ramin Homayouni and Jintao Wang).
</BLOCKQUOTE>



<BLOCKQUOTE><STRONG>2005-03-01: </STRONG><B>Network Graph</B> output has been improved significantly. It is now possible to change the labels from probe set IDs to gene symbols. Nodes can also be color-coded by database. Markers and genotypes can be used as nodes. Literature Correlations can be used to define the lines (edges) between traits. (Implementation by Jintao Wang).
</BLOCKQUOTE>


<BLOCKQUOTE><STRONG>2005-03-01: </STRONG><B>Heritability Weighted Transform</B> method has been published at <A href="http://genomebiology.com/2005/6/3/R27"  target="empty" class="fs14"><I>Genome Biology</I></A>. This method (HWT1PM) provides significantly higher signal than other common transforms. (Design and implemenation by Ken Manly)
</BLOCKQUOTE>


<BLOCKQUOTE><STRONG>2005-02-23: </STRONG><B>Database Schema</B> has been published online at <A HREF="http://www.genenetwork.org/schema.html" target="empty" class="fs14">http://www.genenetwork.org/schema.html</A>. This schema (January 2005 version) was generated using MySQLdump v 9.1.  (Implemenation by Jintao Wang, Bill Bug, and Ken Manly)
</BLOCKQUOTE>

<BLOCKQUOTE><STRONG>2005-02-23: </STRONG><B>Scriptable Interface</B> improved to handle queries from Genome Browser and other systems. The new interface provides a list of links to data from multiple tissues and strains for a single gene. For example, to retrieve expression estimates for <I>Kcnj8</I> the URL query has this form: <A HREF="http://www.genenetwork.org/cgi-bin/beta/main.py?cmd=search&gene=kcnj8" target="empty" class="fs14">http://www.genenetwork.org/cgi-bin/beta/main.py?cmd=search&gene=kcnj8</A>. This query does not resolve the many possible aliases for gene symbols, and requires the use of the preferred or official gene symbol. (RWW, implementation by Jintao Wang)
</BLOCKQUOTE>

<BLOCKQUOTE><STRONG>2005-01-27: </STRONG><B>QTL Reaper</B> 1.0.0 has been released. QTL Reaper is platform-independent program  for rapidly mapping thousands of traits. It is now available to advanced users at <A HREF="http://qtlreaper.sourceforge.net/" target="empty" class="fs14"> SourceForge</A> (241 KB, written in Python and C with sample and help files). QTL Reaper can map well over 50,000 traits in under 12 hours on fast single-processor systems. It includes a sophisticated method (Besage, 1991) to adjust the number of permutation tests to estimate genome-wide p values with reasonable precision down to values of approximately 10^-5 (10^6 permutations). This feature is useful for identifying reproducible QTLs in large transcriptome data sets, that is, sets of QTLs with defined false discovery rates. (Design by Ken Manly, implemenation by Jintao Wang)
<DIR><SMALL>
Besag J, and Clifford P (1991). Sequential Monte Carlo p-values. Biometrika 78: 301-304.
</SMALL></DIR>
</BLOCKQUOTE>


<BLOCKQUOTE><STRONG>2005-01-26: </STRONG> The <B>Pair-scan</B> output tables now include a new analytic tool that provides a breakdown of strains in each genotype category (for example, the four two-locus genotypes: B/B, B/D, D/B, and D/D) either in the form of scatter plots or in the form of a box plot. This new feature is still being tested and refined and is currently available only on the test site (www.genenetwork.org/search3.html). This feature will be moved to the public site in February. (Implemenation by Jintao Wang)
</BLOCKQUOTE>


<BLOCKQUOTE><STRONG>2005-01-22: </STRONG><B>Marker Genotype Databases</B> have been added that complement trait and transcriptome databases for the following groups: AKXD, AXB/BXA, CXB, BXH, BXD, LXS, B6D2F2, and the rat HXB/BXH. These new databases enable you to use any marker genotype as a "trait" to search for transcripts or classical phenotypes that may be influenced by particular genomic regions. This is now possible using the new Genotype databases and the <B>Compare Correlates</B> tool. To find all markers on Chromosome 1 just type in "Chr 1" or "Chromosome 1" into the <B>Search</B> field. These maker genotype databases are currently available on the test site (www.genenetwork.org/search3.html) but will be moved to the public site by late January. (Implemenation by Jing Gu, Lu Lu, Yanhua Qu, Rob Williams, and Jintao Wang)
</BLOCKQUOTE>


<BLOCKQUOTE><STRONG>2005-01-21: </STRONG>New Data Download feature has been added. The <B>Information</B> files for most UTHSC Brain databases (e.g., the <A HREF="http://www.genenetwork.org/dbdoc/BR_U_0304_RR.html" target="empty" class="fs14">RMA Orig</A> transform) now have links to Excel workbooks that include the full Affymetrix U74Av2 data set of 100 arrays for each transform. These Excel workbooks also include a separate spreadsheet with the strain averages for each transform. Look for the word "Download" in the Information pages. (Implemenation by Yanhua Qu)
</BLOCKQUOTE>


<BLOCKQUOTE><STRONG>2005-01-13: </STRONG>We have added a new BLAST probe analysis tool to the<B> Probe Information</B> tables associaed with each Affymetrix probe set. This button-tool aligns any PM 25-mer probe to the GenBank sequence that Affymetrix lists as being the sequence source. When BLAT analysis of concatenated probes does not provide an unequivocal map location for a probe set, this method can be used to verify that the GenBank accession is correct. If so, it may then be appropriate to BLAT the entire GenBank entry to verify probe set map location. (Implemenation by Yanhua Qu)
</BLOCKQUOTE>

<BLOCKQUOTE><STRONG>2005-01-11: </STRONG>Rat HXB/BXH Published Phenotype databases added to the GeneNetwork. The genetic maps that are used in combination with these phenotypes are based on a total of 770 markers. Phenotypes were all provided by Michal Pravenec. We thank Tim Aitman and Pierre Mormede for review of their data sets. (Implementation by RWW, MP, and JW)
</BLOCKQUOTE>


<BLOCKQUOTE>
<STRONG>2005-01-03: </STRONG>We now provide links to entire data files for the U74Av2 brain data set. All DAT, CEL, TXT, RPT, and EXP files can be downloaded. For example, here are data files for five <A HREF="http://www.genenetwork.org/webqtl/main.py?FormID=showStrainInfo&database=BR_U_0304_RR&strain=C57BL/6J" target="empty" class="fs14">C57BL/6J</A> U74Av2 arrays. The complete U74Av2 data set consists of a total of 100 arrays, all of which can be reached from the <A HREF="http://www.genenetwork.org/dbdoc/BR_U_0304_RR.html"  target="empty" class="fs14">Main Table</A> in any of the <B>Information Pages</B> for these different transforms (MAS5, RMA, PDNN, HWT1PM, dChip). The DAT, CEL, RPT and EXP files will be identical among all transforms. The only differences among transforms are the TXT files. The appropriate reference to cite if you make use of these data files is:
<DIR><SMALL>
Chesler EJ, Lu L, Shou S, Qu Y, Gu J, Wang J, Hsu HC, Mountz JD, Baldwin N, Langston MA, Threadgill DW, Manly KF, Williams RW (2005) Genetic dissection of gene expression reveals polygenic and pleiotropic networks modulating brain structure and function. Nature Genetics 37: 233-42.
</SMALL></DIR>

</BLOCKQUOTE>
<p></P>

<BLOCKQUOTE>
<STRONG>2004-12-25: </STRONG>First draft of the <A HREF="http://www.genenetwork.org/glossary.html" class="fs14" target="_new">WebQTL Glossary</A> is completed. Many key terms are now defined. We will be adding links to the glossary from graphs and other pages.</BLOCKQUOTE>
<p></P>

<BLOCKQUOTE>
<STRONG>2004-12-22: </STRONG>An annotated <A HREF="http://www.genenetwork.org/links.html" class="fs14" target="_new">Links</A> list has been added. Email RW Williams at <rwilliams@nb.utmem.edu> if you have suggestions for additional sites that have proved useful in combination with GeneNetwork resources.</BLOCKQUOTE>
<p></P>

<BLOCKQUOTE>
<STRONG>2004-12-21: </STRONG>We have implemented a new method of transforming Affymetrix microarray data called the Heritability Weighted Transform (Manly et al. 2005). When used with large Affymetrix data sets of the type used by WebQTL, this method is considerably more powerful than other common probes-to-probeset transform such as MAS5, PDNN, RMA, or dChip.  To evaluate this new method please try the Mouse/BXD/Brain/Database called <B>UTHSC Brain mRNA U74Av2 (Dec03) HWT1PM</B> (HWT1PM is short for Heritability Weighted Transform Version 1, Perfect Match Probes only). For further detals on this method see the <H REF="http://www.genenetwork.org/dbdoc/BR_U_1203_H2.html" TARGET="_empty" class="fs14">Info</A> page. The reference for this approach to transforming Affymetrix array data is:

<DIR> <SMALL>
Manly KF, Wang J, Williams RW (2005) Weighting by heritability for detection of quantitative trait loci with microarray estimates of gene expression. Genome Biology 6: R27.
</SMALL></DIR>
</BLOCKQUOTE>
<p></P>

<BLOCKQUOTE>
<STRONG>2004-12-17: </STRONG>We have added mouse UniGene identifiers from <A HREF="http://www.ncbi.nlm.nih.gov/UniGene/UGOrg.cgi?TAXID=10090" target="_empty" class="fs14">Build 142</A>. It is therefore now possible to enter search terms such as "Mm.1" to find data on S100 calcium binding protein A10 (S100a10). A total of 38,034 probe sets on the Affymetrix mouse expression array 430 2.0, have UniGene identifiers.</BLOCKQUOTE>
<p></P>

<BLOCKQUOTE>
<STRONG>2004-12-14: </STRONG>First draft of the <A HREF="http://www.genenetwork.org/faq.html" class="fs14" target="_new">WebQTL Frequently Asked Questions</A> is completed. We be happy to answer any other questions you have. Please email RW Williams at <rwilliams@nb.utmem.edu>.</BLOCKQUOTE>
<p></P>

<BLOCKQUOTE>
<STRONG>2004-12-13: </STRONG>Major additions are expected later in December in both the SJUT Cerebellum data set and in the INIA Brain data set. Sample size will be almost doubled in both data sets.</BLOCKQUOTE>
<p></P>

<BLOCKQUOTE>
<STRONG>2004-12-10: </STRONG>Updated positions of Mouse Expression Aglient  G4121A probe using the May 2004 (mm5) assembly of the mouse genome. This work was carried out by Yanhua Qu. </BLOCKQUOTE>
<p></P>

<BLOCKQUOTE>
<STRONG>2004-12-10: </STRONG>We have begun to combine WebQTL and The GeneNetwork. WebQTL is the first and so far only "channel" of the GeneNetwork. However, our hope is that there will soon be other projects that will share use of the GeneNetwork. The main URL is now www.genenetwork.org. Requests to www.webqtl.org will resolve to www.genenetwork.org.
</BLOCKQUOTE>
<p></P>

<Blockquote><STRONG>2004-12-03: </STRONG>Rat HXB/BXH genotype and published phenotype databases added to beta test site of WebQTL. The genetic maps are based on a total of 770 markers. Phenotypes were all provided by Dr. Michal Pravenec. 
</BLOCKQUOTE>

<p></P>

<Blockquote><STRONG>2004-12-02: </STRONG>Important new graphic and analytic tools have been added. 

<P>The first of these is the <B>Compare Correlates</B> tool. This function is available in Selection Windows. It is essentially a Venn diagram set tool. Instead of providing simple graphs, it provides lists of traits in different parts of a virtual Venn diagram. For example, to find traits that covary with Sonic Hedgehog, Indian Hedgehog, Desert Hedgehog, Patched1, and Gli3, you would select five key transcripts into a <B>Selections</B> window (use the "Add Selection" tool and then select a group of traits in the Selections window). Compare Correlates allows you to chose the target database to which the key traits will be correlated. Compare Correlates was designed by Elissa Chesler and Stephen Pitts. Code was written and optimized by Stephen Pitts. </P>

<P>The second new tool is <B>Network Graph</B>. This function displays a set of traits and their correlations in the form of a graph with nodes (traits) and lines (correlations). There are quite a few tunable parameters, including the correlation threshold used to draw (or not draw) a line between nodes. To use this new tool, you again need to have traits loaded into one of the Selections windows. Network Graph was designed by Elissa Chesler and Stephen Pitts. Code was written, optimized, and error-checked by Stephen Pitts.</P>

</BLOCKQUOTE>

<Blockquote><STRONG>2004-10-24: </STRONG>Updated positions of all Mouse Expression U74Av2, 430A, 430B, and 430 2.0 probe sets using the May 2004 (mm5) assembly of the mouse genome. This work was carried out by Yanhua Qu. The M430 data consists of 45,000 probe sets. Positions were obtained using a series of methods: Method 1. A BLAT analysis of the actual probe sequence using a 48-processor cluster (our thanks to Yan Cui). Roughly 90% of all probe sets were mapped using this method. If the probe sequence did not BLAT with a score above 99 AND an identity match of 100, then we used Method 2: We used the position of the probe set given in the <A HREF="http://hgdownload.cse.ucsc.edu/goldenPath/mm5/database/UCSC" target="_blank" class="fs14"> affMOE430.txt.gz</A> data file. This method recovered position data for approximaely 5% of all probe sets. If Method 2 failed, then we used Method 3: We obtained the position given by <A HREF="http://www.affymetrix.com/support/technical/byproduct.affx?product=moe430" target="_blank" class="fs14">Affymetrix</A> in the files called "MOE430A Annotations, CSV (6.3 Mb, 10/12/04)" and  "MOE430B Annotations, CSV (3.9 Mb, 10/12/04)". This method recovered positions on roughly 4%. As a last resort we used Method 4: We retained position data from mm4 or mm3 without interpolation. No position data would be found for 198 records and no chromosome could be found for 46 probe sets.  We estimate that 5 to 10% of position data are unreliable.

</Blockquote>
<p></P>


<Blockquote><STRONG>2004-10-16: </STRONG>Expression data set for the striatum of BXD strains released by Glenn Rosen to the www.webqtl.org/search3.html beta site. This is the first WebQTL database that exploits the Mouse Expression 430 2.0 array from Affymetrix. Four versions were released: MAS5, RMA, PDNN, and the new GCRMA. </Blockquote>
<p></P>


<Blockquote><STRONG>2004-10-11: </STRONG> New hierarchical <B>Search Page</B> interface released to main site (Choose species, cross, type, and database). New <B>Info</B> pages released. More complete annotation and explanation of the use of the pair-scan data is now provided when the "permutation" option is  selected in the <B>Analysis Tools</B> area of the <B>Trait Data and Editing Form</B>. </Blockquote>

<p></P>

<Blockquote><STRONG>2004-09-22: </STRONG> Pair-scan feature is now zoomable. Click on any single chromosome pair region to zoom in. </Blockquote><p></P>



			<Blockquote><STRONG>2004-08-20: </STRONG> Pair-scan permutation test is now available, it takes 90 seconds to do 500 permutations. </Blockquote><p></P>
			<Blockquote><STRONG>2004-07-15: </STRONG> New Pair-scan searches for pairs of chromosomal regions that may be involved in two-locus epistatic interactions is added to WebQTL</Blockquote><p></P>
			<Blockquote><STRONG>2004-06-07: </STRONG> Interval mapping graph in 2X resolution is now available for downloading. </Blockquote><p></P>
			<Blockquote><STRONG>2004-06-02: </STRONG> Three new B6D2F2 database are added to WebQTL. Dominance estimation for interval mapping with F2 data is available.</Blockquote><p></P>
			<Blockquote><STRONG>2004-05-03: </STRONG> Cluster qtl map display is added to WebQTL. These QTL heat maps can be drawn using three different color assignments. </Blockquote><p></P>
			<Blockquote><STRONG>2004-03-18: </STRONG> User is now able to add their own traits to selections, the correlation matrix and multiple mapping and some other features can be included for those traits. </Blockquote>
<P></P>

<BLOCKQUOTE><STRONG>2005-07-15: </STRONG> Database <A HREF="http://www.genenetwork.org/webqtl/main.py?FormID=admin" target="_blank" class="fs14">List Selector</A>  has been implemented for the administrator. This facility is used to select the best databases to use by external resources that link to the GeneNetwork. (Implemented by Jintao Wang.)
</BLOCKQUOTE>

<Blockquote class="subtitle">Information about this text file:

<Blockquote><P>This text file originally generated by RWW, March 2004. 
</P></Blockquote>
</Blockquote>

</span>
<p></P>
			
		</TD>
		</TR></TABLE>
		</TD>
	</TR>
	<TR>
		<TD align=center bgColor=#ddddff class="solidBorder">
		<!--Start of footer-->
		<TABLE width="90%">
		<script language='JavaScript' src='/javascript/footer.js'></script>
		</TABLE>
		<!--End of footer-->
		</TD>
	</TR>
</TABLE>
<!-- /Footer -->
<script language="JavaScript" src="/javascript/menu_new.js"></script>
<script language="JavaScript" src="/javascript/menu_items.js"></script>
<script language="JavaScript" src="/javascript/menu_tpl.js"></script>
<script language="JavaScript">
	<!--//
	new menu (MENU_ITEMS, MENU_POS);
	//-->
</script>
<script src="http://www.google-analytics.com/urchin.js" type="text/javascript">
</script>
<script type="text/javascript">
_uacct = "UA-3782271-1";
urchinTracker();
</script>
</BODY>
</HTML>
